// @teambit/react/dist/templates/react-workspace-data-fetching/index.js
"use strict";

require("core-js/modules/es.promise.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.reactWorkspaceDataFetchingTemplate = void 0;
function _workspaceCommon() {
  const data = require("../workspace-common");
  _workspaceCommon = function () {
    return data;
  };
  return data;
}
const reactWorkspaceDataFetchingTemplate = {
  name: 'react-data-fetching',
  description: 'React workspace with components to show data fetching',
  generateFiles: async context => {
    return (0, _workspaceCommon().generateFiles)(context);
  },
  fork: () => {
    return [{
      id: 'learn-bit-react.data-fetching/pages/books-page',
      targetName: 'pages/books-page',
      path: 'pages/books-page'
    }, {
      id: 'learn-bit-react.data-fetching/ui/hooks/use-books',
      targetName: 'ui/hooks/use-books',
      path: 'ui/hooks/use-books'
    }, {
      id: 'learn-bit-react.data-fetching/ui/book',
      targetName: 'ui/book',
      path: 'ui/book'
    }, {
      id: 'learn-bit-react.data-fetching/ui/book-list',
      targetName: 'ui/book-list',
      path: 'ui/book-list'
    }, {
      id: 'learn-bit-react.data-fetching/models/book',
      targetName: 'models/book',
      path: 'models/book'
    }];
  }
};
exports.reactWorkspaceDataFetchingTemplate = reactWorkspaceDataFetchingTemplate;
