// @teambit/component/dist/show/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ExtensionsFragment", {
  enumerable: true,
  get: function () {
    return _extensions().ExtensionsFragment;
  }
});
Object.defineProperty(exports, "FilesFragment", {
  enumerable: true,
  get: function () {
    return _files().FilesFragment;
  }
});
Object.defineProperty(exports, "IDFragment", {
  enumerable: true,
  get: function () {
    return _id().IDFragment;
  }
});
Object.defineProperty(exports, "MainFileFragment", {
  enumerable: true,
  get: function () {
    return _mainFile().MainFileFragment;
  }
});
Object.defineProperty(exports, "NameFragment", {
  enumerable: true,
  get: function () {
    return _name().NameFragment;
  }
});
Object.defineProperty(exports, "ScopeFragment", {
  enumerable: true,
  get: function () {
    return _scope().ScopeFragment;
  }
});
Object.defineProperty(exports, "ShowCmd", {
  enumerable: true,
  get: function () {
    return _show().ShowCmd;
  }
});
Object.defineProperty(exports, "ShowFragment", {
  enumerable: true,
  get: function () {
    return _showFragment().ShowFragment;
  }
});
Object.defineProperty(exports, "ShowJSONRow", {
  enumerable: true,
  get: function () {
    return _showFragment().JSONRow;
  }
});
Object.defineProperty(exports, "ShowRow", {
  enumerable: true,
  get: function () {
    return _showFragment().ShowRow;
  }
});
function _showFragment() {
  const data = require("./show-fragment");
  _showFragment = function () {
    return data;
  };
  return data;
}
function _show() {
  const data = require("./show.cmd");
  _show = function () {
    return data;
  };
  return data;
}
function _mainFile() {
  const data = require("./main-file.fragment");
  _mainFile = function () {
    return data;
  };
  return data;
}
function _name() {
  const data = require("./name.fragment");
  _name = function () {
    return data;
  };
  return data;
}
function _id() {
  const data = require("./id.fragment");
  _id = function () {
    return data;
  };
  return data;
}
function _scope() {
  const data = require("./scope.fragment");
  _scope = function () {
    return data;
  };
  return data;
}
function _files() {
  const data = require("./files.fragment");
  _files = function () {
    return data;
  };
  return data;
}
function _extensions() {
  const data = require("./extensions.fragment");
  _extensions = function () {
    return data;
  };
  return data;
}
