// @teambit/compositions/dist/ui/compositions-panel/compositions-panel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CompositionsPanel = CompositionsPanel;
function _extends2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
  _extends2 = function () {
    return data;
  };
  return data;
}
function _objectWithoutProperties2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
  _objectWithoutProperties2 = function () {
    return data;
  };
  return data;
}
function _evangelistElements() {
  const data = require("@teambit/evangelist.elements.icon");
  _evangelistElements = function () {
    return data;
  };
  return data;
}
function _classnames() {
  const data = _interopRequireDefault(require("classnames"));
  _classnames = function () {
    return data;
  };
  return data;
}
function _react() {
  const data = _interopRequireWildcard(require("react"));
  _react = function () {
    return data;
  };
  return data;
}
function _compositionsPanelModule() {
  const data = _interopRequireDefault(require("./compositions-panel.module.scss"));
  _compositionsPanelModule = function () {
    return data;
  };
  return data;
}
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function CompositionsPanel(_ref) {
  let {
      url,
      compositions,
      isScaling,
      onSelectComposition: onSelect,
      active,
      className
    } = _ref,
    rest = (0, _objectWithoutProperties2().default)(_ref, ["url", "compositions", "isScaling", "onSelectComposition", "active", "className"]);
  const handleSelect = (0, _react().useCallback)(selected => {
    onSelect && onSelect(selected);
  }, [onSelect]);
  return /*#__PURE__*/_react().default.createElement("ul", (0, _extends2().default)({}, rest, {
    className: (0, _classnames().default)(className)
  }), compositions.map(composition => {
    const href = isScaling ? `${url}&name=${composition.identifier}` : `${url}&${composition.identifier}`;

    // TODO - move to composition panel node
    return /*#__PURE__*/_react().default.createElement("li", {
      key: composition.identifier,
      className: (0, _classnames().default)(_compositionsPanelModule().default.linkWrapper, composition === active && _compositionsPanelModule().default.active)
    }, /*#__PURE__*/_react().default.createElement("a", {
      className: _compositionsPanelModule().default.panelLink,
      onClick: () => handleSelect(composition)
    }, /*#__PURE__*/_react().default.createElement("span", {
      className: _compositionsPanelModule().default.box
    }), /*#__PURE__*/_react().default.createElement("span", {
      className: _compositionsPanelModule().default.name
    }, composition.displayName)), /*#__PURE__*/_react().default.createElement("div", {
      className: _compositionsPanelModule().default.right
    }, /*#__PURE__*/_react().default.createElement("a", {
      className: _compositionsPanelModule().default.panelLink,
      target: "_blank",
      rel: "noopener noreferrer",
      href: href
    }, /*#__PURE__*/_react().default.createElement(_evangelistElements().Icon, {
      className: _compositionsPanelModule().default.icon,
      of: "open-tab"
    }))));
  }));
}
