// @teambit/compiler/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  CompilerAspect: true,
  Compiler: true,
  CompilerOptions: true,
  CompilationInitiator: true,
  TranspileFileOutput: true,
  TranspileFileParams: true,
  TranspileComponentParams: true
};
Object.defineProperty(exports, "CompilationInitiator", {
  enumerable: true,
  get: function () {
    return _types().CompilationInitiator;
  }
});
Object.defineProperty(exports, "Compiler", {
  enumerable: true,
  get: function () {
    return _types().Compiler;
  }
});
Object.defineProperty(exports, "CompilerAspect", {
  enumerable: true,
  get: function () {
    return _compiler().CompilerAspect;
  }
});
Object.defineProperty(exports, "CompilerOptions", {
  enumerable: true,
  get: function () {
    return _types().CompilerOptions;
  }
});
Object.defineProperty(exports, "TranspileComponentParams", {
  enumerable: true,
  get: function () {
    return _types().TranspileComponentParams;
  }
});
Object.defineProperty(exports, "TranspileFileOutput", {
  enumerable: true,
  get: function () {
    return _types().TranspileFileOutput;
  }
});
Object.defineProperty(exports, "TranspileFileParams", {
  enumerable: true,
  get: function () {
    return _types().TranspileFileParams;
  }
});
function _compiler() {
  const data = require("./compiler.aspect");
  _compiler = function () {
    return data;
  };
  return data;
}
function _types() {
  const data = require("./types");
  _types = function () {
    return data;
  };
  return data;
}
var _events = require("./events");
Object.keys(_events).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _events[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _events[key];
    }
  });
});
