// @teambit/compiler/dist/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CompilationInitiator = void 0;
let CompilationInitiator;
exports.CompilationInitiator = CompilationInitiator;
(function (CompilationInitiator) {
  CompilationInitiator[CompilationInitiator["CmdReport"] = 0] = "CmdReport";
  CompilationInitiator[CompilationInitiator["CmdJson"] = 1] = "CmdJson";
  CompilationInitiator[CompilationInitiator["PreStart"] = 2] = "PreStart";
  CompilationInitiator[CompilationInitiator["PreWatch"] = 3] = "PreWatch";
  CompilationInitiator[CompilationInitiator["Start"] = 4] = "Start";
  CompilationInitiator[CompilationInitiator["ComponentChanged"] = 5] = "ComponentChanged";
  CompilationInitiator[CompilationInitiator["AspectLoadFail"] = 6] = "AspectLoadFail";
  CompilationInitiator[CompilationInitiator["ComponentAdded"] = 7] = "ComponentAdded";
  CompilationInitiator[CompilationInitiator["Install"] = 8] = "Install";
})(CompilationInitiator || (exports.CompilationInitiator = CompilationInitiator = {}));
