// @teambit/changelog/dist/changelog.section.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChangelogSection = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _react() {
  const data = _interopRequireDefault(require("react"));
  _react = function () {
    return data;
  };
  return data;
}
function _uiFoundationUi() {
  const data = require("@teambit/ui-foundation.ui.menu-widget-icon");
  _uiFoundationUi = function () {
    return data;
  };
  return data;
}
function _changeLogPage() {
  const data = require("./ui/change-log-page");
  _changeLogPage = function () {
    return data;
  };
  return data;
}
class ChangelogSection {
  constructor() {
    (0, _defineProperty2().default)(this, "route", {
      path: '~changelog',
      element: /*#__PURE__*/_react().default.createElement(_changeLogPage().ChangeLogPage, null)
    });
    (0, _defineProperty2().default)(this, "navigationLink", {
      href: '~changelog',
      children: /*#__PURE__*/_react().default.createElement(_uiFoundationUi().MenuWidgetIcon, {
        icon: "changelog",
        tooltipContent: "Change log"
      }),
      displayName: 'Change log'
    });
    (0, _defineProperty2().default)(this, "order", 40);
  }
}
exports.ChangelogSection = ChangelogSection;
