// @teambit/preview/dist/execution-ref.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ExecutionRef = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
// TODO - use workspace.list() instead of this
class ExecutionRef {
  constructor(executionCtx) {
    this.executionCtx = executionCtx;
    (0, _defineProperty2().default)(this, "currentComponents", void 0);
    this.currentComponents = executionCtx.components;
  }
  add(added) {
    this.currentComponents = this.currentComponents.concat(added);
  }
  remove(removed) {
    this.currentComponents = this.currentComponents.filter(c => c.id.toString() !== removed.toString());
  }
  update(next) {
    this.currentComponents = this.currentComponents.map(c => c.equals(next) ? next : c);
  }
  get(id) {
    return this.currentComponents.find(x => x.id.isEqual(id));
  }
}
exports.ExecutionRef = ExecutionRef;
