// @teambit/preview/dist/preview.graphql.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.previewSchema = previewSchema;
function _graphqlTag() {
  const data = _interopRequireDefault(require("graphql-tag"));
  _graphqlTag = function () {
    return data;
  };
  return data;
}
function previewSchema(previewExtension) {
  return {
    typeDefs: (0, _graphqlTag().default)`
      type Preview {
        # url: String!
        """
        Check if the component supports scaling
        """
        isScaling: Boolean
        includesEnvTemplate: Boolean
        legacyHeader: Boolean
      }

      extend type Component {
        preview: Preview
      }
    `,
    resolvers: {
      Component: {
        preview: component => {
          // return previewExtension.getPreview(component);
          return {
            component
          };
        }
      },
      Preview: {
        includesEnvTemplate: ({
          component
        }) => {
          return previewExtension.isBundledWithEnv(component);
        },
        isScaling: ({
          component
        }) => {
          return previewExtension.doesScaling(component);
        },
        legacyHeader: ({
          component
        }) => {
          return previewExtension.isLegacyHeader(component);
        }
      }
    }
  };
}
