// @teambit/react/dist/templates/react-context/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.reactContext = void 0;
function _context() {
  const data = require("./files/context");
  _context = function () {
    return data;
  };
  return data;
}
function _contextProvider() {
  const data = require("./files/context-provider");
  _contextProvider = function () {
    return data;
  };
  return data;
}
function _composition() {
  const data = require("./files/composition");
  _composition = function () {
    return data;
  };
  return data;
}
function _docs() {
  const data = require("./files/docs");
  _docs = function () {
    return data;
  };
  return data;
}
function _test() {
  const data = require("./files/test");
  _test = function () {
    return data;
  };
  return data;
}
function _indexFile() {
  const data = require("./files/index-file");
  _indexFile = function () {
    return data;
  };
  return data;
}
const reactContext = {
  name: 'react-context',
  description: 'a react context component',
  generateFiles: context => {
    return [(0, _indexFile().indexFile)(context), (0, _context().contextFile)(context), (0, _contextProvider().contextProviderFile)(context), (0, _composition().compositionFile)(context), (0, _docs().docsFile)(context), (0, _test().testFile)(context)];
  },
  config: {
    'teambit.react/react': {},
    'teambit.envs/envs': {
      env: 'teambit.react/react'
    }
  }
};
exports.reactContext = reactContext;
