// @teambit/react-native/dist/templates/react-native-component/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.reactNativeComponent = void 0;
function _component() {
  const data = require("./files/component");
  _component = function () {
    return data;
  };
  return data;
}
function _composition() {
  const data = require("./files/composition");
  _composition = function () {
    return data;
  };
  return data;
}
function _docs() {
  const data = require("./files/docs");
  _docs = function () {
    return data;
  };
  return data;
}
function _test() {
  const data = require("./files/test");
  _test = function () {
    return data;
  };
  return data;
}
function _index() {
  const data = require("./files/index");
  _index = function () {
    return data;
  };
  return data;
}
const reactNativeComponent = {
  name: 'react-native',
  description: 'a basic react native component',
  generateFiles: context => {
    return [(0, _index().indexFile)(context), (0, _component().componentFile)(context), (0, _composition().compositionFile)(context), (0, _docs().docsFile)(context), (0, _test().testFile)(context)];
  },
  config: {
    'teambit.react/react-native': {},
    'teambit.envs/envs': {
      env: 'teambit.react/react-native'
    }
  }
};
exports.reactNativeComponent = reactNativeComponent;
