// @teambit/elements/dist/elements-artifact.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ElementsArtifact = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _utils() {
  const data = require("@teambit/legacy/dist/utils");
  _utils = function () {
    return data;
  };
  return data;
}
class ElementsArtifact {
  constructor(artifacts) {
    this.artifacts = artifacts;
  }
  isEmpty() {
    return !this.artifacts.length;
  }
  getFile(path) {
    return this.artifacts.find(file => {
      return (0, _utils().pathNormalizeToLinux)(file.relative) === path;
    });
  }
  getDefaultMainFilePrefix() {
    return ElementsArtifact.defaultMainFilePrefix;
  }
  getMainElementsBundleFile() {
    return this.artifacts.find(file => {
      return (0, _utils().pathNormalizeToLinux)(file.relative).includes(`${this.getDefaultMainFilePrefix()}.`);
    });
  }
  getMainElementsFileUrl() {
    const mainFile = this.artifacts.find(file => {
      return (0, _utils().pathNormalizeToLinux)(file.relative).includes(`${this.getDefaultMainFilePrefix()}.`);
    });
    return mainFile === null || mainFile === void 0 ? void 0 : mainFile.url;
  }
}
exports.ElementsArtifact = ElementsArtifact;
(0, _defineProperty2().default)(ElementsArtifact, "defaultMainFilePrefix", 'elements');
