// @teambit/user-agent/dist/user-agent.ui.runtime.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.array.iterator.js");
require("core-js/modules/es.promise.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UserAgentUI = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _ui() {
  const data = require("@teambit/ui");
  _ui = function () {
    return data;
  };
  return data;
}
function _react() {
  const data = _interopRequireDefault(require("react"));
  _react = function () {
    return data;
  };
  return data;
}
function _uaParserJs() {
  const data = _interopRequireDefault(require("ua-parser-js"));
  _uaParserJs = function () {
    return data;
  };
  return data;
}
function _uiFoundationUiHooks() {
  const data = require("@teambit/ui-foundation.ui.hooks.use-user-agent");
  _uiFoundationUiHooks = function () {
    return data;
  };
  return data;
}
function _userAgent() {
  const data = require("./user-agent.aspect");
  _userAgent = function () {
    return data;
  };
  return data;
}
/**
 * user agent aspect
 */
class UserAgentUI {
  static async provider([uiUi]) {
    const userAgentUi = new UserAgentUI();
    uiUi.registerRenderHooks({
      serverInit: ({
        browser
      }) => {
        const userAgent = new (_uaParserJs().default)(browser.headers['user-agent']);
        return {
          userAgent
        };
      },
      browserInit: () => {
        return {
          userAgent: new (_uaParserJs().default)(window.navigator.userAgent)
        };
      },
      reactContext: UserAgentReactContext
    });
    return userAgentUi;
  }
}
exports.UserAgentUI = UserAgentUI;
(0, _defineProperty2().default)(UserAgentUI, "slots", []);
(0, _defineProperty2().default)(UserAgentUI, "dependencies", [_ui().UIAspect]);
(0, _defineProperty2().default)(UserAgentUI, "runtime", _ui().UIRuntime);
function UserAgentReactContext({
  children,
  renderCtx
}) {
  return /*#__PURE__*/_react().default.createElement(_uiFoundationUiHooks().UserAgentProvider, {
    value: renderCtx === null || renderCtx === void 0 ? void 0 : renderCtx.userAgent
  }, children);
}
_userAgent().UserAgentAspect.addRuntime(UserAgentUI);
