// @teambit/react-router/dist/route-context.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RootRoute = RootRoute;
exports.RouteContext = RouteContext;
function _react() {
  const data = _interopRequireDefault(require("react"));
  _react = function () {
    return data;
  };
  return data;
}
function _reactRouterDom() {
  const data = require("react-router-dom");
  _reactRouterDom = function () {
    return data;
  };
  return data;
}
function _server() {
  const data = require("react-router-dom/server");
  _server = function () {
    return data;
  };
  return data;
}
function _uiFoundationUiReactRouter() {
  const data = require("@teambit/ui-foundation.ui.react-router.slot-router");
  _uiFoundationUiReactRouter = function () {
    return data;
  };
  return data;
}
function _baseReactNavigation() {
  const data = require("@teambit/base-react.navigation.link");
  _baseReactNavigation = function () {
    return data;
  };
  return data;
}
function _uiFoundationUiNavigation() {
  const data = require("@teambit/ui-foundation.ui.navigation.react-router-adapter");
  _uiFoundationUiNavigation = function () {
    return data;
  };
  return data;
}
function _routingMethod() {
  const data = require("./routing-method");
  _routingMethod = function () {
    return data;
  };
  return data;
}
/**
 * Setup context needed for routing.
 */
function RouteContext({
  routing = _routingMethod().Routing.url,
  children,
  location
}) {
  return /*#__PURE__*/_react().default.createElement(Router, {
    type: routing,
    location: location
  }, /*#__PURE__*/_react().default.createElement(_baseReactNavigation().NavigationProvider, {
    implementation: _uiFoundationUiNavigation().reactRouterAdapter
  }, children));
}
function RootRoute({
  rootRoutes,
  routeSlot
}) {
  return /*#__PURE__*/_react().default.createElement(_uiFoundationUiReactRouter().SlotRouter, {
    slot: routeSlot,
    rootRoutes: rootRoutes
  });
}

/** provides the router engine (browser, inMemory, etc) */
function Router({
  type,
  children,
  location
}) {
  switch (type) {
    case _routingMethod().Routing.static:
      return /*#__PURE__*/_react().default.createElement(_server().StaticRouter, {
        location: location || '/'
      }, children);
    case _routingMethod().Routing.inMemory:
      return /*#__PURE__*/_react().default.createElement(_reactRouterDom().MemoryRouter, {
        initialEntries: [location || '/'],
        initialIndex: 1
      }, children);
    case _routingMethod().Routing.hash:
      // @ts-ignore - https://github.com/teambit/bit/issues/5746
      return /*#__PURE__*/_react().default.createElement(_reactRouterDom().HashRouter, null, children);
    case _routingMethod().Routing.url:
    default:
      // @ts-ignore - https://github.com/teambit/bit/issues/5746
      return /*#__PURE__*/_react().default.createElement(_reactRouterDom().BrowserRouter, null, children);
  }
}
