// @teambit/cli/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  Command: true,
  CLIArgs: true,
  Flags: true,
  GenericObject: true,
  CommandOptions: true,
  CLIAspect: true,
  MainRuntime: true
};
Object.defineProperty(exports, "CLIArgs", {
  enumerable: true,
  get: function () {
    return _command().CLIArgs;
  }
});
Object.defineProperty(exports, "CLIAspect", {
  enumerable: true,
  get: function () {
    return _cli().CLIAspect;
  }
});
Object.defineProperty(exports, "Command", {
  enumerable: true,
  get: function () {
    return _command().Command;
  }
});
Object.defineProperty(exports, "CommandOptions", {
  enumerable: true,
  get: function () {
    return _legacyCommand().CommandOptions;
  }
});
Object.defineProperty(exports, "Flags", {
  enumerable: true,
  get: function () {
    return _command().Flags;
  }
});
Object.defineProperty(exports, "GenericObject", {
  enumerable: true,
  get: function () {
    return _command().GenericObject;
  }
});
Object.defineProperty(exports, "MainRuntime", {
  enumerable: true,
  get: function () {
    return _cli().MainRuntime;
  }
});
function _command() {
  const data = require("@teambit/legacy/dist/cli/command");
  _command = function () {
    return data;
  };
  return data;
}
function _legacyCommand() {
  const data = require("@teambit/legacy/dist/cli/legacy-command");
  _legacyCommand = function () {
    return data;
  };
  return data;
}
var _exceptions = require("./exceptions");
Object.keys(_exceptions).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _exceptions[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _exceptions[key];
    }
  });
});
function _cli() {
  const data = require("./cli.aspect");
  _cli = function () {
    return data;
  };
  return data;
}
