// @teambit/config/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ComponentScopeDirMap", {
  enumerable: true,
  get: function () {
    return _workspaceConfig().ComponentScopeDirMap;
  }
});
Object.defineProperty(exports, "ConfigAspect", {
  enumerable: true,
  get: function () {
    return _config().ConfigAspect;
  }
});
Object.defineProperty(exports, "ConfigMain", {
  enumerable: true,
  get: function () {
    return _configMain().ConfigMain;
  }
});
Object.defineProperty(exports, "ConfigRuntime", {
  enumerable: true,
  get: function () {
    return _config().ConfigRuntime;
  }
});
Object.defineProperty(exports, "getWorkspaceConfigTemplateParsed", {
  enumerable: true,
  get: function () {
    return _workspaceConfig().getWorkspaceConfigTemplateParsed;
  }
});
Object.defineProperty(exports, "stringifyWorkspaceConfig", {
  enumerable: true,
  get: function () {
    return _workspaceConfig().stringifyWorkspaceConfig;
  }
});
function _configMain() {
  const data = require("./config.main.runtime");
  _configMain = function () {
    return data;
  };
  return data;
}
function _config() {
  const data = require("./config.aspect");
  _config = function () {
    return data;
  };
  return data;
}
function _workspaceConfig() {
  const data = require("./workspace-config");
  _workspaceConfig = function () {
    return data;
  };
  return data;
}
