// @teambit/remove/dist/remove.fragment.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.promise.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RemoveFragment = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
class RemoveFragment {
  constructor(remove) {
    this.remove = remove;
    (0, _defineProperty2().default)(this, "title", 'removed');
    (0, _defineProperty2().default)(this, "weight", 3);
  }
  async renderRow(component) {
    const isRemoved = this.remove.isRemoved(component);
    return {
      title: this.title,
      // when it's not removed, set as an empty string so then it won't be shown in bit-show
      content: isRemoved ? isRemoved.toString() : ''
    };
  }
  async json(component) {
    return {
      title: this.title,
      json: this.remove.getRemoveInfo(component)
    };
  }
}
exports.RemoveFragment = RemoveFragment;
