// @teambit/html/dist/templates/html-env/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.htmlEnvTemplate = void 0;
function _index() {
  const data = require("./files/index");
  _index = function () {
    return data;
  };
  return data;
}
function _doc() {
  const data = require("./files/doc");
  _doc = function () {
    return data;
  };
  return data;
}
function _extension() {
  const data = require("./files/extension");
  _extension = function () {
    return data;
  };
  return data;
}
const htmlEnvTemplate = {
  name: 'html-env',
  description: 'customize the base Html env with your configs and tools',
  generateFiles: context => {
    return [{
      relativePath: 'index.ts',
      content: (0, _index().indexFile)(context),
      isMain: true
    }, {
      relativePath: `${context.name}.docs.mdx`,
      content: (0, _doc().docFile)()
    }, {
      relativePath: `${context.name}.extension.ts`,
      content: (0, _extension().extensionFile)(context)
    }];
  },
  config: {
    'teambit.envs/env': {},
    'teambit.envs/envs': {
      env: 'teambit.envs/env'
    }
  }
};
exports.htmlEnvTemplate = htmlEnvTemplate;
