// @teambit/react/dist/webpack/webpack.config.component.prod.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = _default;
function _componentId() {
  const data = require("@teambit/component-id");
  _componentId = function () {
    return data;
  };
  return data;
}
require("@teambit/react.babel.bit-react-transformer");
// Make sure the bit-react-transformer is a dependency
// TODO: remove it once we can set policy from component to component then set it via the component.json

// This is the production and development configuration.
// It is focused on developer experience, fast rebuilds, and a minimal bundle.
// eslint-disable-next-line complexity
function _default() {
  return {
    module: {
      rules: [{
        test: /\.js$/,
        include: [/node_modules/, /\/dist\//],
        exclude: /@teambit\/legacy/,
        descriptionData: {
          componentId: _componentId().ComponentID.isValidObject
        },
        use: [{
          loader: require.resolve('babel-loader'),
          options: {
            babelrc: false,
            configFile: false,
            plugins: [
            // for component highlighting in preview.
            [require.resolve('@teambit/react.babel.bit-react-transformer')]],
            // turn off all optimizations (only slow down for node_modules)
            compact: false,
            minified: false
          }
        }]
      }]
    }
  };
}
