// @teambit/legacy/dist/utils/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Queue", {
  enumerable: true,
  get: function () {
    return _queue().default;
  }
});
Object.defineProperty(exports, "bufferToReadStream", {
  enumerable: true,
  get: function () {
    return _toReadStream().default;
  }
});
Object.defineProperty(exports, "buildCommandMessage", {
  enumerable: true,
  get: function () {
    return _buildCommandMessage().buildCommandMessage;
  }
});
Object.defineProperty(exports, "calculateFileInfo", {
  enumerable: true,
  get: function () {
    return _fileInfo().default;
  }
});
Object.defineProperty(exports, "camelCase", {
  enumerable: true,
  get: function () {
    return _camelCase().default;
  }
});
Object.defineProperty(exports, "checksum", {
  enumerable: true,
  get: function () {
    return _checksum().checksum;
  }
});
Object.defineProperty(exports, "checksumFile", {
  enumerable: true,
  get: function () {
    return _checksum().checksumFile;
  }
});
Object.defineProperty(exports, "cleanBang", {
  enumerable: true,
  get: function () {
    return _cleanBang().default;
  }
});
Object.defineProperty(exports, "cleanChar", {
  enumerable: true,
  get: function () {
    return _cleanChar().default;
  }
});
Object.defineProperty(exports, "cleanObject", {
  enumerable: true,
  get: function () {
    return _objectClean().default;
  }
});
Object.defineProperty(exports, "createSymlinkOrCopy", {
  enumerable: true,
  get: function () {
    return _createSymlinkOrCopy().default;
  }
});
Object.defineProperty(exports, "currentDirName", {
  enumerable: true,
  get: function () {
    return _currentDirName().default;
  }
});
Object.defineProperty(exports, "deflate", {
  enumerable: true,
  get: function () {
    return _zlibDeflate().default;
  }
});
Object.defineProperty(exports, "empty", {
  enumerable: true,
  get: function () {
    return _empty().default;
  }
});
exports.eol = void 0;
Object.defineProperty(exports, "filter", {
  enumerable: true,
  get: function () {
    return _filter().default;
  }
});
Object.defineProperty(exports, "filterAsync", {
  enumerable: true,
  get: function () {
    return _filterAsync().default;
  }
});
Object.defineProperty(exports, "filterObject", {
  enumerable: true,
  get: function () {
    return _filterObject().default;
  }
});
Object.defineProperty(exports, "first", {
  enumerable: true,
  get: function () {
    return _first().default;
  }
});
Object.defineProperty(exports, "flatten", {
  enumerable: true,
  get: function () {
    return _flatten().default;
  }
});
Object.defineProperty(exports, "forEach", {
  enumerable: true,
  get: function () {
    return _foreach().default;
  }
});
Object.defineProperty(exports, "fromBase64", {
  enumerable: true,
  get: function () {
    return _fromBase().default;
  }
});
Object.defineProperty(exports, "generateRandomStr", {
  enumerable: true,
  get: function () {
    return _generateRandom().default;
  }
});
Object.defineProperty(exports, "getExt", {
  enumerable: true,
  get: function () {
    return _getExt().default;
  }
});
Object.defineProperty(exports, "getLatestVersionNumber", {
  enumerable: true,
  get: function () {
    return _resolveLatestVersion().default;
  }
});
Object.defineProperty(exports, "getMissingTestFiles", {
  enumerable: true,
  get: function () {
    return _getMissingTestFiles().default;
  }
});
Object.defineProperty(exports, "getStringifyArgs", {
  enumerable: true,
  get: function () {
    return _getStringifyArgs().default;
  }
});
Object.defineProperty(exports, "getWithoutExt", {
  enumerable: true,
  get: function () {
    return _fsNoExt().default;
  }
});
Object.defineProperty(exports, "glob", {
  enumerable: true,
  get: function () {
    return _glob().default;
  }
});
Object.defineProperty(exports, "hasOwnProperty", {
  enumerable: true,
  get: function () {
    return _hasOwnProperty().default;
  }
});
Object.defineProperty(exports, "identityFile", {
  enumerable: true,
  get: function () {
    return _identityFile().default;
  }
});
Object.defineProperty(exports, "immutableUnshift", {
  enumerable: true,
  get: function () {
    return _immutableUnshift().default;
  }
});
Object.defineProperty(exports, "inflate", {
  enumerable: true,
  get: function () {
    return _zlibInflate().default;
  }
});
Object.defineProperty(exports, "isAutoGeneratedFile", {
  enumerable: true,
  get: function () {
    return _isFileAutoGenerated().default;
  }
});
Object.defineProperty(exports, "isBitUrl", {
  enumerable: true,
  get: function () {
    return _isBitUrl().default;
  }
});
Object.defineProperty(exports, "isDir", {
  enumerable: true,
  get: function () {
    return _isDir().default;
  }
});
Object.defineProperty(exports, "isDirEmpty", {
  enumerable: true,
  get: function () {
    return _isDirEmpty().default;
  }
});
Object.defineProperty(exports, "isDirEmptySync", {
  enumerable: true,
  get: function () {
    return _isDirEmptySync().default;
  }
});
Object.defineProperty(exports, "isNumeric", {
  enumerable: true,
  get: function () {
    return _isNumeric().default;
  }
});
Object.defineProperty(exports, "isRelativeImport", {
  enumerable: true,
  get: function () {
    return _isRelativeImport().default;
  }
});
Object.defineProperty(exports, "isString", {
  enumerable: true,
  get: function () {
    return _isString().default;
  }
});
Object.defineProperty(exports, "isValidPath", {
  enumerable: true,
  get: function () {
    return _isValidPath().default;
  }
});
Object.defineProperty(exports, "mapObject", {
  enumerable: true,
  get: function () {
    return _mapObject().default;
  }
});
Object.defineProperty(exports, "mapToObject", {
  enumerable: true,
  get: function () {
    return _toObject().default;
  }
});
Object.defineProperty(exports, "objectToStringifiedTupleArray", {
  enumerable: true,
  get: function () {
    return _objectToStringifiedTupleArray().default;
  }
});
Object.defineProperty(exports, "outputFile", {
  enumerable: true,
  get: function () {
    return _fsOutputFile().default;
  }
});
Object.defineProperty(exports, "outputJsonFile", {
  enumerable: true,
  get: function () {
    return _fsOutputJsonSync().default;
  }
});
Object.defineProperty(exports, "packCommand", {
  enumerable: true,
  get: function () {
    return _packCommand().packCommand;
  }
});
Object.defineProperty(exports, "parseSSHUrl", {
  enumerable: true,
  get: function () {
    return _parseUrl().default;
  }
});
Object.defineProperty(exports, "pathHas", {
  enumerable: true,
  get: function () {
    return _propogateUntil().pathHas;
  }
});
Object.defineProperty(exports, "pathHasAll", {
  enumerable: true,
  get: function () {
    return _propogateUntil().pathHasAll;
  }
});
Object.defineProperty(exports, "pathIsInside", {
  enumerable: true,
  get: function () {
    return _pathIsInside().pathIsInside;
  }
});
Object.defineProperty(exports, "pathJoinLinux", {
  enumerable: true,
  get: function () {
    return _path().pathJoinLinux;
  }
});
Object.defineProperty(exports, "pathNormalizeToLinux", {
  enumerable: true,
  get: function () {
    return _path().pathNormalizeToLinux;
  }
});
Object.defineProperty(exports, "pathRelativeLinux", {
  enumerable: true,
  get: function () {
    return _path().pathRelativeLinux;
  }
});
Object.defineProperty(exports, "pathResolveToLinux", {
  enumerable: true,
  get: function () {
    return _path().pathResolveToLinux;
  }
});
Object.defineProperty(exports, "prependBang", {
  enumerable: true,
  get: function () {
    return _prependBang().default;
  }
});
Object.defineProperty(exports, "propogateUntil", {
  enumerable: true,
  get: function () {
    return _propogateUntil().propogateUntil;
  }
});
Object.defineProperty(exports, "readDirIgnoreDsStore", {
  enumerable: true,
  get: function () {
    return _readDirIgnoreDsStore().default;
  }
});
Object.defineProperty(exports, "readDirSyncIgnoreDsStore", {
  enumerable: true,
  get: function () {
    return _readDirIgnoreDsStore().readDirSyncIgnoreDsStore;
  }
});
Object.defineProperty(exports, "removeChalkCharacters", {
  enumerable: true,
  get: function () {
    return _removeChalkCharacters().default;
  }
});
Object.defineProperty(exports, "removeFromRequireCache", {
  enumerable: true,
  get: function () {
    return _removeFromRequireCache().default;
  }
});
Object.defineProperty(exports, "render", {
  enumerable: true,
  get: function () {
    return _prettyJsonChalk().default;
  }
});
Object.defineProperty(exports, "resolveGroupId", {
  enumerable: true,
  get: function () {
    return _osResolveGroupId().default;
  }
});
Object.defineProperty(exports, "resolveHomePath", {
  enumerable: true,
  get: function () {
    return _osResolveHomePath().default;
  }
});
Object.defineProperty(exports, "retrieveIgnoreList", {
  enumerable: true,
  get: function () {
    return _ignore().default;
  }
});
Object.defineProperty(exports, "rmDir", {
  enumerable: true,
  get: function () {
    return _fsRmdir().default;
  }
});
Object.defineProperty(exports, "searchFilesIgnoreExt", {
  enumerable: true,
  get: function () {
    return _searchFilesIgnoreExt().default;
  }
});
Object.defineProperty(exports, "sha1", {
  enumerable: true,
  get: function () {
    return _sha().default;
  }
});
Object.defineProperty(exports, "sharedStartOfArray", {
  enumerable: true,
  get: function () {
    return _sharedStart().default;
  }
});
Object.defineProperty(exports, "sortObject", {
  enumerable: true,
  get: function () {
    return _sort().default;
  }
});
Object.defineProperty(exports, "splitBy", {
  enumerable: true,
  get: function () {
    return _splitBy().default;
  }
});
Object.defineProperty(exports, "stripTrailingChar", {
  enumerable: true,
  get: function () {
    return _stripTrailingChar().default;
  }
});
Object.defineProperty(exports, "toBase64", {
  enumerable: true,
  get: function () {
    return _toBase().default;
  }
});
Object.defineProperty(exports, "toBase64ArrayBuffer", {
  enumerable: true,
  get: function () {
    return _toBase64ArrayBuffer().default;
  }
});
Object.defineProperty(exports, "toBoolean", {
  enumerable: true,
  get: function () {
    return _toBoolean().default;
  }
});
Object.defineProperty(exports, "toResultObject", {
  enumerable: true,
  get: function () {
    return _promiseToResultObject().default;
  }
});
Object.defineProperty(exports, "unpackCommand", {
  enumerable: true,
  get: function () {
    return _unpackCommand().unpackCommand;
  }
});
Object.defineProperty(exports, "writeFile", {
  enumerable: true,
  get: function () {
    return _fsWriteFile().default;
  }
});
function _filterAsync() {
  const data = _interopRequireDefault(require("./array/filter-async"));
  _filterAsync = function () {
    return data;
  };
  return data;
}
function _first() {
  const data = _interopRequireDefault(require("./array/first"));
  _first = function () {
    return data;
  };
  return data;
}
function _flatten() {
  const data = _interopRequireDefault(require("./array/flatten"));
  _flatten = function () {
    return data;
  };
  return data;
}
function _sharedStart() {
  const data = _interopRequireDefault(require("./array/shared-start"));
  _sharedStart = function () {
    return data;
  };
  return data;
}
function _splitBy() {
  const data = _interopRequireDefault(require("./array/split-by"));
  _splitBy = function () {
    return data;
  };
  return data;
}
function _toReadStream() {
  const data = _interopRequireDefault(require("./buffer/to-read-stream"));
  _toReadStream = function () {
    return data;
  };
  return data;
}
function _buildCommandMessage() {
  const data = require("./build-command-message");
  _buildCommandMessage = function () {
    return data;
  };
  return data;
}
function _sha() {
  const data = _interopRequireDefault(require("./encryption/sha1"));
  _sha = function () {
    return data;
  };
  return data;
}
function eol() {
  const data = _interopRequireWildcard(require("./eol"));
  eol = function () {
    return data;
  };
  return data;
}
Object.defineProperty(exports, "eol", {
  enumerable: true,
  get: function () {
    return eol();
  }
});
function _filterObject() {
  const data = _interopRequireDefault(require("./filter-object"));
  _filterObject = function () {
    return data;
  };
  return data;
}
function _fsOutputFile() {
  const data = _interopRequireDefault(require("./fs-output-file"));
  _fsOutputFile = function () {
    return data;
  };
  return data;
}
function _fsOutputJsonSync() {
  const data = _interopRequireDefault(require("./fs-output-json-sync"));
  _fsOutputJsonSync = function () {
    return data;
  };
  return data;
}
function _fsRmdir() {
  const data = _interopRequireDefault(require("./fs-rmdir"));
  _fsRmdir = function () {
    return data;
  };
  return data;
}
function _fsWriteFile() {
  const data = _interopRequireDefault(require("./fs-write-file"));
  _fsWriteFile = function () {
    return data;
  };
  return data;
}
function _checksum() {
  const data = require("./checksum");
  _checksum = function () {
    return data;
  };
  return data;
}
function _currentDirName() {
  const data = _interopRequireDefault(require("./fs/current-dir-name"));
  _currentDirName = function () {
    return data;
  };
  return data;
}
function _createSymlinkOrCopy() {
  const data = _interopRequireDefault(require("./fs/create-symlink-or-copy"));
  _createSymlinkOrCopy = function () {
    return data;
  };
  return data;
}
function _fileInfo() {
  const data = _interopRequireDefault(require("./fs/file-info"));
  _fileInfo = function () {
    return data;
  };
  return data;
}
function _fsNoExt() {
  const data = _interopRequireDefault(require("./fs/fs-no-ext"));
  _fsNoExt = function () {
    return data;
  };
  return data;
}
function _getExt() {
  const data = _interopRequireDefault(require("./fs/get-ext"));
  _getExt = function () {
    return data;
  };
  return data;
}
function _isDirEmpty() {
  const data = _interopRequireDefault(require("./fs/is-dir-empty"));
  _isDirEmpty = function () {
    return data;
  };
  return data;
}
function _propogateUntil() {
  const data = require("./fs/propogate-until");
  _propogateUntil = function () {
    return data;
  };
  return data;
}
function _readDirIgnoreDsStore() {
  const data = _interopRequireWildcard(require("./fs/read-dir-ignore-ds-store"));
  _readDirIgnoreDsStore = function () {
    return data;
  };
  return data;
}
function _searchFilesIgnoreExt() {
  const data = _interopRequireDefault(require("./fs/search-files-ignore-ext"));
  _searchFilesIgnoreExt = function () {
    return data;
  };
  return data;
}
function _getMissingTestFiles() {
  const data = _interopRequireDefault(require("./getMissingTestFiles"));
  _getMissingTestFiles = function () {
    return data;
  };
  return data;
}
function _glob() {
  const data = _interopRequireDefault(require("./glob"));
  _glob = function () {
    return data;
  };
  return data;
}
function _ignore() {
  const data = _interopRequireDefault(require("./ignore/ignore"));
  _ignore = function () {
    return data;
  };
  return data;
}
function _immutableUnshift() {
  const data = _interopRequireDefault(require("./immutable-unshift"));
  _immutableUnshift = function () {
    return data;
  };
  return data;
}
function _isBitUrl() {
  const data = _interopRequireDefault(require("./is-bit-url"));
  _isBitUrl = function () {
    return data;
  };
  return data;
}
function _isDir() {
  const data = _interopRequireDefault(require("./is-dir"));
  _isDir = function () {
    return data;
  };
  return data;
}
function _isDirEmptySync() {
  const data = _interopRequireDefault(require("./is-dir-empty-sync"));
  _isDirEmptySync = function () {
    return data;
  };
  return data;
}
function _isFileAutoGenerated() {
  const data = _interopRequireDefault(require("./is-file-auto-generated"));
  _isFileAutoGenerated = function () {
    return data;
  };
  return data;
}
function _isRelativeImport() {
  const data = _interopRequireDefault(require("./is-relative-import"));
  _isRelativeImport = function () {
    return data;
  };
  return data;
}
function _isValidPath() {
  const data = _interopRequireDefault(require("./is-valid-path"));
  _isValidPath = function () {
    return data;
  };
  return data;
}
function _mapObject() {
  const data = _interopRequireDefault(require("./map-object"));
  _mapObject = function () {
    return data;
  };
  return data;
}
function _toObject() {
  const data = _interopRequireDefault(require("./map/to-object"));
  _toObject = function () {
    return data;
  };
  return data;
}
function _isNumeric() {
  const data = _interopRequireDefault(require("./number/is-numeric"));
  _isNumeric = function () {
    return data;
  };
  return data;
}
function _objectClean() {
  const data = _interopRequireDefault(require("./object-clean"));
  _objectClean = function () {
    return data;
  };
  return data;
}
function _objectToStringifiedTupleArray() {
  const data = _interopRequireDefault(require("./object-to-stringified-tuple-array"));
  _objectToStringifiedTupleArray = function () {
    return data;
  };
  return data;
}
function _empty() {
  const data = _interopRequireDefault(require("./object/empty"));
  _empty = function () {
    return data;
  };
  return data;
}
function _filter() {
  const data = _interopRequireDefault(require("./object/filter"));
  _filter = function () {
    return data;
  };
  return data;
}
function _foreach() {
  const data = _interopRequireDefault(require("./object/foreach"));
  _foreach = function () {
    return data;
  };
  return data;
}
function _hasOwnProperty() {
  const data = _interopRequireDefault(require("./object/has-own-property"));
  _hasOwnProperty = function () {
    return data;
  };
  return data;
}
function _sort() {
  const data = _interopRequireDefault(require("./object/sort"));
  _sort = function () {
    return data;
  };
  return data;
}
function _osResolveGroupId() {
  const data = _interopRequireDefault(require("./os-resolve-group-id"));
  _osResolveGroupId = function () {
    return data;
  };
  return data;
}
function _osResolveHomePath() {
  const data = _interopRequireDefault(require("./os-resolve-home-path"));
  _osResolveHomePath = function () {
    return data;
  };
  return data;
}
function _packCommand() {
  const data = require("./pack-command");
  _packCommand = function () {
    return data;
  };
  return data;
}
function _path() {
  const data = require("./path");
  _path = function () {
    return data;
  };
  return data;
}
function _pathIsInside() {
  const data = require("./path-is-inside");
  _pathIsInside = function () {
    return data;
  };
  return data;
}
function _prependBang() {
  const data = _interopRequireDefault(require("./prepend-bang"));
  _prependBang = function () {
    return data;
  };
  return data;
}
function _prettyJsonChalk() {
  const data = _interopRequireDefault(require("./pretty-json-chalk"));
  _prettyJsonChalk = function () {
    return data;
  };
  return data;
}
function _promiseToResultObject() {
  const data = _interopRequireDefault(require("./promise-to-result-object"));
  _promiseToResultObject = function () {
    return data;
  };
  return data;
}
function _queue() {
  const data = _interopRequireDefault(require("./queue"));
  _queue = function () {
    return data;
  };
  return data;
}
function _removeFromRequireCache() {
  const data = _interopRequireDefault(require("./remove-from-require-cache"));
  _removeFromRequireCache = function () {
    return data;
  };
  return data;
}
function _resolveLatestVersion() {
  const data = _interopRequireDefault(require("./resolveLatestVersion"));
  _resolveLatestVersion = function () {
    return data;
  };
  return data;
}
function _identityFile() {
  const data = _interopRequireDefault(require("./ssh/identity-file"));
  _identityFile = function () {
    return data;
  };
  return data;
}
function _parseUrl() {
  const data = _interopRequireDefault(require("./ssh/parse-url"));
  _parseUrl = function () {
    return data;
  };
  return data;
}
function _camelCase() {
  const data = _interopRequireDefault(require("./string/camel-case"));
  _camelCase = function () {
    return data;
  };
  return data;
}
function _cleanBang() {
  const data = _interopRequireDefault(require("./string/clean-bang"));
  _cleanBang = function () {
    return data;
  };
  return data;
}
function _cleanChar() {
  const data = _interopRequireDefault(require("./string/clean-char"));
  _cleanChar = function () {
    return data;
  };
  return data;
}
function _fromBase() {
  const data = _interopRequireDefault(require("./string/from-base64"));
  _fromBase = function () {
    return data;
  };
  return data;
}
function _generateRandom() {
  const data = _interopRequireDefault(require("./string/generate-random"));
  _generateRandom = function () {
    return data;
  };
  return data;
}
function _getStringifyArgs() {
  const data = _interopRequireDefault(require("./string/get-stringify-args"));
  _getStringifyArgs = function () {
    return data;
  };
  return data;
}
function _isString() {
  const data = _interopRequireDefault(require("./string/is-string"));
  _isString = function () {
    return data;
  };
  return data;
}
function _removeChalkCharacters() {
  const data = _interopRequireDefault(require("./string/remove-chalk-characters"));
  _removeChalkCharacters = function () {
    return data;
  };
  return data;
}
function _stripTrailingChar() {
  const data = _interopRequireDefault(require("./string/strip-trailing-char"));
  _stripTrailingChar = function () {
    return data;
  };
  return data;
}
function _toBase() {
  const data = _interopRequireDefault(require("./string/to-base64"));
  _toBase = function () {
    return data;
  };
  return data;
}
function _toBase64ArrayBuffer() {
  const data = _interopRequireDefault(require("./string/to-base64-array-buffer"));
  _toBase64ArrayBuffer = function () {
    return data;
  };
  return data;
}
function _toBoolean() {
  const data = _interopRequireDefault(require("./to-boolean"));
  _toBoolean = function () {
    return data;
  };
  return data;
}
function _unpackCommand() {
  const data = require("./unpack-command");
  _unpackCommand = function () {
    return data;
  };
  return data;
}
function _zlibDeflate() {
  const data = _interopRequireDefault(require("./zlib-deflate"));
  _zlibDeflate = function () {
    return data;
  };
  return data;
}
function _zlibInflate() {
  const data = _interopRequireDefault(require("./zlib-inflate"));
  _zlibInflate = function () {
    return data;
  };
  return data;
}
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
