// @teambit/legacy/dist/utils/queue.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _pQueue() {
  const data = _interopRequireDefault(require("p-queue"));
  _pQueue = function () {
    return data;
  };
  return data;
}
class Queue {
  constructor(concurrency = 5, autoStart = true) {
    this.concurrency = concurrency;
    this.autoStart = autoStart;
    (0, _defineProperty2().default)(this, "queue", void 0);
    this.queue = new (_pQueue().default)({
      concurrency,
      autoStart
    });
  }
  addAll(fns) {
    return this.queue.addAll(fns);
  }
  add(fn, priority) {
    return this.queue.add(fn, {
      priority
    });
  }
  pause() {
    this.queue.pause();
  }
  size() {
    return this.queue.size;
  }
  onEmpty() {
    return this.queue.onEmpty();
  }
  onIdle() {
    return this.queue.onIdle();
  }
}
exports.default = Queue;
