// @teambit/aspect-loader/dist/plugin.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Plugin = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
class Plugin {
  constructor(def, path) {
    this.def = def;
    this.path = path;
    (0, _defineProperty2().default)(this, "_instance", void 0);
  }

  // consider adding a more abstract type here to allow users to ask for dependencies.

  /**
   * determines whether the plugin supports a certain runtime.
   */
  supportsRuntime(runtime) {
    return this.def.runtimes.includes(runtime);
  }
  register(sourceAspect) {
    const object = this.require();
    this.def.register(object, sourceAspect);
  }
  require() {
    // eslint-disable-next-line global-require, import/no-dynamic-require
    this._instance = require(this.path).default;
    return this._instance;
  }
}
exports.Plugin = Plugin;
