// @teambit/harmony-ui-app/dist/harmony-ui-app.main.runtime.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.array.iterator.js");
require("core-js/modules/es.promise.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HarmonyUiAppMain = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _application() {
  const data = require("@teambit/application");
  _application = function () {
    return data;
  };
  return data;
}
function _cli() {
  const data = require("@teambit/cli");
  _cli = function () {
    return data;
  };
  return data;
}
function _ui() {
  const data = _interopRequireDefault(require("@teambit/ui"));
  _ui = function () {
    return data;
  };
  return data;
}
function _component() {
  const data = require("@teambit/component");
  _component = function () {
    return data;
  };
  return data;
}
function _harmonyUiApp() {
  const data = require("./harmony-ui-app.aspect");
  _harmonyUiApp = function () {
    return data;
  };
  return data;
}
// import { HarmonyUIApp } from './ui.application';

class HarmonyUiAppMain {
  constructor(application, ui, componentAspect) {
    this.application = application;
    this.ui = ui;
    this.componentAspect = componentAspect;
  }

  /**
   * register a new harmony UI application.
   */
  registerHarmonyApp(options) {
    this.ui.registerUiRoot({
      name: options.name,
      path: this.componentAspect.getHost().path,
      configFile: '',
      async resolvePattern() {
        return [];
      },
      async resolveAspects() {
        return options.aspectDefs;
      }
    });

    // this.application.registerApp(new HarmonyUIApp(options.name, this.ui));

    return this;
  }
  static async provider([application, ui, componentAspect]) {
    return new HarmonyUiAppMain(application, ui, componentAspect);
  }
}
exports.HarmonyUiAppMain = HarmonyUiAppMain;
(0, _defineProperty2().default)(HarmonyUiAppMain, "slots", []);
(0, _defineProperty2().default)(HarmonyUiAppMain, "dependencies", [_application().ApplicationAspect, _ui().default, _component().ComponentAspect]);
(0, _defineProperty2().default)(HarmonyUiAppMain, "runtime", _cli().MainRuntime);
_harmonyUiApp().HarmonyUiAppAspect.addRuntime(HarmonyUiAppMain);
