// @teambit/react/dist/templates/react-workspace-deprecated/index.js
"use strict";

require("core-js/modules/es.promise.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.deprecatedReactWorkspaceTemplate = void 0;
function _workspaceCommon() {
  const data = require("../workspace-common");
  _workspaceCommon = function () {
    return data;
  };
  return data;
}
const deprecatedReactWorkspaceTemplate = {
  name: 'react-workspace',
  description: 'React workspace with demo components',
  hidden: true,
  generateFiles: async context => {
    return (0, _workspaceCommon().generateFiles)(context);
  },
  fork: () => {
    return [{
      id: 'teambit.react/templates/envs/my-react',
      targetName: 'envs/my-react',
      path: 'demo/envs/my-react'
    }, {
      id: 'teambit.react/templates/ui/text',
      targetName: 'ui/text',
      path: 'demo/ui/text'
    }, {
      id: 'teambit.react/templates/ui/heading',
      targetName: 'ui/heading',
      path: 'demo/ui/heading'
    }, {
      id: 'teambit.react/templates/ui/card',
      targetName: 'ui/card',
      path: 'demo/ui/card'
    }, {
      id: 'teambit.react/templates/pages/welcome',
      targetName: 'pages/welcome',
      path: 'demo/pages/welcome'
    }];
  }
};
exports.deprecatedReactWorkspaceTemplate = deprecatedReactWorkspaceTemplate;
