// @teambit/community/dist/community.main.runtime.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.promise.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CommunityMain = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _cli() {
  const data = require("@teambit/cli");
  _cli = function () {
    return data;
  };
  return data;
}
function _constants() {
  const data = require("@teambit/legacy/dist/constants");
  _constants = function () {
    return data;
  };
  return data;
}
function _community() {
  const data = require("./community.aspect");
  _community = function () {
    return data;
  };
  return data;
}
class CommunityMain {
  constructor(config) {
    this.config = config;
  }
  getBaseDomain() {
    return this.config.communityDomain;
  }
  getDocsDomain() {
    return this.config.docsDomain;
  }
  static async provider(_deps, config) {
    return new CommunityMain(config);
  }
}
exports.CommunityMain = CommunityMain;
(0, _defineProperty2().default)(CommunityMain, "slots", []);
(0, _defineProperty2().default)(CommunityMain, "dependencies", []);
(0, _defineProperty2().default)(CommunityMain, "runtime", _cli().MainRuntime);
(0, _defineProperty2().default)(CommunityMain, "defaultConfig", {
  communityDomain: _constants().BASE_COMMUNITY_DOMAIN,
  docsDomain: _constants().BASE_DOCS_DOMAIN
});
_community().CommunityAspect.addRuntime(CommunityMain);
