// @teambit/pkg/dist/pack.task.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.promise.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PackTask = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
/**
 * pack components to a .tgz file
 */
class PackTask {
  constructor(aspectId, packer, logger) {
    this.aspectId = aspectId;
    this.packer = packer;
    this.logger = logger;
    (0, _defineProperty2().default)(this, "name", 'PackComponents');
    (0, _defineProperty2().default)(this, "location", 'end');
  }
  async execute(context) {
    const capsules = context.capsuleNetwork.seedersCapsules;
    this.logger.info(`going to run pack on ${capsules.length} capsules`);
    const packResults = await this.packer.packMultipleCapsules(capsules, {
      override: true
    }, false, true);
    const packArtifactsDefs = this.packer.getArtifactDefInCapsule();
    return {
      componentsResults: packResults,
      artifacts: [packArtifactsDefs]
    };
  }
}
exports.PackTask = PackTask;
