// @teambit/pkg/dist/pkg.graphql.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.pkgSchema = pkgSchema;
function _graphqlTag() {
  const data = _interopRequireDefault(require("graphql-tag"));
  _graphqlTag = function () {
    return data;
  };
  return data;
}
function pkgSchema(pkg) {
  return {
    typeDefs: (0, _graphqlTag().default)`
      extend type Component {
        packageManifest: PackageManifest

        # package name of the component.
        packageName: String!
      }

      type TarDist {
        tarball: String
        shasum: String
      }

      type VersionsPackageManifest {
        name: String
        version: String
        main: String
        dependencies: JSONObject
        devDependencies: JSONObject
        peerDependencies: JSONObject
        scripts: JSONObject
        dist: TarDist
      }

      type PackageManifest {
        name: String
        distTags: JSONObject
        externalRegistry: Boolean
        versions(version: String): [VersionsPackageManifest]
      }
    `,
    resolvers: {
      Component: {
        packageName: component => {
          return pkg.getPackageName(component);
        },
        packageManifest: component => {
          return pkg.getManifest(component);
        }
      },
      PackageManifest: {
        versions: (parent, {
          version
        }) => {
          if (version) {
            return parent.versions.filter(v => v.version === version);
          }
          return parent.versions;
        }
      }
    }
  };
}
