// @teambit/pkg/dist/pkg.ui.runtime.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.array.iterator.js");
require("core-js/modules/es.promise.js");
require("core-js/modules/es.regexp.exec.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.PkgUI = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _react() {
  const data = _interopRequireDefault(require("react"));
  _react = function () {
    return data;
  };
  return data;
}
function _ui() {
  const data = require("@teambit/ui");
  _ui = function () {
    return data;
  };
  return data;
}
function _uiFoundationUiUseBox() {
  const data = require("@teambit/ui-foundation.ui.use-box.menu");
  _uiFoundationUiUseBox = function () {
    return data;
  };
  return data;
}
function _component() {
  const data = _interopRequireDefault(require("@teambit/component"));
  _component = function () {
    return data;
  };
  return data;
}
function _pkg() {
  const data = require("./pkg.aspect");
  _pkg = function () {
    return data;
  };
  return data;
}
class PkgUI {
  constructor() {
    (0, _defineProperty2().default)(this, "npmConsumeMethod", (comp, options) => {
      if (options !== null && options !== void 0 && options.currentLane) return undefined;
      const registry = comp.packageName.split('/')[0];
      const packageVersion = comp.version === comp.latest ? '' : `@${comp.version}`;
      return {
        Title: /*#__PURE__*/_react().default.createElement("img", {
          style: {
            width: '30px'
          },
          src: "https://static.bit.dev/brands/logo-npm-new.svg"
        }),
        Component: /*#__PURE__*/_react().default.createElement(_uiFoundationUiUseBox().Install, {
          config: `npm config set '${registry}:registry' https://node.bit.cloud`,
          componentName: comp.id.name,
          packageManager: "npm",
          copyString: `npm i ${comp.packageName}${packageVersion}`,
          registryName: registry
        }),
        order: 10
      };
    });
  }
  static async provider([componentUI]) {
    const pkg = new PkgUI();
    componentUI.registerConsumeMethod(pkg.npmConsumeMethod);
    return pkg;
  }
}
exports.PkgUI = PkgUI;
(0, _defineProperty2().default)(PkgUI, "runtime", _ui().UIRuntime);
(0, _defineProperty2().default)(PkgUI, "dependencies", [_component().default]);
var _default = PkgUI;
exports.default = _default;
_pkg().PkgAspect.addRuntime(PkgUI);
