// @teambit/pkg/dist/publish-dry-run.task.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.promise.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PublishDryRunTask = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
/**
 * publish build task is running "publish --dry-run" to avoid later npm errors during export
 */
class PublishDryRunTask {
  constructor(aspectId, publisher, packer, logger) {
    this.aspectId = aspectId;
    this.publisher = publisher;
    this.packer = packer;
    this.logger = logger;
    (0, _defineProperty2().default)(this, "name", 'PublishDryRun');
    (0, _defineProperty2().default)(this, "location", 'end');
    (0, _defineProperty2().default)(this, "dependencies", void 0);
  }
  async execute(context) {
    this.publisher.options.dryRun = true;
    const capsules = context.capsuleNetwork.seedersCapsules;
    // const capsulesToPublish = capsules.filter((c) => this.publisher.shouldPublish(c.component.config.extensions));
    const capsulesToPublish = [];
    capsules.forEach(c => {
      const shouldPublish = this.publisher.shouldPublish(c.component.config.extensions);
      if (shouldPublish) {
        capsulesToPublish.push(c);
      }
    });
    this.logger.info(`going to run publish dry-run on ${capsulesToPublish.length} out of ${capsules.length}`);
    const publishResults = await this.publisher.publishMultipleCapsules(capsulesToPublish);
    this.logger.info(`going to run pack dry-run on ${capsules.length} capsules`);
    const packResults = await this.packer.packMultipleCapsules(capsules, {
      override: true
    }, true, true);
    return {
      componentsResults: publishResults.concat(packResults),
      artifacts: []
    };
  }
}
exports.PublishDryRunTask = PublishDryRunTask;
