// @teambit/pkg/dist/publish.task.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.promise.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PublishTask = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
/**
 * publish components by running "npm publish"
 */
class PublishTask {
  constructor(aspectId, publisher, logger) {
    this.aspectId = aspectId;
    this.publisher = publisher;
    this.logger = logger;
    (0, _defineProperty2().default)(this, "name", 'PublishComponents');
    (0, _defineProperty2().default)(this, "location", 'end');
  }
  async execute(context) {
    this.publisher.options.dryRun = false;
    const capsules = context.capsuleNetwork.seedersCapsules;
    // const capsulesToPublish = capsules.filter((c) => this.publisher.shouldPublish(c.component.config.extensions));
    const capsulesToPublish = [];
    capsules.forEach(c => {
      const shouldPublish = this.publisher.shouldPublish(c.component.config.extensions);
      if (shouldPublish) {
        capsulesToPublish.push(c);
      }
    });
    this.logger.info(`going to run publish on ${capsulesToPublish.length} out of ${capsules.length}`);
    const publishResults = await this.publisher.publishMultipleCapsules(capsulesToPublish);
    return {
      componentsResults: publishResults,
      artifacts: []
    };
  }
}
exports.PublishTask = PublishTask;
