// @teambit/dependency-resolver/dist/exceptions/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "CoreAspectLinkError", {
  enumerable: true,
  get: function () {
    return _coreAspectLinkError().CoreAspectLinkError;
  }
});
Object.defineProperty(exports, "InvalidVersionWithPrefix", {
  enumerable: true,
  get: function () {
    return _invalidVersionWithPrefix().InvalidVersionWithPrefix;
  }
});
Object.defineProperty(exports, "MainAspectNotInstallable", {
  enumerable: true,
  get: function () {
    return _mainAspectNotInstallable().MainAspectNotInstallable;
  }
});
Object.defineProperty(exports, "MainAspectNotLinkable", {
  enumerable: true,
  get: function () {
    return _mainAspectNotLinkable().MainAspectNotLinkable;
  }
});
Object.defineProperty(exports, "NonAspectCorePackageLinkError", {
  enumerable: true,
  get: function () {
    return _nonAspectCorePackageLinkError().NonAspectCorePackageLinkError;
  }
});
Object.defineProperty(exports, "PackageManagerNotFound", {
  enumerable: true,
  get: function () {
    return _packageManagerNotFound().PackageManagerNotFound;
  }
});
Object.defineProperty(exports, "RootDirNotDefined", {
  enumerable: true,
  get: function () {
    return _rootDirNotDefined().RootDirNotDefined;
  }
});
function _packageManagerNotFound() {
  const data = require("./package-manager-not-found");
  _packageManagerNotFound = function () {
    return data;
  };
  return data;
}
function _rootDirNotDefined() {
  const data = require("./root-dir-not-defined");
  _rootDirNotDefined = function () {
    return data;
  };
  return data;
}
function _mainAspectNotInstallable() {
  const data = require("./main-aspect-not-installable");
  _mainAspectNotInstallable = function () {
    return data;
  };
  return data;
}
function _mainAspectNotLinkable() {
  const data = require("./main-aspect-not-linkable");
  _mainAspectNotLinkable = function () {
    return data;
  };
  return data;
}
function _coreAspectLinkError() {
  const data = require("./core-aspect-link-error");
  _coreAspectLinkError = function () {
    return data;
  };
  return data;
}
function _nonAspectCorePackageLinkError() {
  const data = require("./non-aspect-core-package-link-error");
  _nonAspectCorePackageLinkError = function () {
    return data;
  };
  return data;
}
function _invalidVersionWithPrefix() {
  const data = require("./invalid-version-with-prefix");
  _invalidVersionWithPrefix = function () {
    return data;
  };
  return data;
}
