// @teambit/envs/dist/env.fragment.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.promise.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EnvFragment = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
class EnvFragment {
  constructor(envs) {
    this.envs = envs;
    (0, _defineProperty2().default)(this, "title", 'env');
    (0, _defineProperty2().default)(this, "weight", 3);
  }
  async renderRow(component) {
    return {
      title: this.title,
      content: this.getEnvId(component)
    };
  }
  async json(component) {
    return {
      title: this.title,
      json: this.getEnvId(component)
    };
  }
  getEnvId(component) {
    // don't use this.envs.getEnv(). otherwise, it'll throw an error when running bit-show on a remote component
    // where the env can't register to the slot.
    return this.envs.getEnvId(component);
  }
}
exports.EnvFragment = EnvFragment;
