// @teambit/envs/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  EnvsAspect: true,
  DEFAULT_ENV: true,
  ExecutionContext: true,
  EnvService: true,
  ConcreteService: true,
  EnvRuntime: true,
  EnvsExecutionResult: true,
  EnvDefinition: true
};
Object.defineProperty(exports, "ConcreteService", {
  enumerable: true,
  get: function () {
    return _services().ConcreteService;
  }
});
Object.defineProperty(exports, "DEFAULT_ENV", {
  enumerable: true,
  get: function () {
    return _environmentsMain().DEFAULT_ENV;
  }
});
Object.defineProperty(exports, "EnvDefinition", {
  enumerable: true,
  get: function () {
    return _envDefinition().EnvDefinition;
  }
});
Object.defineProperty(exports, "EnvRuntime", {
  enumerable: true,
  get: function () {
    return _envRuntime().EnvRuntime;
  }
});
Object.defineProperty(exports, "EnvService", {
  enumerable: true,
  get: function () {
    return _services().EnvService;
  }
});
Object.defineProperty(exports, "EnvsAspect", {
  enumerable: true,
  get: function () {
    return _environments().EnvsAspect;
  }
});
Object.defineProperty(exports, "EnvsExecutionResult", {
  enumerable: true,
  get: function () {
    return _envsExecutionResult().EnvsExecutionResult;
  }
});
Object.defineProperty(exports, "ExecutionContext", {
  enumerable: true,
  get: function () {
    return _context().ExecutionContext;
  }
});
exports.default = void 0;
function _environments() {
  const data = require("./environments.aspect");
  _environments = function () {
    return data;
  };
  return data;
}
function _environmentsMain() {
  const data = require("./environments.main.runtime");
  _environmentsMain = function () {
    return data;
  };
  return data;
}
var _environment = require("./environment");
Object.keys(_environment).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _environment[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _environment[key];
    }
  });
});
function _context() {
  const data = require("./context");
  _context = function () {
    return data;
  };
  return data;
}
function _services() {
  const data = require("./services");
  _services = function () {
    return data;
  };
  return data;
}
function _envRuntime() {
  const data = require("./runtime/env-runtime");
  _envRuntime = function () {
    return data;
  };
  return data;
}
function _envsExecutionResult() {
  const data = require("./runtime/envs-execution-result");
  _envsExecutionResult = function () {
    return data;
  };
  return data;
}
function _envDefinition() {
  const data = require("./env-definition");
  _envDefinition = function () {
    return data;
  };
  return data;
}
var _default = _environments().EnvsAspect;
exports.default = _default;
