// @teambit/command-bar/dist/searchers/command-searcher/command-searcher.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CommandSearcher = void 0;
function _react() {
  const data = _interopRequireDefault(require("react"));
  _react = function () {
    return data;
  };
  return data;
}
function _explorerUi() {
  const data = require("@teambit/explorer.ui.command-bar");
  _explorerUi = function () {
    return data;
  };
  return data;
}
function _commandResult() {
  const data = require("./command-result");
  _commandResult = function () {
    return data;
  };
  return data;
}
const searchKeys = ['displayName'];
class CommandSearcher extends _explorerUi().FuzzySearcher {
  constructor(commands) {
    super({
      searchKeys
    });
    this.update(commands);
  }
  test(term) {
    return term.startsWith('>');
  }
  toSearchableItem(item) {
    return item;
  }
  toSearchResult({
    item
  }) {
    return {
      id: item.id,
      action: item.action,
      children: /*#__PURE__*/_react().default.createElement(_commandResult().CommandResult, {
        command: item
      })
    };
  }
}
exports.CommandSearcher = CommandSearcher;
