// @teambit/code/dist/code.section.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CodeSection = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _react() {
  const data = _interopRequireDefault(require("react"));
  _react = function () {
    return data;
  };
  return data;
}
function _uiFoundationUi() {
  const data = require("@teambit/ui-foundation.ui.menu-widget-icon");
  _uiFoundationUi = function () {
    return data;
  };
  return data;
}
class CodeSection {
  constructor(codeUI) {
    this.codeUI = codeUI;
    (0, _defineProperty2().default)(this, "route", {
      path: '~code/*',
      element: this.codeUI.getCodePage()
    });
    (0, _defineProperty2().default)(this, "navigationLink", {
      href: '~code',
      children: /*#__PURE__*/_react().default.createElement(_uiFoundationUi().MenuWidgetIcon, {
        icon: "Code",
        tooltipContent: "Code"
      }),
      displayName: 'Code'
    });
    (0, _defineProperty2().default)(this, "order", 30);
  }
}
exports.CodeSection = CodeSection;
