// @teambit/code/dist/code.compare.section.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CodeCompareSection = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _react() {
  const data = _interopRequireDefault(require("react"));
  _react = function () {
    return data;
  };
  return data;
}
function _codeUi() {
  const data = require("@teambit/code.ui.code-compare");
  _codeUi = function () {
    return data;
  };
  return data;
}
class CodeCompareSection {
  constructor(fileIconSlot) {
    this.fileIconSlot = fileIconSlot;
    (0, _defineProperty2().default)(this, "navigationLink", {
      href: 'code',
      children: 'Code',
      order: 4
    });
    (0, _defineProperty2().default)(this, "route", {
      path: 'code/*',
      element: /*#__PURE__*/_react().default.createElement(_codeUi().CodeCompare, {
        fileIconSlot: this.fileIconSlot
      })
    });
  }
}
exports.CodeCompareSection = CodeCompareSection;
