// @teambit/ui/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  UIAspect: true,
  UIRuntime: true,
  UIRoot: true,
  PostStartOptions: true,
  ProxyEntry: true,
  DataQueryResult: true,
  useDataQuery: true,
  useMutation: true
};
Object.defineProperty(exports, "DataQueryResult", {
  enumerable: true,
  get: function () {
    return _uiFoundationUiHooks().DataQueryResult;
  }
});
Object.defineProperty(exports, "PostStartOptions", {
  enumerable: true,
  get: function () {
    return _uiRoot().PostStartOptions;
  }
});
Object.defineProperty(exports, "ProxyEntry", {
  enumerable: true,
  get: function () {
    return _uiRoot().ProxyEntry;
  }
});
Object.defineProperty(exports, "UIAspect", {
  enumerable: true,
  get: function () {
    return _ui().UIAspect;
  }
});
Object.defineProperty(exports, "UIRoot", {
  enumerable: true,
  get: function () {
    return _uiRoot().UIRoot;
  }
});
Object.defineProperty(exports, "UIRuntime", {
  enumerable: true,
  get: function () {
    return _ui().UIRuntime;
  }
});
Object.defineProperty(exports, "default", {
  enumerable: true,
  get: function () {
    return _ui().UIAspect;
  }
});
Object.defineProperty(exports, "useDataQuery", {
  enumerable: true,
  get: function () {
    return _uiFoundationUiHooks().useDataQuery;
  }
});
Object.defineProperty(exports, "useMutation", {
  enumerable: true,
  get: function () {
    return _client().useMutation;
  }
});
function _ui() {
  const data = require("./ui.aspect");
  _ui = function () {
    return data;
  };
  return data;
}
var _events = require("./events");
Object.keys(_events).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _events[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _events[key];
    }
  });
});
function _uiRoot() {
  const data = require("./ui-root");
  _uiRoot = function () {
    return data;
  };
  return data;
}
function _uiFoundationUiHooks() {
  const data = require("@teambit/ui-foundation.ui.hooks.use-data-query");
  _uiFoundationUiHooks = function () {
    return data;
  };
  return data;
}
function _client() {
  const data = require("@apollo/client");
  _client = function () {
    return data;
  };
  return data;
}
