// @teambit/legacy/dist/e2e-helper/e2e-scope-json-helper.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _fsExtra() {
  const data = _interopRequireDefault(require("fs-extra"));
  _fsExtra = function () {
    return data;
  };
  return data;
}
function path() {
  const data = _interopRequireWildcard(require("path"));
  path = function () {
    return data;
  };
  return data;
}
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
class ScopeJsonHelper {
  constructor(scopes) {
    (0, _defineProperty2().default)(this, "scopes", void 0);
    this.scopes = scopes;
  }
  read(scopeJsonDir = composeScopePathForWorkspace(this.scopes.localPath)) {
    const scopeJsonPath = path().join(scopeJsonDir, 'scope.json');
    return _fsExtra().default.existsSync(scopeJsonPath) ? _fsExtra().default.readJSONSync(scopeJsonPath) : {};
  }
  write(scopeJson, scopeJsonDir = composeScopePathForWorkspace(this.scopes.localPath)) {
    const scopeJsonPath = path().join(scopeJsonDir, 'scope.json');
    return _fsExtra().default.writeJSONSync(scopeJsonPath, scopeJson, {
      spaces: 2
    });
  }
  addKeyVal(key, val, scopeJsonDir = composeScopePathForWorkspace(this.scopes.localPath)) {
    const scopeJson = this.read(scopeJsonDir);
    scopeJson[key] = val;
    this.write(scopeJson, scopeJsonDir);
  }
}
exports.default = ScopeJsonHelper;
function composeScopePathForWorkspace(workspacePath) {
  return path().join(workspacePath, '.bit');
}
