// @teambit/workspace/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  WorkspaceAspect: true,
  CheckTypes: true,
  ComponentStatus: true,
  WorkspaceModelComponent: true,
  WorkspaceModel: true,
  WorkspaceContext: true
};
Object.defineProperty(exports, "CheckTypes", {
  enumerable: true,
  get: function () {
    return _checkTypes().CheckTypes;
  }
});
Object.defineProperty(exports, "ComponentStatus", {
  enumerable: true,
  get: function () {
    return _workspaceComponent().ComponentStatus;
  }
});
Object.defineProperty(exports, "WorkspaceAspect", {
  enumerable: true,
  get: function () {
    return _workspace().WorkspaceAspect;
  }
});
Object.defineProperty(exports, "WorkspaceContext", {
  enumerable: true,
  get: function () {
    return _workspaceContext().WorkspaceContext;
  }
});
Object.defineProperty(exports, "WorkspaceModel", {
  enumerable: true,
  get: function () {
    return _workspaceModel().Workspace;
  }
});
Object.defineProperty(exports, "WorkspaceModelComponent", {
  enumerable: true,
  get: function () {
    return _workspaceModel().WorkspaceModelComponent;
  }
});
exports.default = void 0;
function _workspace() {
  const data = require("./workspace.aspect");
  _workspace = function () {
    return data;
  };
  return data;
}
function _checkTypes() {
  const data = require("./watch/check-types");
  _checkTypes = function () {
    return data;
  };
  return data;
}
var _events = require("./events");
Object.keys(_events).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _events[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _events[key];
    }
  });
});
function _workspaceComponent() {
  const data = require("./workspace-component");
  _workspaceComponent = function () {
    return data;
  };
  return data;
}
function _workspaceModel() {
  const data = require("./ui/workspace/workspace-model");
  _workspaceModel = function () {
    return data;
  };
  return data;
}
function _workspaceContext() {
  const data = require("./ui/workspace/workspace-context");
  _workspaceContext = function () {
    return data;
  };
  return data;
}
var _default = _workspace().WorkspaceAspect;
exports.default = _default;
