// @teambit/react/dist/templates/react-workspace-app/index.js
"use strict";

require("core-js/modules/es.promise.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.reactWorkspaceAppTemplate = void 0;
function _commentJson() {
  const data = require("comment-json");
  _commentJson = function () {
    return data;
  };
  return data;
}
function _workspaceCommon() {
  const data = require("../workspace-common");
  _workspaceCommon = function () {
    return data;
  };
  return data;
}
function _constants() {
  const data = require("../workspace-common/constants");
  _constants = function () {
    return data;
  };
  return data;
}
const reactWorkspaceAppTemplate = {
  name: 'react-app',
  description: 'EXPERIMENTAL. react workspace for an app',
  hidden: true,
  appName: 'my-app',
  generateFiles: async context => {
    const scope = context.defaultScope || _constants().DEFAULT_SCOPE_NAME;
    const extensions = {
      [`${scope}/apps/my-app`]: (0, _commentJson().parse)(`{}`)
    };
    return (0, _workspaceCommon().generateFiles)(context, extensions);
  },
  fork: () => {
    return [{
      id: 'teambit.react/templates/apps/my-app',
      targetName: 'apps/my-app'
    },
    // { id: 'teambit.react/templates/envs/my-react', targetName: 'envs/my-react' }, // TODO: uncomment when ready
    {
      id: 'teambit.react/templates/themes/theme',
      targetName: 'themes/theme'
    }, {
      id: 'teambit.react/templates/pages/home',
      targetName: 'pages/home'
    }];
  }
};
exports.reactWorkspaceAppTemplate = reactWorkspaceAppTemplate;
