// @teambit/ui/dist/webpack/webpack.ssr.config.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createWebpackConfig;
function _path() {
  const data = _interopRequireDefault(require("path"));
  _path = function () {
    return data;
  };
  return data;
}
function _webpackMerge() {
  const data = require("webpack-merge");
  _webpackMerge = function () {
    return data;
  };
  return data;
}
function _webpackBase() {
  const data = _interopRequireDefault(require("./webpack.base.config"));
  _webpackBase = function () {
    return data;
  };
  return data;
}
function createWebpackConfig(workspaceDir, entryFiles, publicDir) {
  const baseConfig = (0, _webpackBase().default)(workspaceDir, entryFiles);
  const ssrConfig = createSsrConfig(workspaceDir, publicDir);
  const combined = (0, _webpackMerge().merge)(baseConfig, ssrConfig);
  return combined;
}
function createSsrConfig(workspaceDir, publicDir) {
  const ssrConfig = {
    target: 'node',
    devtool: 'eval-cheap-source-map',
    // TODO

    output: {
      path: _path().default.resolve(workspaceDir, publicDir, 'ssr'),
      publicPath: '/public/ssr/',
      libraryTarget: 'commonjs',
      filename: 'index.js'
    }

    // // no optimizations for ssr at this point,
    // // especially no chunks.
    // optimization: { },
  };

  return ssrConfig;
}
