// @teambit/webpack/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  runTransformersWithContext: true,
  WebpackAspect: true,
  WebpackDevServer: true,
  WebpackBundler: true,
  Configuration: true,
  WebpackConfigMutator: true,
  WebpackBitReporterPlugin: true,
  fallbacks: true,
  fallbacksAliases: true,
  fallbacksProvidePluginConfig: true,
  GenerateBodyInjectionTransformer: true,
  BodyInjectionOptions: true,
  generateAddAliasesFromPeersTransformer: true,
  generateExposePeersTransformer: true,
  generateExternalsTransformer: true,
  GenerateHeadInjectionTransformer: true,
  HeadInjectionOptions: true
};
Object.defineProperty(exports, "BodyInjectionOptions", {
  enumerable: true,
  get: function () {
    return _transformers().BodyInjectionOptions;
  }
});
Object.defineProperty(exports, "Configuration", {
  enumerable: true,
  get: function () {
    return _webpack4().Configuration;
  }
});
Object.defineProperty(exports, "GenerateBodyInjectionTransformer", {
  enumerable: true,
  get: function () {
    return _transformers().GenerateBodyInjectionTransformer;
  }
});
Object.defineProperty(exports, "GenerateHeadInjectionTransformer", {
  enumerable: true,
  get: function () {
    return _transformers().GenerateHeadInjectionTransformer;
  }
});
Object.defineProperty(exports, "HeadInjectionOptions", {
  enumerable: true,
  get: function () {
    return _transformers().HeadInjectionOptions;
  }
});
Object.defineProperty(exports, "WebpackAspect", {
  enumerable: true,
  get: function () {
    return _webpack().WebpackAspect;
  }
});
Object.defineProperty(exports, "WebpackBitReporterPlugin", {
  enumerable: true,
  get: function () {
    return _webpackBitReporterPlugin().WebpackBitReporterPlugin;
  }
});
Object.defineProperty(exports, "WebpackBundler", {
  enumerable: true,
  get: function () {
    return _webpack3().WebpackBundler;
  }
});
Object.defineProperty(exports, "WebpackConfigMutator", {
  enumerable: true,
  get: function () {
    return _webpackModules().WebpackConfigMutator;
  }
});
Object.defineProperty(exports, "WebpackDevServer", {
  enumerable: true,
  get: function () {
    return _webpack2().WebpackDevServer;
  }
});
Object.defineProperty(exports, "fallbacks", {
  enumerable: true,
  get: function () {
    return _webpackFallbacks().fallbacks;
  }
});
Object.defineProperty(exports, "fallbacksAliases", {
  enumerable: true,
  get: function () {
    return _webpackFallbacksAliases().fallbacksAliases;
  }
});
Object.defineProperty(exports, "fallbacksProvidePluginConfig", {
  enumerable: true,
  get: function () {
    return _webpackFallbacksProvidePluginConfig().fallbacksProvidePluginConfig;
  }
});
Object.defineProperty(exports, "generateAddAliasesFromPeersTransformer", {
  enumerable: true,
  get: function () {
    return _transformers().generateAddAliasesFromPeersTransformer;
  }
});
Object.defineProperty(exports, "generateExposePeersTransformer", {
  enumerable: true,
  get: function () {
    return _transformers().generateExposePeersTransformer;
  }
});
Object.defineProperty(exports, "generateExternalsTransformer", {
  enumerable: true,
  get: function () {
    return _transformers().generateExternalsTransformer;
  }
});
Object.defineProperty(exports, "runTransformersWithContext", {
  enumerable: true,
  get: function () {
    return _webpackMain().runTransformersWithContext;
  }
});
function _webpackMain() {
  const data = require("./webpack.main.runtime");
  _webpackMain = function () {
    return data;
  };
  return data;
}
function _webpack() {
  const data = require("./webpack.aspect");
  _webpack = function () {
    return data;
  };
  return data;
}
function _webpack2() {
  const data = require("./webpack.dev-server");
  _webpack2 = function () {
    return data;
  };
  return data;
}
function _webpack3() {
  const data = require("./webpack.bundler");
  _webpack3 = function () {
    return data;
  };
  return data;
}
var _events = require("./events");
Object.keys(_events).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _events[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _events[key];
    }
  });
});
function _webpack4() {
  const data = require("webpack");
  _webpack4 = function () {
    return data;
  };
  return data;
}
function _webpackModules() {
  const data = require("@teambit/webpack.modules.config-mutator");
  _webpackModules = function () {
    return data;
  };
  return data;
}
function _webpackBitReporterPlugin() {
  const data = require("./plugins/webpack-bit-reporter-plugin");
  _webpackBitReporterPlugin = function () {
    return data;
  };
  return data;
}
function _webpackFallbacks() {
  const data = require("./config/webpack-fallbacks");
  _webpackFallbacks = function () {
    return data;
  };
  return data;
}
function _webpackFallbacksAliases() {
  const data = require("./config/webpack-fallbacks-aliases");
  _webpackFallbacksAliases = function () {
    return data;
  };
  return data;
}
function _webpackFallbacksProvidePluginConfig() {
  const data = require("./config/webpack-fallbacks-provide-plugin-config");
  _webpackFallbacksProvidePluginConfig = function () {
    return data;
  };
  return data;
}
function _transformers() {
  const data = require("./transformers");
  _transformers = function () {
    return data;
  };
  return data;
}
