// @teambit/aspect/dist/templates/aspect/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.aspectTemplate = void 0;
function _index() {
  const data = require("./files/index");
  _index = function () {
    return data;
  };
  return data;
}
function _aspectFile() {
  const data = require("./files/aspect-file");
  _aspectFile = function () {
    return data;
  };
  return data;
}
function _mainRuntime() {
  const data = require("./files/main-runtime");
  _mainRuntime = function () {
    return data;
  };
  return data;
}
const aspectTemplate = {
  name: 'aspect',
  description: 'extend Bit capabilities',
  generateFiles: context => {
    return [{
      relativePath: 'index.ts',
      content: (0, _index().indexFile)(context),
      isMain: true
    }, {
      relativePath: `${context.name}.aspect.ts`,
      content: (0, _aspectFile().aspectFile)(context)
    }, {
      relativePath: `${context.name}.main.runtime.ts`,
      content: (0, _mainRuntime().mainRuntime)(context)
    }];
  },
  config: {
    'teambit.harmony/aspect': {},
    'teambit.envs/envs': {
      env: 'teambit.harmony/aspect'
    }
  }
};
exports.aspectTemplate = aspectTemplate;
