// @teambit/dependency-resolver/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "BIT_DEV_REGISTRY", {
  enumerable: true,
  get: function () {
    return _dependencyResolverMain().BIT_DEV_REGISTRY;
  }
});
Object.defineProperty(exports, "BaseDependency", {
  enumerable: true,
  get: function () {
    return _dependencies().BaseDependency;
  }
});
Object.defineProperty(exports, "ComponentDependency", {
  enumerable: true,
  get: function () {
    return _dependencies().ComponentDependency;
  }
});
Object.defineProperty(exports, "ComponentManifest", {
  enumerable: true,
  get: function () {
    return _manifest().ComponentManifest;
  }
});
Object.defineProperty(exports, "ComponentsManifestsMap", {
  enumerable: true,
  get: function () {
    return _types().ComponentsManifestsMap;
  }
});
Object.defineProperty(exports, "CoreAspectLinkResult", {
  enumerable: true,
  get: function () {
    return _dependencyLinker().CoreAspectLinkResult;
  }
});
Object.defineProperty(exports, "CreateFromComponentsOptions", {
  enumerable: true,
  get: function () {
    return _manifest().CreateFromComponentsOptions;
  }
});
Object.defineProperty(exports, "DependenciesManifest", {
  enumerable: true,
  get: function () {
    return _dependencies().DependenciesManifest;
  }
});
Object.defineProperty(exports, "Dependency", {
  enumerable: true,
  get: function () {
    return _dependencies().Dependency;
  }
});
Object.defineProperty(exports, "DependencyDetector", {
  enumerable: true,
  get: function () {
    return _dependencyDetector().DependencyDetector;
  }
});
Object.defineProperty(exports, "DependencyFactory", {
  enumerable: true,
  get: function () {
    return _dependencies().DependencyFactory;
  }
});
Object.defineProperty(exports, "DependencyInstaller", {
  enumerable: true,
  get: function () {
    return _dependencyInstaller().DependencyInstaller;
  }
});
Object.defineProperty(exports, "DependencyLifecycleType", {
  enumerable: true,
  get: function () {
    return _dependencies().DependencyLifecycleType;
  }
});
Object.defineProperty(exports, "DependencyList", {
  enumerable: true,
  get: function () {
    return _dependencies().DependencyList;
  }
});
Object.defineProperty(exports, "DependencyResolverAspect", {
  enumerable: true,
  get: function () {
    return _dependencyResolver().DependencyResolverAspect;
  }
});
Object.defineProperty(exports, "DependencySource", {
  enumerable: true,
  get: function () {
    return _variantPolicy().DependencySource;
  }
});
Object.defineProperty(exports, "DepsLinkedToEnvResult", {
  enumerable: true,
  get: function () {
    return _dependencyLinker().DepsLinkedToEnvResult;
  }
});
Object.defineProperty(exports, "EnvPolicy", {
  enumerable: true,
  get: function () {
    return _policy().EnvPolicy;
  }
});
Object.defineProperty(exports, "EnvPolicyConfigObject", {
  enumerable: true,
  get: function () {
    return _policy().EnvPolicyConfigObject;
  }
});
Object.defineProperty(exports, "FileContext", {
  enumerable: true,
  get: function () {
    return _dependencyDetector().FileContext;
  }
});
Object.defineProperty(exports, "InstallArgs", {
  enumerable: true,
  get: function () {
    return _dependencyInstaller().InstallArgs;
  }
});
Object.defineProperty(exports, "InstallOptions", {
  enumerable: true,
  get: function () {
    return _dependencyInstaller().InstallOptions;
  }
});
Object.defineProperty(exports, "InstallationContext", {
  enumerable: true,
  get: function () {
    return _packageManager().InstallationContext;
  }
});
Object.defineProperty(exports, "KEY_NAME_BY_LIFECYCLE_TYPE", {
  enumerable: true,
  get: function () {
    return _dependencies().KEY_NAME_BY_LIFECYCLE_TYPE;
  }
});
Object.defineProperty(exports, "LinkDetail", {
  enumerable: true,
  get: function () {
    return _dependencyLinker().LinkDetail;
  }
});
Object.defineProperty(exports, "LinkResults", {
  enumerable: true,
  get: function () {
    return _dependencyLinker().LinkResults;
  }
});
Object.defineProperty(exports, "LinkToDirResult", {
  enumerable: true,
  get: function () {
    return _dependencyLinker().LinkToDirResult;
  }
});
Object.defineProperty(exports, "LinkingOptions", {
  enumerable: true,
  get: function () {
    return _dependencyLinker().LinkingOptions;
  }
});
Object.defineProperty(exports, "ManifestDependenciesObject", {
  enumerable: true,
  get: function () {
    return _manifest().ManifestDependenciesObject;
  }
});
Object.defineProperty(exports, "NPM_REGISTRY", {
  enumerable: true,
  get: function () {
    return _dependencyResolverMain().NPM_REGISTRY;
  }
});
Object.defineProperty(exports, "NestedNMDepsLinksResult", {
  enumerable: true,
  get: function () {
    return _dependencyLinker().NestedNMDepsLinksResult;
  }
});
Object.defineProperty(exports, "OutdatedPkg", {
  enumerable: true,
  get: function () {
    return _getAllPolicyPkgs().OutdatedPkg;
  }
});
Object.defineProperty(exports, "PackageManager", {
  enumerable: true,
  get: function () {
    return _packageManager().PackageManager;
  }
});
Object.defineProperty(exports, "PackageManagerInstallOptions", {
  enumerable: true,
  get: function () {
    return _packageManager().PackageManagerInstallOptions;
  }
});
Object.defineProperty(exports, "PackageManagerNetworkConfig", {
  enumerable: true,
  get: function () {
    return _dependencyResolverMain().NetworkConfig;
  }
});
Object.defineProperty(exports, "PackageManagerProxyConfig", {
  enumerable: true,
  get: function () {
    return _dependencyResolverMain().ProxyConfig;
  }
});
Object.defineProperty(exports, "PackageManagerResolveRemoteVersionOptions", {
  enumerable: true,
  get: function () {
    return _packageManager().PackageManagerResolveRemoteVersionOptions;
  }
});
Object.defineProperty(exports, "PeersAutoDetectPolicy", {
  enumerable: true,
  get: function () {
    return _policy().PeersAutoDetectPolicy;
  }
});
Object.defineProperty(exports, "Policy", {
  enumerable: true,
  get: function () {
    return _policy().Policy;
  }
});
Object.defineProperty(exports, "PolicyConfigKeys", {
  enumerable: true,
  get: function () {
    return _policy().PolicyConfigKeys;
  }
});
Object.defineProperty(exports, "PolicyConfigKeysNames", {
  enumerable: true,
  get: function () {
    return _policy().PolicyConfigKeysNames;
  }
});
Object.defineProperty(exports, "PolicyEntry", {
  enumerable: true,
  get: function () {
    return _policy().PolicyEntry;
  }
});
Object.defineProperty(exports, "PolicySemver", {
  enumerable: true,
  get: function () {
    return _policy().PolicySemver;
  }
});
Object.defineProperty(exports, "RawComponentState", {
  enumerable: true,
  get: function () {
    return _types().RawComponentState;
  }
});
Object.defineProperty(exports, "Registries", {
  enumerable: true,
  get: function () {
    return _registry().Registries;
  }
});
Object.defineProperty(exports, "RegistriesMap", {
  enumerable: true,
  get: function () {
    return _types().RegistriesMap;
  }
});
Object.defineProperty(exports, "Registry", {
  enumerable: true,
  get: function () {
    return _registry().Registry;
  }
});
Object.defineProperty(exports, "ResolvedPackageVersion", {
  enumerable: true,
  get: function () {
    return _packageManager().ResolvedPackageVersion;
  }
});
Object.defineProperty(exports, "SemverVersion", {
  enumerable: true,
  get: function () {
    return _dependencies().SemverVersion;
  }
});
Object.defineProperty(exports, "SerializedDependency", {
  enumerable: true,
  get: function () {
    return _dependencies().SerializedDependency;
  }
});
Object.defineProperty(exports, "SerializedVariantPolicy", {
  enumerable: true,
  get: function () {
    return _policy().SerializedVariantPolicy;
  }
});
Object.defineProperty(exports, "VariantPolicy", {
  enumerable: true,
  get: function () {
    return _policy().VariantPolicy;
  }
});
Object.defineProperty(exports, "VariantPolicyConfigObject", {
  enumerable: true,
  get: function () {
    return _policy().VariantPolicyConfigObject;
  }
});
Object.defineProperty(exports, "VariantPolicyFactory", {
  enumerable: true,
  get: function () {
    return _policy().VariantPolicyFactory;
  }
});
Object.defineProperty(exports, "WorkspaceDependencyLifecycleType", {
  enumerable: true,
  get: function () {
    return _dependencies().WorkspaceDependencyLifecycleType;
  }
});
Object.defineProperty(exports, "WorkspaceManifest", {
  enumerable: true,
  get: function () {
    return _manifest().WorkspaceManifest;
  }
});
Object.defineProperty(exports, "WorkspacePolicy", {
  enumerable: true,
  get: function () {
    return _policy().WorkspacePolicy;
  }
});
Object.defineProperty(exports, "WorkspacePolicyEntry", {
  enumerable: true,
  get: function () {
    return _policy().WorkspacePolicyEntry;
  }
});
Object.defineProperty(exports, "extendWithComponentsFromDir", {
  enumerable: true,
  get: function () {
    return _extendWithComponentsFromDir().extendWithComponentsFromDir;
  }
});
function _types() {
  const data = require("./types");
  _types = function () {
    return data;
  };
  return data;
}
function _manifest() {
  const data = require("./manifest");
  _manifest = function () {
    return data;
  };
  return data;
}
function _registry() {
  const data = require("./registry");
  _registry = function () {
    return data;
  };
  return data;
}
function _packageManager() {
  const data = require("./package-manager");
  _packageManager = function () {
    return data;
  };
  return data;
}
function _dependencyResolverMain() {
  const data = require("./dependency-resolver.main.runtime");
  _dependencyResolverMain = function () {
    return data;
  };
  return data;
}
function _dependencyResolver() {
  const data = require("./dependency-resolver.aspect");
  _dependencyResolver = function () {
    return data;
  };
  return data;
}
function _dependencies() {
  const data = require("./dependencies");
  _dependencies = function () {
    return data;
  };
  return data;
}
function _policy() {
  const data = require("./policy");
  _policy = function () {
    return data;
  };
  return data;
}
function _dependencyLinker() {
  const data = require("./dependency-linker");
  _dependencyLinker = function () {
    return data;
  };
  return data;
}
function _dependencyInstaller() {
  const data = require("./dependency-installer");
  _dependencyInstaller = function () {
    return data;
  };
  return data;
}
function _dependencyDetector() {
  const data = require("./dependency-detector");
  _dependencyDetector = function () {
    return data;
  };
  return data;
}
function _variantPolicy() {
  const data = require("./policy/variant-policy/variant-policy");
  _variantPolicy = function () {
    return data;
  };
  return data;
}
function _getAllPolicyPkgs() {
  const data = require("./get-all-policy-pkgs");
  _getAllPolicyPkgs = function () {
    return data;
  };
  return data;
}
function _extendWithComponentsFromDir() {
  const data = require("./extend-with-components-from-dir");
  _extendWithComponentsFromDir = function () {
    return data;
  };
  return data;
}
