// @teambit/graph/dist/graph.compare.section.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GraphCompareSection = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _react() {
  const data = _interopRequireDefault(require("react"));
  _react = function () {
    return data;
  };
  return data;
}
function _graph() {
  const data = require("@teambit/graph");
  _graph = function () {
    return data;
  };
  return data;
}
class GraphCompareSection {
  constructor() {
    (0, _defineProperty2().default)(this, "navigationLink", {
      href: 'dependencies',
      children: 'Dependencies',
      order: 2
    });
    (0, _defineProperty2().default)(this, "route", {
      path: 'dependencies/*',
      element: /*#__PURE__*/_react().default.createElement(_graph().DependenciesCompare, null)
    });
  }
}
exports.GraphCompareSection = GraphCompareSection;
