// @teambit/graph/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "DependenciesCompare", {
  enumerable: true,
  get: function () {
    return _dependenciesCompare().DependenciesCompare;
  }
});
Object.defineProperty(exports, "Dependency", {
  enumerable: true,
  get: function () {
    return _dependency().Dependency;
  }
});
Object.defineProperty(exports, "DuplicateDependency", {
  enumerable: true,
  get: function () {
    return _duplicateDependency().DuplicateDependency;
  }
});
Object.defineProperty(exports, "EdgeModel", {
  enumerable: true,
  get: function () {
    return _query().EdgeModel;
  }
});
Object.defineProperty(exports, "EdgeType", {
  enumerable: true,
  get: function () {
    return _edgeType().EdgeType;
  }
});
Object.defineProperty(exports, "GraphAspect", {
  enumerable: true,
  get: function () {
    return _graph().GraphAspect;
  }
});
Object.defineProperty(exports, "GraphFilter", {
  enumerable: true,
  get: function () {
    return _graphFilters().GraphFilter;
  }
});
Object.defineProperty(exports, "GraphFilters", {
  enumerable: true,
  get: function () {
    return _graphPage().GraphFilters;
  }
});
Object.defineProperty(exports, "GraphModel", {
  enumerable: true,
  get: function () {
    return _query().GraphModel;
  }
});
Object.defineProperty(exports, "IdGraph", {
  enumerable: true,
  get: function () {
    return _objectListToGraph().IdGraph;
  }
});
Object.defineProperty(exports, "NodeModel", {
  enumerable: true,
  get: function () {
    return _query().NodeModel;
  }
});
Object.defineProperty(exports, "RawGraph", {
  enumerable: true,
  get: function () {
    return _query().RawGraph;
  }
});
Object.defineProperty(exports, "VersionSubgraph", {
  enumerable: true,
  get: function () {
    return _duplicateDependency().VersionSubgraph;
  }
});
Object.defineProperty(exports, "bitObjectListToGraph", {
  enumerable: true,
  get: function () {
    return _objectListToGraph().bitObjectListToGraph;
  }
});
Object.defineProperty(exports, "calcElements", {
  enumerable: true,
  get: function () {
    return _dependenciesGraph().calcElements;
  }
});
Object.defineProperty(exports, "calcLayout", {
  enumerable: true,
  get: function () {
    return _dependenciesGraph().calcLayout;
  }
});
Object.defineProperty(exports, "calcMinimapColors", {
  enumerable: true,
  get: function () {
    return _dependenciesGraph().calcMinimapColors;
  }
});
Object.defineProperty(exports, "componentNodeStyles", {
  enumerable: true,
  get: function () {
    return _componentNode().styles;
  }
});
Object.defineProperty(exports, "default", {
  enumerable: true,
  get: function () {
    return _graph().GraphAspect;
  }
});
Object.defineProperty(exports, "defaultNode", {
  enumerable: true,
  get: function () {
    return _componentNode().defaultNode;
  }
});
Object.defineProperty(exports, "depTypeToClass", {
  enumerable: true,
  get: function () {
    return _dependenciesGraph().depTypeToClass;
  }
});
Object.defineProperty(exports, "depTypeToLabel", {
  enumerable: true,
  get: function () {
    return _dependenciesGraph().depTypeToLabel;
  }
});
Object.defineProperty(exports, "dependenciesGraphStyles", {
  enumerable: true,
  get: function () {
    return _dependenciesGraph().styles;
  }
});
Object.defineProperty(exports, "external", {
  enumerable: true,
  get: function () {
    return _componentNode().external;
  }
});
Object.defineProperty(exports, "graphPageStyles", {
  enumerable: true,
  get: function () {
    return _graphPage().styles;
  }
});
Object.defineProperty(exports, "objectListToGraph", {
  enumerable: true,
  get: function () {
    return _objectListToGraph().objectListToGraph;
  }
});
Object.defineProperty(exports, "root", {
  enumerable: true,
  get: function () {
    return _componentNode().root;
  }
});
Object.defineProperty(exports, "useGraph", {
  enumerable: true,
  get: function () {
    return _query().useGraph;
  }
});
Object.defineProperty(exports, "useGraphQuery", {
  enumerable: true,
  get: function () {
    return _query().useGraphQuery;
  }
});
function _duplicateDependency() {
  const data = require("./duplicate-dependency");
  _duplicateDependency = function () {
    return data;
  };
  return data;
}
function _edgeType() {
  const data = require("./edge-type");
  _edgeType = function () {
    return data;
  };
  return data;
}
function _graph() {
  const data = require("./graph.aspect");
  _graph = function () {
    return data;
  };
  return data;
}
function _dependency() {
  const data = require("./model/dependency");
  _dependency = function () {
    return data;
  };
  return data;
}
function _graphFilters() {
  const data = require("./model/graph-filters");
  _graphFilters = function () {
    return data;
  };
  return data;
}
function _objectListToGraph() {
  const data = require("./object-list-to-graph");
  _objectListToGraph = function () {
    return data;
  };
  return data;
}
function _dependenciesCompare() {
  const data = require("./ui/dependencies-compare");
  _dependenciesCompare = function () {
    return data;
  };
  return data;
}
function _dependenciesGraph() {
  const data = require("./ui/dependencies-graph");
  _dependenciesGraph = function () {
    return data;
  };
  return data;
}
function _graphPage() {
  const data = require("./ui/graph-page");
  _graphPage = function () {
    return data;
  };
  return data;
}
function _query() {
  const data = require("./ui/query");
  _query = function () {
    return data;
  };
  return data;
}
function _componentNode() {
  const data = require("./ui/component-node");
  _componentNode = function () {
    return data;
  };
  return data;
}
