// @teambit/preview/dist/strategies/generate-component-link.js
"use strict";

require("core-js/modules/es.array.iterator.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.generateComponentLink = generateComponentLink;
function _toolboxPath() {
  const data = require("@teambit/toolbox.path.to-windows-compatible-path");
  _toolboxPath = function () {
    return data;
  };
  return data;
}
function generateComponentLink(modules) {
  const links = modules.map(({
    prefix,
    paths
  }) => ({
    name: prefix,
    entries: paths.map((path, idx) => ({
      path: (0, _toolboxPath().toWindowsCompatiblePath)(path),
      linkName: `${prefix}_${idx}`
    }))
  }));

  // import per preview file
  const importStr = links.map(({
    entries
  }) => entries.map(({
    path,
    linkName
  }) => `import * as ${linkName} from '${path}'`).join(';\n')).join(';\n');

  // export files group per preview
  const exportsString = links.map(({
    name,
    entries
  }) => `export const ${name} = [${entries.map(entry => entry.linkName).join(', ')}]`).join(';\n');
  const exportsMetadataString = modules.filter(mod => mod.metadata).map(mod => `export const ${mod.prefix}_metadata = ${JSON.stringify(mod.metadata)}`).join(';\n');
  return `${importStr};

${exportsString};

${exportsMetadataString};
`;
}
