// @teambit/legacy/dist/scope/actions/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Action", {
  enumerable: true,
  get: function () {
    return _action().Action;
  }
});
Object.defineProperty(exports, "ExportPersist", {
  enumerable: true,
  get: function () {
    return _exportPersist().ExportPersist;
  }
});
Object.defineProperty(exports, "ExportValidate", {
  enumerable: true,
  get: function () {
    return _exportValidate().ExportValidate;
  }
});
Object.defineProperty(exports, "FetchMissingDeps", {
  enumerable: true,
  get: function () {
    return _fetchMissingDeps().FetchMissingDeps;
  }
});
Object.defineProperty(exports, "FetchMissingHistory", {
  enumerable: true,
  get: function () {
    return _fetchMissingHistory().FetchMissingHistory;
  }
});
Object.defineProperty(exports, "PostSign", {
  enumerable: true,
  get: function () {
    return _postSign().PostSign;
  }
});
Object.defineProperty(exports, "RemovePendingDir", {
  enumerable: true,
  get: function () {
    return _removePendingDir().RemovePendingDir;
  }
});
function _action() {
  const data = require("./action");
  _action = function () {
    return data;
  };
  return data;
}
function _exportPersist() {
  const data = require("./export-persist");
  _exportPersist = function () {
    return data;
  };
  return data;
}
function _exportValidate() {
  const data = require("./export-validate");
  _exportValidate = function () {
    return data;
  };
  return data;
}
function _fetchMissingHistory() {
  const data = require("./fetch-missing-history");
  _fetchMissingHistory = function () {
    return data;
  };
  return data;
}
function _removePendingDir() {
  const data = require("./remove-pending-dir");
  _removePendingDir = function () {
    return data;
  };
  return data;
}
function _fetchMissingDeps() {
  const data = require("./fetch-missing-deps");
  _fetchMissingDeps = function () {
    return data;
  };
  return data;
}
function _postSign() {
  const data = require("./post-sign");
  _postSign = function () {
    return data;
  };
  return data;
}
