// @teambit/graph/dist/ui/component-node/component-node.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.array.iterator.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ComponentNode = ComponentNode;
function _extends2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
  _extends2 = function () {
    return data;
  };
  return data;
}
function _objectWithoutProperties2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
  _objectWithoutProperties2 = function () {
    return data;
  };
  return data;
}
function _react() {
  const data = _interopRequireWildcard(require("react"));
  _react = function () {
    return data;
  };
  return data;
}
function _classnames() {
  const data = _interopRequireDefault(require("classnames"));
  _classnames = function () {
    return data;
  };
  return data;
}
function _baseUiText() {
  const data = require("@teambit/base-ui.text.muted-text");
  _baseUiText = function () {
    return data;
  };
  return data;
}
function _componentUi() {
  const data = require("@teambit/component.ui.deprecation-icon");
  _componentUi = function () {
    return data;
  };
  return data;
}
function _envsUi() {
  const data = require("@teambit/envs.ui.env-icon");
  _envsUi = function () {
    return data;
  };
  return data;
}
function _designUiStyles() {
  const data = require("@teambit/design.ui.styles.ellipsis");
  _designUiStyles = function () {
    return data;
  };
  return data;
}
function _baseUiSurfaces() {
  const data = require("@teambit/base-ui.surfaces.card");
  _baseUiSurfaces = function () {
    return data;
  };
  return data;
}
function _baseUiRouting() {
  const data = require("@teambit/base-ui.routing.nav-link");
  _baseUiRouting = function () {
    return data;
  };
  return data;
}
function _componentModules() {
  const data = require("@teambit/component.modules.component-url");
  _componentModules = function () {
    return data;
  };
  return data;
}
function _dependenciesGraph() {
  const data = require("../dependencies-graph/");
  _dependenciesGraph = function () {
    return data;
  };
  return data;
}
function _componentNodeModule() {
  const data = _interopRequireDefault(require("./component-node.module.scss"));
  _componentNodeModule = function () {
    return data;
  };
  return data;
}
function _variantsModule() {
  const data = _interopRequireDefault(require("./variants.module.scss"));
  _variantsModule = function () {
    return data;
  };
  return data;
}
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
// keep order: styles, then variants

function ComponentNode(_ref) {
  let {
      node,
      type = 'defaultNode'
    } = _ref,
    rest = (0, _objectWithoutProperties2().default)(_ref, ["node", "type"]);
  const graphContext = (0, _react().useContext)(_dependenciesGraph().ComponentGraphContext);
  const {
    component
  } = node;
  const {
    id
  } = component;
  return /*#__PURE__*/_react().default.createElement(_baseUiSurfaces().Card, (0, _extends2().default)({
    className: (0, _classnames().default)(_componentNodeModule().default.compNode, _variantsModule().default[type]),
    elevation: "none"
  }, rest), /*#__PURE__*/_react().default.createElement("div", {
    className: _componentNodeModule().default.firstRow
  }, /*#__PURE__*/_react().default.createElement(_envsUi().EnvIcon, {
    component: component,
    className: _componentNodeModule().default.envIcon
  }), /*#__PURE__*/_react().default.createElement(Breadcrumbs, {
    componentId: id,
    className: _baseUiText().mutedText
  })), /*#__PURE__*/_react().default.createElement("div", {
    className: _componentNodeModule().default.nameLine
  }, /*#__PURE__*/_react().default.createElement(_baseUiRouting().NavLink, {
    className: _componentNodeModule().default.link,
    external: true,
    href: _componentModules().ComponentUrl.toUrl(id, {
      includeVersion: false
    })
  }, /*#__PURE__*/_react().default.createElement("span", {
    className: (0, _classnames().default)(_componentNodeModule().default.name, _designUiStyles().ellipsis)
  }, id.name)), id.version && /*#__PURE__*/_react().default.createElement("span", {
    className: (0, _classnames().default)(_componentNodeModule().default.version, _designUiStyles().ellipsis)
  }, id.version), /*#__PURE__*/_react().default.createElement("div", {
    className: _componentNodeModule().default.buffs
  }, /*#__PURE__*/_react().default.createElement(_componentUi().DeprecationIcon, {
    component: component
  }), graphContext && graphContext.componentWidgets.toArray().map(([widgetId, Widget]) => /*#__PURE__*/_react().default.createElement(Widget, {
    key: widgetId,
    component: component
  })))));
}
function Breadcrumbs(_ref2) {
  let {
      componentId,
      className
    } = _ref2,
    rest = (0, _objectWithoutProperties2().default)(_ref2, ["componentId", "className"]);
  const {
    scope,
    namespace
  } = componentId;
  const showSep = !!scope && !!namespace;
  return /*#__PURE__*/_react().default.createElement("div", (0, _extends2().default)({}, rest, {
    className: (0, _classnames().default)(_componentNodeModule().default.breadcrumbs, _designUiStyles().ellipsis, className)
  }), scope, showSep && '/', namespace);
}
