// @teambit/graph/dist/ui/component-node/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ComponentNode", {
  enumerable: true,
  get: function () {
    return _componentNode().ComponentNode;
  }
});
Object.defineProperty(exports, "defaultNode", {
  enumerable: true,
  get: function () {
    return _variants().defaultNode;
  }
});
Object.defineProperty(exports, "defaultNodeColor", {
  enumerable: true,
  get: function () {
    return _variants().defaultNodeColor;
  }
});
Object.defineProperty(exports, "external", {
  enumerable: true,
  get: function () {
    return _variants().external;
  }
});
Object.defineProperty(exports, "externalNodeColor", {
  enumerable: true,
  get: function () {
    return _variants().externalNodeColor;
  }
});
Object.defineProperty(exports, "root", {
  enumerable: true,
  get: function () {
    return _variants().root;
  }
});
Object.defineProperty(exports, "rootNodeColor", {
  enumerable: true,
  get: function () {
    return _variants().rootNodeColor;
  }
});
exports.styles = void 0;
function _componentNodeModule() {
  const data = _interopRequireDefault(require("./component-node.module.scss"));
  _componentNodeModule = function () {
    return data;
  };
  return data;
}
function _componentNode() {
  const data = require("./component-node");
  _componentNode = function () {
    return data;
  };
  return data;
}
function _variants() {
  const data = require("./variants");
  _variants = function () {
    return data;
  };
  return data;
}
const {
  compNode,
  firstRow,
  envIcon,
  breadcrumbs,
  nameLine,
  name,
  version,
  buffs,
  link
} = _componentNodeModule().default;
const styles = {
  compNode,
  firstRow,
  envIcon,
  breadcrumbs,
  nameLine,
  name,
  version,
  buffs,
  link
};
exports.styles = styles;
