// @teambit/compiler/dist/events/component-compilation-on-done.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ComponentCompilationOnDoneEvent = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _pubsub() {
  const data = require("@teambit/pubsub");
  _pubsub = function () {
    return data;
  };
  return data;
}
/* eslint-disable max-classes-per-file */

class ComponentCompilationOnDoneEventData {
  constructor(errors, component, buildResults) {
    this.errors = errors;
    this.component = component;
    this.buildResults = buildResults;
  }
}
class ComponentCompilationOnDoneEvent extends _pubsub().BitBaseEvent {
  constructor(errors, component, buildResults, timestamp = Date.now()) {
    super(ComponentCompilationOnDoneEvent.TYPE, '0.0.1', timestamp, new ComponentCompilationOnDoneEventData(errors, component, buildResults));
    this.errors = errors;
    this.component = component;
    this.buildResults = buildResults;
    this.timestamp = timestamp;
  }
}
exports.ComponentCompilationOnDoneEvent = ComponentCompilationOnDoneEvent;
(0, _defineProperty2().default)(ComponentCompilationOnDoneEvent, "TYPE", 'component-compilation-on-done');
