// @teambit/legacy/dist/scope/component-dependencies.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.array.iterator.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
class ComponentWithDependencies {
  constructor(props) {
    (0, _defineProperty2().default)(this, "component", void 0);
    (0, _defineProperty2().default)(this, "dependencies", void 0);
    (0, _defineProperty2().default)(this, "devDependencies", void 0);
    (0, _defineProperty2().default)(this, "extensionDependencies", void 0);
    (0, _defineProperty2().default)(this, "missingDependencies", void 0);
    this.component = props.component;
    this.dependencies = props.dependencies || [];
    this.devDependencies = props.devDependencies || [];
    this.extensionDependencies = props.extensionDependencies || [];
    this.missingDependencies = props.missingDependencies || [];
  }

  // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
  get allDependencies() {
    return [...this.dependencies, ...this.devDependencies, ...this.extensionDependencies];
  }
  hasDependency(id) {
    this.allDependencies.some(dependency => dependency.id.isEqual(id));
  }
}
exports.default = ComponentWithDependencies;
