// @teambit/legacy/dist/scope/lane-objects.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.array.iterator.js");
require("core-js/modules/es.promise.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _utils() {
  const data = require("../utils");
  _utils = function () {
    return data;
  };
  return data;
}
function _object() {
  const data = _interopRequireDefault(require("./objects/object"));
  _object = function () {
    return data;
  };
  return data;
}
class LaneObjects {
  constructor(lane, objects) {
    (0, _defineProperty2().default)(this, "lane", void 0);
    (0, _defineProperty2().default)(this, "objects", void 0);
    this.lane = lane;
    this.objects = objects;
  }
  toString() {
    return JSON.stringify({
      lane: (0, _utils().toBase64ArrayBuffer)(this.lane),
      objects: this.objects.map(_utils().toBase64ArrayBuffer)
    });
  }
  static fromString(str) {
    return LaneObjects.fromObject(JSON.parse(str));
  }
  static fromObject(object) {
    const {
      lane,
      objects
    } = object;
    return new LaneObjects(_from64Buffer(lane), objects.map(_from64Buffer));
  }

  /**
   * prefer using `this.toObjectsAsync()` if not must to be sync.
   */
  toObjects() {
    return {
      // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
      lane: _object().default.parseSync(this.lane),
      objects: this.objects.map(obj => _object().default.parseSync(obj))
    };
  }
  /**
   * see `this.toObject()` for the sync version
   */
  async toObjectsAsync() {
    return {
      // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
      lane: await _object().default.parseObject(this.lane),
      objects: await Promise.all(this.objects.map(obj => _object().default.parseObject(obj)))
    };
  }
}
exports.default = LaneObjects;
function _from64Buffer(val) {
  return Buffer.from(val, 'base64');
}
