// @teambit/graph/dist/ui/graph.section.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GraphSection = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _react() {
  const data = _interopRequireDefault(require("react"));
  _react = function () {
    return data;
  };
  return data;
}
function _graphPage() {
  const data = require("./graph-page");
  _graphPage = function () {
    return data;
  };
  return data;
}
class GraphSection {
  constructor(componentWidgetSlot) {
    this.componentWidgetSlot = componentWidgetSlot;
    (0, _defineProperty2().default)(this, "route", {
      path: '~dependencies',
      element: /*#__PURE__*/_react().default.createElement(_graphPage().GraphPage, {
        componentWidgets: this.componentWidgetSlot
      })
    });
    (0, _defineProperty2().default)(this, "navigationLink", {
      href: '~dependencies',
      children: 'Dependencies'
    });
    (0, _defineProperty2().default)(this, "order", 40);
  }
}
exports.GraphSection = GraphSection;
