// @teambit/react/dist/react.graphql.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.promise.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.reactSchema = reactSchema;
function _graphqlTag() {
  const data = _interopRequireDefault(require("graphql-tag"));
  _graphqlTag = function () {
    return data;
  };
  return data;
}
function reactSchema(react) {
  return {
    typeDefs: (0, _graphqlTag().default)`
      extend type ComponentHost {
        getDocs(id: String!): ReactDocs
      }

      type ReactDocs {
        abstract: String
        filePath: String
        properties: [Property]
      }

      type Property {
        name: String
        description: String
        required: Boolean
        type: String
        defaultValue: DefaultValue
      }

      type DefaultValue {
        value: String
        computed: Boolean
      }
    `,
    resolvers: {
      ComponentHost: {
        getDocs: async (host, {
          id
        }) => {
          const componentId = await host.resolveComponentId(id);
          const component = await host.get(componentId);
          const empty = {
            abstract: '',
            filePath: '',
            properties: []
          };
          if (!component) return empty;
          const docs = react.getDocs(component);
          if (!docs) return empty;
          return docs;
        }
      }
    }
  };
}
