// @teambit/deprecation/dist/deprecation.graphql.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.deprecationSchema = deprecationSchema;
function _graphqlTag() {
  const data = _interopRequireDefault(require("graphql-tag"));
  _graphqlTag = function () {
    return data;
  };
  return data;
}
function deprecationSchema(deprecation) {
  return {
    typeDefs: (0, _graphqlTag().default)`
      extend type Component {
        deprecation: DeprecationInfo
      }

      type DeprecationInfo {
        isDeprecate: Boolean
        newId: String
      }

      type DeprecationResult {
        bitIds: [String]
        missingComponents: [String]
      }

      type Mutation {
        # deprecate components
        deprecate(bitIds: [String!]!): DeprecationResult

        # undo deprecate to components
        undeprecate(bitIds: [String!]!): DeprecationResult
      }
    `,
    resolvers: {
      // Mutation: {
      //   deprecate: (req: any, { bitIds }: { bitIds: string[] }, context: { verb: string }) => {
      //     if (context.verb !== 'write') throw new Error('You are not authorized');
      //     return deprecation.deprecate(bitIds);
      //   },

      //   undeprecate: (req: any, { bitIds }: { bitIds: string[] }, context: { verb: string }) => {
      //     if (context.verb !== 'write') throw new Error('You are not authorized');
      //     return deprecation.unDeprecate(bitIds);
      //   },
      // },
      Component: {
        deprecation: component => {
          return deprecation.getDeprecationInfo(component);
        }
      }
    }
  };
}
