// @teambit/deprecation/dist/deprecation.ui.runtime.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.array.iterator.js");
require("core-js/modules/es.promise.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DeprecationUIRuntime = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _react() {
  const data = _interopRequireDefault(require("react"));
  _react = function () {
    return data;
  };
  return data;
}
function _ui() {
  const data = require("@teambit/ui");
  _ui = function () {
    return data;
  };
  return data;
}
function _docs() {
  const data = require("@teambit/docs");
  _docs = function () {
    return data;
  };
  return data;
}
function _componentUi() {
  const data = require("@teambit/component.ui.component-deprecated");
  _componentUi = function () {
    return data;
  };
  return data;
}
function _deprecation() {
  const data = require("./deprecation.aspect");
  _deprecation = function () {
    return data;
  };
  return data;
}
class DeprecationUIRuntime {
  static async provider([docsUI]) {
    docsUI.registerTitleBadge({
      component: function Badge({
        legacyComponentModel
      }) {
        return /*#__PURE__*/_react().default.createElement(_componentUi().ComponentDeprecated, {
          deprecation: legacyComponentModel.deprecation
        });
      },
      weight: 40
    });
  }
}
exports.DeprecationUIRuntime = DeprecationUIRuntime;
(0, _defineProperty2().default)(DeprecationUIRuntime, "dependencies", [_docs().DocsAspect]);
(0, _defineProperty2().default)(DeprecationUIRuntime, "runtime", _ui().UIRuntime);
_deprecation().DeprecationAspect.addRuntime(DeprecationUIRuntime);
