// @teambit/deprecation/dist/deprecation.fragment.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.promise.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DeprecationFragment = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
class DeprecationFragment {
  constructor(deprecation) {
    this.deprecation = deprecation;
    (0, _defineProperty2().default)(this, "title", 'deprecated');
    (0, _defineProperty2().default)(this, "weight", 3);
  }
  async renderRow(component) {
    const deprecationInfo = await this.deprecation.getDeprecationInfo(component);
    const isDeprecate = deprecationInfo.isDeprecate.toString();
    const newId = deprecationInfo.newId ? ` (new-id: ${deprecationInfo.newId})` : '';
    return {
      title: this.title,
      content: isDeprecate + newId
    };
  }
  async json(component) {
    return {
      title: this.title,
      json: await this.deprecation.getDeprecationInfo(component)
    };
  }
}
exports.DeprecationFragment = DeprecationFragment;
