// @teambit/docs/dist/docs.compare.section.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OverviewCompareSection = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _react() {
  const data = _interopRequireDefault(require("react"));
  _react = function () {
    return data;
  };
  return data;
}
function _docsUi() {
  const data = require("@teambit/docs.ui.overview-compare");
  _docsUi = function () {
    return data;
  };
  return data;
}
class OverviewCompareSection {
  constructor(titleBadgeSlot) {
    this.titleBadgeSlot = titleBadgeSlot;
    (0, _defineProperty2().default)(this, "navigationLink", {
      href: '.',
      children: 'Overview',
      exact: true,
      order: 0
    });
    (0, _defineProperty2().default)(this, "route", {
      path: '*',
      element: /*#__PURE__*/_react().default.createElement(_docsUi().OverviewCompare, {
        titleBadges: this.titleBadgeSlot
      })
    });
  }
}
exports.OverviewCompareSection = OverviewCompareSection;
