// @teambit/docs/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "BadgePosition", {
  enumerable: true,
  get: function () {
    return _overview().BadgePosition;
  }
});
Object.defineProperty(exports, "Doc", {
  enumerable: true,
  get: function () {
    return _docsEntities().Doc;
  }
});
Object.defineProperty(exports, "DocProp", {
  enumerable: true,
  get: function () {
    return _docsEntities().DocProp;
  }
});
Object.defineProperty(exports, "DocPropList", {
  enumerable: true,
  get: function () {
    return _docsEntities().DocPropList;
  }
});
Object.defineProperty(exports, "DocsAspect", {
  enumerable: true,
  get: function () {
    return _docs().DocsAspect;
  }
});
Object.defineProperty(exports, "Overview", {
  enumerable: true,
  get: function () {
    return _overview().Overview;
  }
});
exports.default = void 0;
Object.defineProperty(exports, "defaultDocs", {
  enumerable: true,
  get: function () {
    return _docs2().defaultDocs;
  }
});
function _docs() {
  const data = require("./docs.aspect");
  _docs = function () {
    return data;
  };
  return data;
}
function _docsEntities() {
  const data = require("@teambit/docs.entities.doc");
  _docsEntities = function () {
    return data;
  };
  return data;
}
function _docs2() {
  const data = require("./docs");
  _docs2 = function () {
    return data;
  };
  return data;
}
function _overview() {
  const data = require("./overview");
  _overview = function () {
    return data;
  };
  return data;
}
/**
 * Doc entity.
 * @deprecated - please dont use this. use directly from @teambit/docs.entities.doc.
 */
var _default = _docs().DocsAspect;
exports.default = _default;
