// @teambit/generator/dist/create.cmd.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.array.iterator.js");
require("core-js/modules/es.promise.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CreateCmd = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _chalk() {
  const data = _interopRequireDefault(require("chalk"));
  _chalk = function () {
    return data;
  };
  return data;
}
class CreateCmd {
  constructor(generator, docsDomain) {
    this.generator = generator;
    this.docsDomain = docsDomain;
    (0, _defineProperty2().default)(this, "name", 'create <template-name> <component-names...>');
    (0, _defineProperty2().default)(this, "description", 'create a new component (source files and config) using a template.');
    (0, _defineProperty2().default)(this, "alias", '');
    (0, _defineProperty2().default)(this, "loader", true);
    (0, _defineProperty2().default)(this, "arguments", [{
      name: 'template-name',
      description: "the template for generating the component \n(run 'bit templates' for a list of available templates)"
    }, {
      name: 'component-names...',
      description: 'a list of component names to generate'
    }]);
    (0, _defineProperty2().default)(this, "examples", [{
      cmd: 'bit create react ui/button',
      description: "creates a component named 'ui/button' using the 'react' template"
    }, {
      cmd: 'bit create react ui/button pages/register',
      description: "creates two components, 'ui/button' and 'pages/register', using the 'react' template"
    }, {
      cmd: 'bit create react ui/button --scope my-org.my-scope',
      description: "creates a component named 'ui/button' and sets it scope to 'my-org.my-scope'. \nby default, the scope is the `defaultScope` value, configured in your `workspace.jsonc`."
    }, {
      cmd: 'bit create react ui/button --env teambit.community/envs/community-react@1.95.13',
      description: "creates a component named 'ui/button' and sets it to use the 'community-react' env. \n(the template's default env is 'teambit.react/react')."
    }]);
    (0, _defineProperty2().default)(this, "group", 'development');
    (0, _defineProperty2().default)(this, "options", [['n', 'namespace <string>', `sets the component's namespace and nested dirs inside the scope`], ['s', 'scope <string>', `sets the component's scope-name. if not entered, the default-scope will be used`], ['a', 'aspect <string>', 'aspect-id of the template. helpful when multiple aspects use the same template name'], ['p', 'path <string>', 'relative path in the workspace. by default the path is `<scope>/<namespace>/<name>`'], ['e', 'env <string>', "set the component's environment. (overrides the env from variants and the template)"]]);
  }
  async report([templateName, componentNames], options) {
    const results = await this.generator.generateComponentTemplate(componentNames, templateName, options);
    const title = `${results.length} component(s) were created`;
    const componentsData = results.map(result => {
      return `${_chalk().default.bold(result.id.toString())}
    location: ${result.dir}
    env:      ${result.envId} (set by ${result.envSetBy})
`;
    }).join('\n');
    const footer = `env configuration is according to workspace variants, template config or --env flag. learn more at https://${this.docsDomain}/envs/using-envs`;
    return `${_chalk().default.green(title)}\n\n${componentsData}\n\n${footer}`;
  }
}
exports.CreateCmd = CreateCmd;
