// @teambit/generator/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ComponentConfig", {
  enumerable: true,
  get: function () {
    return _componentTemplate().ComponentConfig;
  }
});
Object.defineProperty(exports, "ComponentContext", {
  enumerable: true,
  get: function () {
    return _componentTemplate().ComponentContext;
  }
});
Object.defineProperty(exports, "ComponentFile", {
  enumerable: true,
  get: function () {
    return _componentTemplate().ComponentFile;
  }
});
Object.defineProperty(exports, "ComponentTemplate", {
  enumerable: true,
  get: function () {
    return _componentTemplate().ComponentTemplate;
  }
});
Object.defineProperty(exports, "ConfigContext", {
  enumerable: true,
  get: function () {
    return _componentTemplate().ConfigContext;
  }
});
Object.defineProperty(exports, "GeneratorAspect", {
  enumerable: true,
  get: function () {
    return _generator().GeneratorAspect;
  }
});
Object.defineProperty(exports, "Starter", {
  enumerable: true,
  get: function () {
    return _workspaceTemplate().WorkspaceTemplate;
  }
});
Object.defineProperty(exports, "WorkspaceContext", {
  enumerable: true,
  get: function () {
    return _workspaceTemplate().WorkspaceContext;
  }
});
Object.defineProperty(exports, "WorkspaceFile", {
  enumerable: true,
  get: function () {
    return _workspaceTemplate().WorkspaceFile;
  }
});
Object.defineProperty(exports, "WorkspaceTemplate", {
  enumerable: true,
  get: function () {
    return _workspaceTemplate().WorkspaceTemplate;
  }
});
function _componentTemplate() {
  const data = require("./component-template");
  _componentTemplate = function () {
    return data;
  };
  return data;
}
function _workspaceTemplate() {
  const data = require("./workspace-template");
  _workspaceTemplate = function () {
    return data;
  };
  return data;
}
function _generator() {
  const data = require("./generator.aspect");
  _generator = function () {
    return data;
  };
  return data;
}
