// @teambit/component-compare/dist/component-compare.section.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ComponentCompareSection = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _react() {
  const data = _interopRequireDefault(require("react"));
  _react = function () {
    return data;
  };
  return data;
}
function _uiFoundationUi() {
  const data = require("@teambit/ui-foundation.ui.menu-widget-icon");
  _uiFoundationUi = function () {
    return data;
  };
  return data;
}
class ComponentCompareSection {
  constructor(componentCompare) {
    this.componentCompare = componentCompare;
    (0, _defineProperty2().default)(this, "navigationLink", {
      href: '~compare',
      displayName: 'Compare',
      children: /*#__PURE__*/_react().default.createElement(_uiFoundationUi().MenuWidgetIcon, {
        icon: "compare",
        tooltipContent: "Compare"
      })
    });
    (0, _defineProperty2().default)(this, "route", {
      path: '~compare/*',
      element: this.componentCompare.getComponentComparePage()
    });
    (0, _defineProperty2().default)(this, "order", 35);
  }
}
exports.ComponentCompareSection = ComponentCompareSection;
