// @teambit/component-compare/dist/component-compare-changelog.section.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CompareChangelogSection = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _react() {
  const data = _interopRequireDefault(require("react"));
  _react = function () {
    return data;
  };
  return data;
}
function _componentUi() {
  const data = require("@teambit/component.ui.compare");
  _componentUi = function () {
    return data;
  };
  return data;
}
class CompareChangelogSection {
  constructor() {
    (0, _defineProperty2().default)(this, "navigationLink", {
      href: 'changelog',
      children: 'Changelog',
      order: 5
    });
    (0, _defineProperty2().default)(this, "route", {
      path: 'changelog/*',
      element: /*#__PURE__*/_react().default.createElement(_componentUi().ComponentCompareChangelog, null)
    });
  }
}
exports.CompareChangelogSection = CompareChangelogSection;
