// @teambit/component-compare/dist/component-compare.ui.runtime.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.array.iterator.js");
require("core-js/modules/es.promise.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ComponentCompareUI = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _component() {
  const data = _interopRequireDefault(require("@teambit/component"));
  _component = function () {
    return data;
  };
  return data;
}
function _componentUi() {
  const data = require("@teambit/component.ui.compare");
  _componentUi = function () {
    return data;
  };
  return data;
}
function _harmony() {
  const data = require("@teambit/harmony");
  _harmony = function () {
    return data;
  };
  return data;
}
function _ui() {
  const data = require("@teambit/ui");
  _ui = function () {
    return data;
  };
  return data;
}
function _react() {
  const data = _interopRequireDefault(require("react"));
  _react = function () {
    return data;
  };
  return data;
}
function _componentCompareAspects() {
  const data = require("./component-compare-aspects.section");
  _componentCompareAspects = function () {
    return data;
  };
  return data;
}
function _componentCompare() {
  const data = require("./component-compare.aspect");
  _componentCompare = function () {
    return data;
  };
  return data;
}
function _componentCompare2() {
  const data = require("./component-compare.section");
  _componentCompare2 = function () {
    return data;
  };
  return data;
}
function _componentCompareChangelog() {
  const data = require("./component-compare-changelog.section");
  _componentCompareChangelog = function () {
    return data;
  };
  return data;
}
class ComponentCompareUI {
  constructor(host, navSlot, routeSlot) {
    this.host = host;
    this.navSlot = navSlot;
    this.routeSlot = routeSlot;
    (0, _defineProperty2().default)(this, "getComponentComparePage", () => /*#__PURE__*/_react().default.createElement(_componentUi().ComponentCompare, {
      navSlot: this.navSlot,
      routeSlot: this.routeSlot,
      host: this.host
    }));
  }
  registerNavigation(route) {
    if (Array.isArray(route)) {
      this.navSlot.register(route);
    } else {
      this.navSlot.register([route]);
    }
    return this;
  }
  registerRoutes(routes) {
    this.routeSlot.register(routes);
    return this;
  }
  static async provider([componentUi], _, [navSlot, routeSlot], harmony) {
    const {
      config
    } = harmony;
    const host = String(config.get('teambit.harmony/bit'));
    const componentCompareUI = new ComponentCompareUI(host, navSlot, routeSlot);
    const componentCompareSection = new (_componentCompare2().ComponentCompareSection)(componentCompareUI);
    componentUi.registerRoute([componentCompareSection.route]);
    componentUi.registerWidget(componentCompareSection.navigationLink, componentCompareSection.order);
    const aspectCompareSection = new (_componentCompareAspects().AspectsCompareSection)(host);
    const compareChangelog = new (_componentCompareChangelog().CompareChangelogSection)();
    componentCompareUI.registerNavigation([{
      props: aspectCompareSection.navigationLink,
      order: aspectCompareSection.navigationLink.order
    }, {
      props: compareChangelog.navigationLink,
      order: compareChangelog.navigationLink.order
    }]);
    componentCompareUI.registerRoutes([aspectCompareSection.route, compareChangelog.route]);
    return componentCompareUI;
  }
}
exports.ComponentCompareUI = ComponentCompareUI;
(0, _defineProperty2().default)(ComponentCompareUI, "runtime", _ui().UIRuntime);
(0, _defineProperty2().default)(ComponentCompareUI, "slots", [_harmony().Slot.withType(), _harmony().Slot.withType()]);
(0, _defineProperty2().default)(ComponentCompareUI, "dependencies", [_component().default]);
_componentCompare().ComponentCompareAspect.addRuntime(ComponentCompareUI);
