// @teambit/component-compare/dist/component-compare-aspects.section.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AspectsCompareSection = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _react() {
  const data = _interopRequireDefault(require("react"));
  _react = function () {
    return data;
  };
  return data;
}
function _componentUi() {
  const data = require("@teambit/component.ui.compare");
  _componentUi = function () {
    return data;
  };
  return data;
}
class AspectsCompareSection {
  constructor(host) {
    this.host = host;
    (0, _defineProperty2().default)(this, "navigationLink", {
      href: 'aspects',
      children: 'Aspects',
      order: 6
    });
    (0, _defineProperty2().default)(this, "route", {
      path: 'aspects/*',
      element: /*#__PURE__*/_react().default.createElement(_componentUi().ComponentCompareAspects, {
        host: this.host
      })
    });
  }
}
exports.AspectsCompareSection = AspectsCompareSection;
