// @teambit/node/dist/templates/node/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.nodeTemplate = void 0;
function _index() {
  const data = require("./files/index");
  _index = function () {
    return data;
  };
  return data;
}
function _doc() {
  const data = require("./files/doc");
  _doc = function () {
    return data;
  };
  return data;
}
function _main() {
  const data = require("./files/main");
  _main = function () {
    return data;
  };
  return data;
}
function _test() {
  const data = require("./files/test");
  _test = function () {
    return data;
  };
  return data;
}
function _composition() {
  const data = require("./files/composition");
  _composition = function () {
    return data;
  };
  return data;
}
const nodeTemplate = {
  name: 'node',
  description: 'a Node.js module',
  generateFiles: context => {
    return [(0, _index().indexFile)(context), (0, _doc().docFile)(context), (0, _main().mainFile)(context), (0, _composition().compositionFile)(context), (0, _test().testFile)(context)];
  },
  config: {
    'teambit.harmony/node': {},
    'teambit.envs/envs': {
      env: 'teambit.harmony/node'
    }
  }
};
exports.nodeTemplate = nodeTemplate;
