// @teambit/envs/dist/context/context.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.array.iterator.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ExecutionContext = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _envDefinition() {
  const data = require("../env-definition");
  _envDefinition = function () {
    return data;
  };
  return data;
}
class ExecutionContext {
  constructor(
  /**
   * upper scope of all environment contexts.
   */
  upper,
  /**
   * runtime instance of the environment.
   */
  envRuntime,
  /**
   * components applied in the execution context.
   */
  components = envRuntime.components) {
    this.upper = upper;
    this.envRuntime = envRuntime;
    this.components = components;
    (0, _defineProperty2().default)(this, "relatedContexts", []);
  }
  /**
   * extension ID of the environment
   */
  get id() {
    return this.envRuntime.id;
  }

  /**
   * environment instance.
   */
  get env() {
    return this.envRuntime.env;
  }
  get envDefinition() {
    return new (_envDefinition().EnvDefinition)(this.id, this.env);
  }
  apply(name, args) {
    if (!this.env[name]) {
      throw new Error(`method ${name} not implemented`);
    }
    return this.env[name].apply(this.env, ...args);
  }
}
exports.ExecutionContext = ExecutionContext;
