// @teambit/isolator/dist/capsule/container-exec.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.promise.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _capsule() {
  const data = require("@teambit/capsule");
  _capsule = function () {
    return data;
  };
  return data;
}
function _events() {
  const data = require("events");
  _events = function () {
    return data;
  };
  return data;
}
class ContainerExec extends _events().EventEmitter {
  constructor(_code = 0) {
    super();
    this._code = _code;
    (0, _defineProperty2().default)(this, "stdout", new (_capsule().DuplexBufferStream)());
    (0, _defineProperty2().default)(this, "stderr", new (_capsule().DuplexBufferStream)());
    (0, _defineProperty2().default)(this, "stdin", new (_capsule().DuplexBufferStream)());
  }
  setStatus(status) {
    this._code = status;
    this.emit('close');
  }
  get code() {
    return this._code;
  }
  inspect() {
    return Promise.resolve({
      running: true,
      pid: 1
    });
  }
  abort() {
    throw new Error('Method not implemented.');
  }
}
exports.default = ContainerExec;
