// @teambit/dependency-resolver/dist/policy/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "EnvPolicy", {
  enumerable: true,
  get: function () {
    return _envPolicy().EnvPolicy;
  }
});
Object.defineProperty(exports, "EnvPolicyConfigObject", {
  enumerable: true,
  get: function () {
    return _envPolicy().EnvPolicyConfigObject;
  }
});
Object.defineProperty(exports, "PeersAutoDetectPolicy", {
  enumerable: true,
  get: function () {
    return _envPolicy().PeersAutoDetectPolicy;
  }
});
Object.defineProperty(exports, "Policy", {
  enumerable: true,
  get: function () {
    return _policy().Policy;
  }
});
Object.defineProperty(exports, "PolicyConfigKeys", {
  enumerable: true,
  get: function () {
    return _policy().PolicyConfigKeys;
  }
});
Object.defineProperty(exports, "PolicyConfigKeysNames", {
  enumerable: true,
  get: function () {
    return _policy().PolicyConfigKeysNames;
  }
});
Object.defineProperty(exports, "PolicyEntry", {
  enumerable: true,
  get: function () {
    return _policy().PolicyEntry;
  }
});
Object.defineProperty(exports, "PolicySemver", {
  enumerable: true,
  get: function () {
    return _policy().PolicySemver;
  }
});
Object.defineProperty(exports, "SerializedVariantPolicy", {
  enumerable: true,
  get: function () {
    return _variantPolicy().SerializedVariantPolicy;
  }
});
Object.defineProperty(exports, "VariantPolicy", {
  enumerable: true,
  get: function () {
    return _variantPolicy().VariantPolicy;
  }
});
Object.defineProperty(exports, "VariantPolicyConfigObject", {
  enumerable: true,
  get: function () {
    return _variantPolicy().VariantPolicyConfigObject;
  }
});
Object.defineProperty(exports, "VariantPolicyFactory", {
  enumerable: true,
  get: function () {
    return _variantPolicy().VariantPolicyFactory;
  }
});
Object.defineProperty(exports, "WorkspacePolicy", {
  enumerable: true,
  get: function () {
    return _workspacePolicy().WorkspacePolicy;
  }
});
Object.defineProperty(exports, "WorkspacePolicyAddEntryOptions", {
  enumerable: true,
  get: function () {
    return _workspacePolicy().AddEntryOptions;
  }
});
Object.defineProperty(exports, "WorkspacePolicyConfigObject", {
  enumerable: true,
  get: function () {
    return _workspacePolicy().WorkspacePolicyConfigObject;
  }
});
Object.defineProperty(exports, "WorkspacePolicyEntry", {
  enumerable: true,
  get: function () {
    return _workspacePolicy().WorkspacePolicyEntry;
  }
});
Object.defineProperty(exports, "WorkspacePolicyFactory", {
  enumerable: true,
  get: function () {
    return _workspacePolicy().WorkspacePolicyFactory;
  }
});
function _policy() {
  const data = require("./policy");
  _policy = function () {
    return data;
  };
  return data;
}
function _workspacePolicy() {
  const data = require("./workspace-policy");
  _workspacePolicy = function () {
    return data;
  };
  return data;
}
function _variantPolicy() {
  const data = require("./variant-policy");
  _variantPolicy = function () {
    return data;
  };
  return data;
}
function _envPolicy() {
  const data = require("./env-policy");
  _envPolicy = function () {
    return data;
  };
  return data;
}
