// @teambit/builder/dist/artifact/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Artifact", {
  enumerable: true,
  get: function () {
    return _artifact().Artifact;
  }
});
Object.defineProperty(exports, "ArtifactDefinition", {
  enumerable: true,
  get: function () {
    return _artifactDefinition().ArtifactDefinition;
  }
});
Object.defineProperty(exports, "ArtifactFactory", {
  enumerable: true,
  get: function () {
    return _artifactFactory().ArtifactFactory;
  }
});
Object.defineProperty(exports, "ArtifactList", {
  enumerable: true,
  get: function () {
    return _artifactList().ArtifactList;
  }
});
Object.defineProperty(exports, "ArtifactMap", {
  enumerable: true,
  get: function () {
    return _artifactFactory().ArtifactMap;
  }
});
Object.defineProperty(exports, "ArtifactModelDefinition", {
  enumerable: true,
  get: function () {
    return _artifactDefinition().ArtifactModelDefinition;
  }
});
Object.defineProperty(exports, "ExtensionArtifact", {
  enumerable: true,
  get: function () {
    return _extensionArtifact().ExtensionArtifact;
  }
});
Object.defineProperty(exports, "FsArtifact", {
  enumerable: true,
  get: function () {
    return _fsArtifact().FsArtifact;
  }
});
function _artifactDefinition() {
  const data = require("./artifact-definition");
  _artifactDefinition = function () {
    return data;
  };
  return data;
}
function _extensionArtifact() {
  const data = require("./extension-artifact");
  _extensionArtifact = function () {
    return data;
  };
  return data;
}
function _artifact() {
  const data = require("./artifact");
  _artifact = function () {
    return data;
  };
  return data;
}
function _artifactList() {
  const data = require("./artifact-list");
  _artifactList = function () {
    return data;
  };
  return data;
}
function _artifactFactory() {
  const data = require("./artifact-factory");
  _artifactFactory = function () {
    return data;
  };
  return data;
}
function _fsArtifact() {
  const data = require("./fs-artifact");
  _fsArtifact = function () {
    return data;
  };
  return data;
}
