// @teambit/component/dist/snap/snap.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Snap = void 0;
/**
 * `Snap` represents a sealed state of the component in the working tree.
 */
class Snap {
  constructor(
  /**
   * hash of the snap.
   */
  hash,
  /**
   * date time of the snap.
   */
  timestamp,
  /**
   * parent snap
   */
  parents,
  /**
   * author of the component `Snap`.
   */
  author,
  /**
   * message added by the `Snap` author.
   */
  message) {
    this.hash = hash;
    this.timestamp = timestamp;
    this.parents = parents;
    this.author = author;
    this.message = message;
  }
  static fromObject(snapObject) {
    const parents = snapObject.parents || [];
    return new Snap(snapObject.hash, new Date(parseInt(snapObject.timestamp)), parents, snapObject.author, snapObject.message);
  }
  toObject() {
    return {
      timestamp: this.timestamp.getTime().toString(),
      hash: this.hash,
      author: this.author,
      message: this.message,
      parents: this.parents
    };
  }
}
exports.Snap = Snap;
