// @teambit/react/dist/templates/react-hook/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.reactHook = void 0;
function _component() {
  const data = require("./files/component");
  _component = function () {
    return data;
  };
  return data;
}
function _composition() {
  const data = require("./files/composition");
  _composition = function () {
    return data;
  };
  return data;
}
function _docs() {
  const data = require("./files/docs");
  _docs = function () {
    return data;
  };
  return data;
}
function _test() {
  const data = require("./files/test");
  _test = function () {
    return data;
  };
  return data;
}
function _indexFile() {
  const data = require("./files/index-file");
  _indexFile = function () {
    return data;
  };
  return data;
}
const reactHook = {
  name: 'react-hook',
  description: 'a react hook component',
  generateFiles: context => {
    return [(0, _indexFile().indexFile)(context), (0, _component().componentFile)(context), (0, _composition().compositionFile)(context), (0, _docs().docsFile)(context), (0, _test().testFile)(context)];
  },
  config: {
    'teambit.react/react': {},
    'teambit.envs/envs': {
      env: 'teambit.react/react'
    }
  }
};
exports.reactHook = reactHook;
