// @teambit/react/dist/templates/react-workspace-analytics/index.js
"use strict";

require("core-js/modules/es.promise.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.reactWorkspaceAnalyticsTemplate = void 0;
function _workspaceCommon() {
  const data = require("../workspace-common");
  _workspaceCommon = function () {
    return data;
  };
  return data;
}
const reactWorkspaceAnalyticsTemplate = {
  name: 'react-analytics',
  description: 'React workspace with components for an Analytics Application',
  generateFiles: async context => {
    return (0, _workspaceCommon().generateFiles)(context);
  },
  fork: () => {
    return [{
      id: 'teambit.analytics/examples/dashboard',
      targetName: 'examples/dashboard',
      path: 'examples/dashboard'
    }, {
      id: 'teambit.analytics/examples/new-users',
      targetName: 'examples/new-users',
      path: 'examples/new-users'
    }, {
      id: 'teambit.analytics/examples/revenues',
      targetName: 'examples/revenues',
      path: 'examples/revenues'
    }, {
      id: 'teambit.analytics/examples/top-frameworks',
      targetName: 'examples/top-frameworks',
      path: 'examples/top-frameworks'
    }, {
      id: 'teambit.analytics/examples/top-members',
      targetName: 'examples/top-members',
      path: 'examples/top-members'
    }];
  }
};
exports.reactWorkspaceAnalyticsTemplate = reactWorkspaceAnalyticsTemplate;
