// @teambit/react/dist/templates/react-workspace-wiki/index.js
"use strict";

require("core-js/modules/es.promise.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.reactWorkspaceWikiTemplate = void 0;
function _workspaceCommon() {
  const data = require("../workspace-common");
  _workspaceCommon = function () {
    return data;
  };
  return data;
}
const reactWorkspaceWikiTemplate = {
  name: 'react-wiki',
  description: 'React workspace with components for a Wiki',
  generateFiles: async context => {
    return (0, _workspaceCommon().generateFiles)(context);
  },
  fork: () => {
    return [{
      id: 'teambit.wiki/apps/wiki',
      targetName: 'apps/wiki',
      path: 'apps/wiki'
    }, {
      id: 'teambit.wiki/sections/wiki',
      targetName: 'sections/wiki',
      path: 'sections/wiki'
    }, {
      id: 'teambit.wiki/blocks/header',
      targetName: 'blocks/header',
      path: 'blocks/header'
    }, {
      id: 'teambit.wiki/blocks/footer',
      targetName: 'blocks/footer',
      path: 'blocks/footer'
    }, {
      id: 'teambit.wiki/content/welcome',
      targetName: 'content/welcome',
      path: 'content/welcome'
    }];
  }
};
exports.reactWorkspaceWikiTemplate = reactWorkspaceWikiTemplate;
