// @teambit/react/dist/apps/web/webpack/webpack.app.ssr.config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SSR_ENTRY_FILE = exports.PUBLIC_PATH = void 0;
exports.buildConfig = buildConfig;
exports.calcOutputPath = calcOutputPath;
exports.clientConfig = clientConfig;
exports.ssrBuildConfig = ssrBuildConfig;
exports.ssrConfig = ssrConfig;
function _path() {
  const data = require("path");
  _path = function () {
    return data;
  };
  return data;
}
const OUTPUT_FOLDER = 'public/bit-apps';
const SSR_ENTRY_FILE = 'index.js';
/** _external_ base url for static files. (is not a folder)
 * @example
 * "http://localhost:3000/public/main.88a88d47d86aaabc0230.js"
 */
exports.SSR_ENTRY_FILE = SSR_ENTRY_FILE;
const PUBLIC_PATH = '/public';
exports.PUBLIC_PATH = PUBLIC_PATH;
var Targets;
(function (Targets) {
  Targets["ssr"] = "ssr";
  Targets["browser"] = "browser";
})(Targets || (Targets = {}));
function calcOutputPath(appName, target) {
  return (0, _path().join)(OUTPUT_FOLDER, appName, target);
}
function clientConfig() {
  return {
    output: {
      publicPath: PUBLIC_PATH,
      filename: '[name].[chunkhash].js'
    }
  };
}
function ssrConfig() {
  return {
    target: 'node',
    devtool: 'eval-cheap-source-map',
    output: {
      libraryTarget: 'commonjs',
      filename: SSR_ENTRY_FILE
    }
  };
}
function buildConfig({
  outputPath
}) {
  return {
    output: {
      path: outputPath,
      publicPath: `/`,
      filename: '[name].[chunkhash].js'
    }
  };
}
function ssrBuildConfig({
  outputPath
}) {
  return {
    target: 'node',
    devtool: 'eval-cheap-source-map',
    output: {
      path: outputPath,
      publicPath: `/`,
      libraryTarget: 'commonjs',
      filename: 'index.js'
    }
  };
}
