// @teambit/jest/dist/jest.main.runtime.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.array.iterator.js");
require("core-js/modules/es.promise.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WORKER_NAME = exports.JestMain = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _cli() {
  const data = require("@teambit/cli");
  _cli = function () {
    return data;
  };
  return data;
}
function _logger() {
  const data = require("@teambit/logger");
  _logger = function () {
    return data;
  };
  return data;
}
function _worker() {
  const data = require("@teambit/worker");
  _worker = function () {
    return data;
  };
  return data;
}
function _jest() {
  const data = require("./jest.aspect");
  _jest = function () {
    return data;
  };
  return data;
}
function _jest2() {
  const data = require("./jest.tester");
  _jest2 = function () {
    return data;
  };
  return data;
}
const WORKER_NAME = 'jest';
exports.WORKER_NAME = WORKER_NAME;
class JestMain {
  constructor(jestWorker, logger) {
    this.jestWorker = jestWorker;
    this.logger = logger;
  }
  createTester(jestConfig, jestModulePath = require.resolve('jest'), opts) {
    return new (_jest2().JestTester)(_jest().JestAspect.id, jestConfig, jestModulePath, this.jestWorker, this.logger, opts);
  }
  static async provider([worker, loggerAspect]) {
    const logger = loggerAspect.createLogger(_jest().JestAspect.id);
    const jestWorker = await worker.declareWorker(WORKER_NAME, require.resolve('./jest.worker'));
    return new JestMain(jestWorker, logger);
  }
}
exports.JestMain = JestMain;
(0, _defineProperty2().default)(JestMain, "runtime", _cli().MainRuntime);
(0, _defineProperty2().default)(JestMain, "dependencies", [_worker().WorkerAspect, _logger().LoggerAspect]);
_jest().JestAspect.addRuntime(JestMain);
