// @teambit/html/dist/html.env.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.promise.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HtmlEnvType = exports.HtmlEnv = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
const HtmlEnvType = 'html';
exports.HtmlEnvType = HtmlEnvType;
class HtmlEnv {
  constructor() {
    (0, _defineProperty2().default)(this, "icon", 'https://static.bit.dev/file_type_html5.svg');
  }
  getDocsTemplate() {
    return require.resolve('./html-docs-app');
  }
  getMounter() {
    return require.resolve('./mount');
  }
  getDependencies() {
    return {
      devDependencies: {
        '@types/jest': '26.0.20'
      }
    };
  }
  getPreviewConfig() {
    return {
      strategyName: 'component',
      splitComponentBundle: true
    };
  }
  async __getDescriptor() {
    return {
      type: HtmlEnvType
    };
  }
}
exports.HtmlEnv = HtmlEnv;
