// @teambit/component-sizer/dist/component-sizer.ui.runtime.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.array.iterator.js");
require("core-js/modules/es.promise.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SizerUIRuntime = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _react() {
  const data = _interopRequireDefault(require("react"));
  _react = function () {
    return data;
  };
  return data;
}
function _ui() {
  const data = require("@teambit/ui");
  _ui = function () {
    return data;
  };
  return data;
}
function _docs() {
  const data = require("@teambit/docs");
  _docs = function () {
    return data;
  };
  return data;
}
function _componentUi() {
  const data = require("@teambit/component.ui.component-size");
  _componentUi = function () {
    return data;
  };
  return data;
}
function _componentSizer() {
  const data = require("./component-sizer.aspect");
  _componentSizer = function () {
    return data;
  };
  return data;
}
/**
 * Component size aspect.
 */
class SizerUIRuntime {
  static async provider([docs]) {
    docs.registerTitleBadge({
      component: function badge({
        legacyComponentModel
      }) {
        return /*#__PURE__*/_react().default.createElement(_componentUi().ComponentSize, {
          legacyComponentModel: legacyComponentModel
        });
      },
      weight: 30
    });
  }
}
exports.SizerUIRuntime = SizerUIRuntime;
(0, _defineProperty2().default)(SizerUIRuntime, "dependencies", [_docs().DocsAspect]);
(0, _defineProperty2().default)(SizerUIRuntime, "runtime", _ui().UIRuntime);
_componentSizer().ComponentSizerAspect.addRuntime(SizerUIRuntime);
