// @teambit/component-sizer/dist/component-sizer.graphql.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.componentSizerSchema = componentSizerSchema;
function _graphqlTag() {
  const data = _interopRequireDefault(require("graphql-tag"));
  _graphqlTag = function () {
    return data;
  };
  return data;
}
function componentSizerSchema(componentSizerMain) {
  return {
    typeDefs: (0, _graphqlTag().default)`
      type ComponentSizedFile {
        name: String
        size: Int
        compressedSize: Int
      }

      type ComponentSize {
        # breakdown for the component files sizes
        files: [ComponentSizedFile]
        # breakdown for the component assets sizes
        assets: [ComponentSizedFile]
        # total size of the component files (like js and css) (without assets)
        totalFiles: Int
        # total size of the component assets (like pngs) (without files)
        totalAssets: Int
        # total size of the component files and assets
        total: Int

        compressedTotalFiles: Int
        compressedTotalAssets: Int
        compressedTotal: Int
      }

      extend type Component {
        # size of the component bundle
        size: ComponentSize
      }
    `,
    resolvers: {
      Component: {
        size: component => {
          return componentSizerMain.getComponentSize(component);
        }
      }
    }
  };
}
