// @teambit/tester/dist/tests.section.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TestsSection = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _react() {
  const data = _interopRequireDefault(require("react"));
  _react = function () {
    return data;
  };
  return data;
}
function _defenderUi() {
  const data = require("@teambit/defender.ui.test-page");
  _defenderUi = function () {
    return data;
  };
  return data;
}
class TestsSection {
  constructor(emptyStateSlot) {
    this.emptyStateSlot = emptyStateSlot;
    (0, _defineProperty2().default)(this, "route", {
      path: '~tests',
      element: /*#__PURE__*/_react().default.createElement(_defenderUi().TestsPage, {
        emptyState: this.emptyStateSlot
      })
    });
    (0, _defineProperty2().default)(this, "navigationLink", {
      href: '~tests',
      children: 'Tests'
    });
    (0, _defineProperty2().default)(this, "order", 40);
  }
}
exports.TestsSection = TestsSection;
