// @teambit/workspace/dist/ui/workspace/workspace-model.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Workspace = void 0;
function _component() {
  const data = require("@teambit/component");
  _component = function () {
    return data;
  };
  return data;
}
function _componentDescriptor() {
  const data = require("@teambit/component-descriptor");
  _componentDescriptor = function () {
    return data;
  };
  return data;
}
// import { ComponentMeta } from '@teambit/component';

class Workspace {
  constructor(
  /**
   * name of the workspace.
   */
  name,
  /**
   * absolute path of the workspace.
   */
  path,
  /**
   * icon of the workspace.
   */
  icon,
  /**
   * components container in the workspace.
   */
  components,
  /**
   * components contained in the workspace.
   */
  componentDescriptors) {
    this.name = name;
    this.path = path;
    this.icon = icon;
    this.components = components;
    this.componentDescriptors = componentDescriptors;
  }

  /**
   * return a component from the workspace.
   */
  getComponent(id) {
    return this.components.find(component => component.id.fullName === id.fullName);
  }
  static from({
    name,
    path,
    components,
    icon
  }) {
    const componentDescriptors = components.map(component => {
      const id = _component().ComponentID.fromObject(component.id);
      return _componentDescriptor().ComponentDescriptor.fromObject({
        id: id.toString()
      });
    });
    return new Workspace(name, path, icon, components.map(value => {
      return _component().ComponentModel.from(value);
    }), componentDescriptors);
  }
  static empty() {
    return new Workspace('', '', '', [], []);
  }
}
exports.Workspace = Workspace;
