// @teambit/harmony/dist/extension/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createHook = exports.provider = exports.HookRegistry = exports.hook = exports.register = exports.ExtensionDecorator = exports.Extension = void 0;
var extension_1 = require("./extension");
Object.defineProperty(exports, "Extension", { enumerable: true, get: function () { return extension_1.Extension; } });
var decorator_1 = require("./decorator");
Object.defineProperty(exports, "ExtensionDecorator", { enumerable: true, get: function () { return decorator_1.ExtensionDecorator; } });
Object.defineProperty(exports, "register", { enumerable: true, get: function () { return decorator_1.register; } });
Object.defineProperty(exports, "hook", { enumerable: true, get: function () { return decorator_1.hook; } });
Object.defineProperty(exports, "HookRegistry", { enumerable: true, get: function () { return decorator_1.HookRegistry; } });
Object.defineProperty(exports, "provider", { enumerable: true, get: function () { return decorator_1.provider; } });
Object.defineProperty(exports, "createHook", { enumerable: true, get: function () { return decorator_1.createHook; } });
