// @teambit/react/dist/templates/react-component-js/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.reactComponentJS = exports.deprecatedReactComponentJS = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _component() {
  const data = require("./files/component");
  _component = function () {
    return data;
  };
  return data;
}
function _composition() {
  const data = require("./files/composition");
  _composition = function () {
    return data;
  };
  return data;
}
function _docs() {
  const data = require("./files/docs");
  _docs = function () {
    return data;
  };
  return data;
}
function _test() {
  const data = require("./files/test");
  _test = function () {
    return data;
  };
  return data;
}
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2().default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
const reactComponentJS = {
  name: 'react-js',
  description: 'a basic react component in js',
  generateFiles: context => {
    const {
      name,
      namePascalCase: Name
    } = context;
    const indexFile = {
      relativePath: 'index.js',
      content: `export { ${Name} } from './${name}';
`
    };
    return [indexFile, (0, _component().componentFile)(context), (0, _composition().compositionFile)(context), (0, _docs().docsFile)(context), (0, _test().testFile)(context)];
  },
  config: {
    'teambit.react/react': {},
    'teambit.envs/envs': {
      env: 'teambit.react/react'
    }
  }
};
exports.reactComponentJS = reactComponentJS;
const deprecatedReactComponentJS = _objectSpread(_objectSpread({}, reactComponentJS), {}, {
  name: 'react-component-js',
  hidden: true
});
exports.deprecatedReactComponentJS = deprecatedReactComponentJS;
