// @teambit/graph/dist/ui/query/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "EdgeModel", {
  enumerable: true,
  get: function () {
    return _edgeModel().EdgeModel;
  }
});
Object.defineProperty(exports, "GraphModel", {
  enumerable: true,
  get: function () {
    return _graphModel().GraphModel;
  }
});
Object.defineProperty(exports, "NodeModel", {
  enumerable: true,
  get: function () {
    return _nodeModel().NodeModel;
  }
});
Object.defineProperty(exports, "RawGraph", {
  enumerable: true,
  get: function () {
    return _getGraph().RawGraph;
  }
});
Object.defineProperty(exports, "useGraph", {
  enumerable: true,
  get: function () {
    return _useGraph().useGraph;
  }
});
Object.defineProperty(exports, "useGraphQuery", {
  enumerable: true,
  get: function () {
    return _useGraphQuery().useGraphQuery;
  }
});
function _useGraphQuery() {
  const data = require("./use-graph-query");
  _useGraphQuery = function () {
    return data;
  };
  return data;
}
function _useGraph() {
  const data = require("./use-graph");
  _useGraph = function () {
    return data;
  };
  return data;
}
function _graphModel() {
  const data = require("./graph-model");
  _graphModel = function () {
    return data;
  };
  return data;
}
function _edgeModel() {
  const data = require("./edge-model");
  _edgeModel = function () {
    return data;
  };
  return data;
}
function _nodeModel() {
  const data = require("./node-model");
  _nodeModel = function () {
    return data;
  };
  return data;
}
function _getGraph() {
  const data = require("./get-graph.query");
  _getGraph = function () {
    return data;
  };
  return data;
}
