// @teambit/scope/dist/scope.ui-root.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.promise.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ScopeUIRoot = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
class ScopeUIRoot {
  constructor(
  /**
   * scope extension.
   */
  scope) {
    this.scope = scope;
    (0, _defineProperty2().default)(this, "buildOptions", {
      ssr: true
    });
  }
  get name() {
    return this.scope.name;
  }
  get path() {
    return this.scope.path;
  }
  get configFile() {
    return 'scope.jsonc';
  }
  get devServers() {
    return Promise.resolve([]);
  }
  resolveAspects(runtime, componentIds, opts) {
    return this.scope.resolveAspects(runtime, componentIds, opts);
  }
  async resolvePattern() {
    return [];
  }
}
exports.ScopeUIRoot = ScopeUIRoot;
