// @teambit/isolator/dist/capsule-list.js
"use strict";

require("core-js/modules/es.array.iterator.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
function _path() {
  const data = require("path");
  _path = function () {
    return data;
  };
  return data;
}
class CapsuleList extends Array {
  getCapsule(id) {
    return this.find(capsule => capsule.component.id.isEqual(id));
  }
  getCapsuleByLegacyId(id) {
    return this.find(capsule => capsule.component.id._legacy.isEqual(id));
  }
  getCapsuleIgnoreVersion(id) {
    return this.find(capsule => capsule.component.id.isEqual(id, {
      ignoreVersion: true
    }));
  }
  getCapsuleIgnoreScopeAndVersion(id) {
    return this.find(capsule => capsule.component.id._legacy.isEqualWithoutScopeAndVersion(id._legacy));
  }
  getAllCapsuleDirs() {
    return this.map(capsule => capsule.path);
  }
  getIdByPathInCapsule(pathInCapsule) {
    const normalizedPathInCapsule = (0, _path().normalize)(pathInCapsule);
    const found = this.find(capsule => normalizedPathInCapsule === (0, _path().normalize)(capsule.path));
    return found ? found.component.id : null;
  }
  getAllComponents() {
    return this.map(c => c.component);
  }
  static fromArray(capsules) {
    return new CapsuleList(...capsules);
  }
}
exports.default = CapsuleList;
