// @teambit/isolator/dist/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Capsule", {
  enumerable: true,
  get: function () {
    return _capsule().Capsule;
  }
});
Object.defineProperty(exports, "CapsuleList", {
  enumerable: true,
  get: function () {
    return _capsuleList().default;
  }
});
Object.defineProperty(exports, "ContainerExec", {
  enumerable: true,
  get: function () {
    return _capsule().ContainerExec;
  }
});
Object.defineProperty(exports, "FsContainer", {
  enumerable: true,
  get: function () {
    return _capsule().FsContainer;
  }
});
Object.defineProperty(exports, "IsolateComponentsOptions", {
  enumerable: true,
  get: function () {
    return _isolatorMain().IsolateComponentsOptions;
  }
});
Object.defineProperty(exports, "IsolatorAspect", {
  enumerable: true,
  get: function () {
    return _isolator().IsolatorAspect;
  }
});
Object.defineProperty(exports, "Network", {
  enumerable: true,
  get: function () {
    return _network().Network;
  }
});
function _isolatorMain() {
  const data = require("./isolator.main.runtime");
  _isolatorMain = function () {
    return data;
  };
  return data;
}
function _network() {
  const data = require("./network");
  _network = function () {
    return data;
  };
  return data;
}
function _capsule() {
  const data = require("./capsule");
  _capsule = function () {
    return data;
  };
  return data;
}
function _isolator() {
  const data = require("./isolator.aspect");
  _isolator = function () {
    return data;
  };
  return data;
}
function _capsuleList() {
  const data = _interopRequireDefault(require("./capsule-list"));
  _capsuleList = function () {
    return data;
  };
  return data;
}
