// @teambit/node/dist/templates/express-route/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.expressRouteTemplate = void 0;
function _index() {
  const data = require("./files/index");
  _index = function () {
    return data;
  };
  return data;
}
function _doc() {
  const data = require("./files/doc");
  _doc = function () {
    return data;
  };
  return data;
}
function _main() {
  const data = require("./files/main");
  _main = function () {
    return data;
  };
  return data;
}
// import { testFile } from './files/test';
// import { compositionFile } from './files/composition';

const expressRouteTemplate = {
  name: 'express-route',
  description: 'an express route',
  generateFiles: context => {
    return [(0, _index().indexFile)(context), (0, _doc().docFile)(context), (0, _main().mainFile)(context)
    // compositionFile(context),
    // testFile(context)
    ];
  },

  config: {
    'teambit.harmony/node': {},
    'teambit.envs/envs': {
      env: 'teambit.harmony/node'
    }
  }
};
exports.expressRouteTemplate = expressRouteTemplate;
