// @teambit/legacy/dist/consumer/component/dependencies/dependency-resolver/dependencies-resolver.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.array.iterator.js");
require("core-js/modules/es.promise.js");
require("core-js/modules/es.regexp.exec.js");
require("core-js/modules/es.string.replace.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
function _forEachObjIndexed2() {
  const data = _interopRequireDefault(require("ramda/src/forEachObjIndexed"));
  _forEachObjIndexed2 = function () {
    return data;
  };
  return data;
}
function _differenceWith2() {
  const data = _interopRequireDefault(require("ramda/src/differenceWith"));
  _differenceWith2 = function () {
    return data;
  };
  return data;
}
function _pick2() {
  const data = _interopRequireDefault(require("ramda/src/pick"));
  _pick2 = function () {
    return data;
  };
  return data;
}
function _keys2() {
  const data = _interopRequireDefault(require("ramda/src/keys"));
  _keys2 = function () {
    return data;
  };
  return data;
}
function _difference2() {
  const data = _interopRequireDefault(require("ramda/src/difference"));
  _difference2 = function () {
    return data;
  };
  return data;
}
function _flatten2() {
  const data = _interopRequireDefault(require("ramda/src/flatten"));
  _flatten2 = function () {
    return data;
  };
  return data;
}
function _isEmpty2() {
  const data = _interopRequireDefault(require("ramda/src/isEmpty"));
  _isEmpty2 = function () {
    return data;
  };
  return data;
}
function _pickBy2() {
  const data = _interopRequireDefault(require("ramda/src/pickBy"));
  _pickBy2 = function () {
    return data;
  };
  return data;
}
function _uniq2() {
  const data = _interopRequireDefault(require("ramda/src/uniq"));
  _uniq2 = function () {
    return data;
  };
  return data;
}
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function path() {
  const data = _interopRequireWildcard(require("path"));
  path = function () {
    return data;
  };
  return data;
}
function _fsExtra() {
  const data = _interopRequireDefault(require("fs-extra"));
  _fsExtra = function () {
    return data;
  };
  return data;
}
function _semver() {
  const data = _interopRequireDefault(require("semver"));
  _semver = function () {
    return data;
  };
  return data;
}
function _lodash() {
  const data = require("lodash");
  _lodash = function () {
    return data;
  };
  return data;
}
function _componentIssues() {
  const data = require("@teambit/component-issues");
  _componentIssues = function () {
    return data;
  };
  return data;
}
function _bitId() {
  const data = require("../../../../bit-id");
  _bitId = function () {
    return data;
  };
  return data;
}
function _constants() {
  const data = require("../../../../constants");
  _constants = function () {
    return data;
  };
  return data;
}
function _generalError() {
  const data = _interopRequireDefault(require("../../../../error/general-error"));
  _generalError = function () {
    return data;
  };
  return data;
}
function _showDoctorError() {
  const data = _interopRequireDefault(require("../../../../error/show-doctor-error"));
  _showDoctorError = function () {
    return data;
  };
  return data;
}
function _logger() {
  const data = _interopRequireDefault(require("../../../../logger/logger"));
  _logger = function () {
    return data;
  };
  return data;
}
function _utils() {
  const data = require("../../../../utils");
  _utils = function () {
    return data;
  };
  return data;
}
function _dependencies() {
  const data = _interopRequireDefault(require("../dependencies"));
  _dependencies = function () {
    return data;
  };
  return data;
}
function _filesDependencyBuilder() {
  const data = require("../files-dependency-builder");
  _filesDependencyBuilder = function () {
    return data;
  };
  return data;
}
function _overridesDependencies() {
  const data = _interopRequireDefault(require("./overrides-dependencies"));
  _overridesDependencies = function () {
    return data;
  };
  return data;
}
function _dependenciesData() {
  const data = require("./dependencies-data");
  _dependenciesData = function () {
    return data;
  };
  return data;
}
function _packageToDefinetlyTyped() {
  const data = require("./package-to-definetly-typed");
  _packageToDefinetlyTyped = function () {
    return data;
  };
  return data;
}
function _packageJsonFile() {
  const data = _interopRequireDefault(require("../../../../consumer/component/package-json-file"));
  _packageJsonFile = function () {
    return data;
  };
  return data;
}
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
class DependencyResolver {
  static registerWorkspacePolicyGetter(func) {
    this.getWorkspacePolicy = func;
  }

  /**
   * This will get the peers policy provided by the env of the component
   */

  static registerHarmonyEnvPeersPolicyForComponentGetter(func) {
    this.getHarmonyEnvPeersPolicyForComponent = func;
  }

  /**
   * This will get the peers policy provided by the env of the component
   */

  static registerHarmonyEnvPeersPolicyForEnvItselfGetter(func) {
    this.getHarmonyEnvPeersPolicyForEnvItself = func;
  }
  constructor(component, consumer) {
    (0, _defineProperty2().default)(this, "component", void 0);
    (0, _defineProperty2().default)(this, "consumer", void 0);
    (0, _defineProperty2().default)(this, "componentId", void 0);
    (0, _defineProperty2().default)(this, "componentMap", void 0);
    (0, _defineProperty2().default)(this, "componentFromModel", void 0);
    (0, _defineProperty2().default)(this, "consumerPath", void 0);
    (0, _defineProperty2().default)(this, "tree", void 0);
    (0, _defineProperty2().default)(this, "allDependencies", void 0);
    (0, _defineProperty2().default)(this, "allPackagesDependencies", void 0);
    (0, _defineProperty2().default)(this, "issues", void 0);
    (0, _defineProperty2().default)(this, "coreAspects", []);
    (0, _defineProperty2().default)(this, "processedFiles", void 0);
    (0, _defineProperty2().default)(this, "overridesDependencies", void 0);
    (0, _defineProperty2().default)(this, "debugDependenciesData", void 0);
    this.component = component;
    this.consumer = consumer;
    this.componentId = component.id;
    // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
    this.componentMap = this.component.componentMap;
    // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
    this.componentFromModel = this.component.componentFromModel;
    this.consumerPath = this.consumer.getPath();
    this.allDependencies = {
      dependencies: [],
      devDependencies: []
    };
    this.allPackagesDependencies = {
      packageDependencies: {},
      devPackageDependencies: {},
      peerPackageDependencies: {}
    };
    this.processedFiles = [];
    this.issues = component.issues;
    this.setLegacyInsideHarmonyIssue();
    this.overridesDependencies = new (_overridesDependencies().default)(component, consumer);
    this.debugDependenciesData = {
      components: []
    };
  }
  setTree(tree) {
    this.tree = tree;
    // console.log(JSON.stringify(tree, null, 4)); // uncomment to easily watch the tree received from bit-javascript
  }

  /**
   * Resolve components and packages dependencies for a component.
   * This method should NOT have any side-effect on the component. the DependenciesLoader class is
   * responsible for saving this data on the component object.
   *
   * The process is as follows:
   * 1) Use the language driver to parse the component files and find for each file its dependencies.
   * 2) The results we get from the driver per file tells us what are the files and packages that depend on our file.
   * and also whether there are missing packages and files.
   * 3) Using the information from the driver, we go over each one of the dependencies files and find its counterpart
   * component. The way how we find it, is by using the bit.map file which has a mapping between the component name and
   * the file paths.
   * 4) If we find a component to the file dependency, we add it to component.dependencies. Otherwise, it's added to
   * component.issues.untrackedDependencies
   * 5) Similarly, when we find the packages dependencies, they are added to component.packageDependencies. Otherwise,
   * they're added to component.issues.missingPackagesDependenciesOnFs
   * 6) In case the driver found a file dependency that is not on the file-system, we add that file to
   * component.issues.missingDependenciesOnFs
   */
  async getDependenciesData(cacheResolvedDependencies, cacheProjectAst) {
    var _this$componentMap, _this$componentMap2;
    const componentDir = (_this$componentMap = this.componentMap) !== null && _this$componentMap !== void 0 && _this$componentMap.rootDir ? path().join(this.consumerPath, (_this$componentMap2 = this.componentMap) === null || _this$componentMap2 === void 0 ? void 0 : _this$componentMap2.rootDir) : this.consumerPath;
    const {
      nonTestsFiles,
      testsFiles
    } = this.componentMap.getFilesGroupedByBeingTests();
    const allFiles = [...nonTestsFiles, ...testsFiles];
    // find the dependencies (internal files and packages) through automatic dependency resolution
    const dependenciesTree = await (0, _filesDependencyBuilder().getDependencyTree)({
      componentDir,
      workspacePath: this.consumerPath,
      filePaths: allFiles,
      bindingPrefix: this.component.bindingPrefix,
      resolveModulesConfig: this.consumer.config._resolveModules,
      visited: cacheResolvedDependencies,
      cacheProjectAst
    });
    // we have the files dependencies, these files should be components that are registered in bit.map. Otherwise,
    // they are referred as "untracked components" and the user should add them later on in order to tag
    this.setTree(dependenciesTree.tree);
    const devFiles = await DependencyResolver.getDevFiles(this.component);
    await this.populateDependencies(allFiles, devFiles);
    return new (_dependenciesData().DependenciesData)(this.allDependencies, this.allPackagesDependencies, this.issues, this.coreAspects, {
      manuallyRemovedDependencies: this.overridesDependencies.manuallyRemovedDependencies,
      manuallyAddedDependencies: this.overridesDependencies.manuallyAddedDependencies,
      missingPackageDependencies: this.overridesDependencies.missingPackageDependencies
    });
  }

  /**
   * Given the tree of file dependencies from the driver, find the components of these files.
   * Each dependency file has a path, use bit.map to search for the component name by that path.
   * If the component is found, add it to "this.allDependencies.dependencies". Otherwise, add it to "this.issues.untrackedDependencies".
   *
   * For the found components, add their sourceRelativePath and destinationRelativePath, they are being used for
   * generating links upon import:
   * sourceRelativePath - location of the link file.
   * destinationRelativePath - destination written inside the link file.
   *
   * When a dependency is found in a regular (implementation) file, it goes to `dependencies`. If
   * it found on a test file, it goes to `devDependencies`.
   * Similarly, when a package is found in a regular file, it goes to `packageDependencies`. When
   * if found in a test file, it goes to `devPackageDependencies`.
   * An exception for the above is when a package is required in a regular or test file but is also
   * mentioned in the `package.json` file as a peerDependency, in that case, the package is added
   * to `peerPackageDependencies` and removed from other places. Unless this package is overridden
   * and marked as ignored in the consumer or component config file.
   */
  async populateDependencies(files, testsFiles) {
    files.forEach(file => {
      const fileType = {
        isTestFile: testsFiles.includes(file)
      };
      this.throwForNonExistFile(file);
      if (this.overridesDependencies.shouldIgnoreFile(file, fileType)) {
        return;
      }
      this.processCoreAspects(file);
      this.processMissing(file, fileType);
      this.processErrors(file);
      this.processPackages(file, fileType);
      this.processComponents(file, fileType);
      this.processDepFiles(file, fileType);
      this.processUnidentifiedPackages(file, fileType);
    });
    this.removeIgnoredPackagesByOverrides();
    this.removeDevAndEnvDepsIfTheyAlsoRegulars();
    this.addCustomResolvedIssues();
    this.applyPeersFromComponentModel();
    this.applyPackageJson();
    this.applyWorkspacePolicy();
    await this.applyAutoDetectedPeersFromEnvOnComponent();
    this.manuallyAddDependencies();
    // Doing this here (after manuallyAddDependencies) because usually the env of the env is adding dependencies as peer of the env
    // which will make this not work if it come before
    // example:
    // custom react has peers with react 16.4.0.
    // the custom react uses the "teambit.envs/env" env, which will add react ^17.0.0 to every component that uses it
    // we want to make sure that the custom react is using 16.4.0 not 17.
    await this.applyAutoDetectedPeersFromEnvOnEnvItSelf();
    this.coreAspects = (0, _uniq2().default)(this.coreAspects);
  }
  addCustomResolvedIssues() {
    const deps = [...this.allDependencies.dependencies, ...this.allDependencies.devDependencies];
    const customModulesDeps = deps.filter(dep => dep.relativePaths.some(r => r.isCustomResolveUsed));
    if (customModulesDeps.length) {
      const importSources = customModulesDeps.reduce((acc, current) => {
        current.relativePaths.forEach(relativePath => {
          if (relativePath.isCustomResolveUsed) {
            // @ts-ignore
            acc[relativePath.importSource] = current.id.toStringWithoutVersion();
          }
        });
        return acc;
      }, {});
      this.issues.getOrCreate(_componentIssues().IssuesClasses.CustomModuleResolutionUsed).data = importSources;
    }
  }
  removeIgnoredPackagesByOverrides() {
    const shouldBeIncluded = (pkgVersion, pkgName) => !this.overridesDependencies.shouldIgnorePackageByType(pkgName, 'dependencies');
    const shouldBeIncludedDev = (pkgVersion, pkgName) => !this.overridesDependencies.shouldIgnorePackageByType(pkgName, 'devDependencies');
    this.allPackagesDependencies.packageDependencies = (0, _pickBy2().default)(shouldBeIncluded, this.allPackagesDependencies.packageDependencies);
    this.allPackagesDependencies.devPackageDependencies = (0, _pickBy2().default)(shouldBeIncludedDev, this.allPackagesDependencies.devPackageDependencies);
  }
  throwForNonExistFile(file) {
    if (!this.tree[file]) {
      throw new Error(`DependencyResolver: a file "${file}" was not returned from the driver, its dependencies are unknown`);
    }
  }
  manuallyAddDependencies() {
    const packageJson = this._getPackageJson();
    const dependencies = this.overridesDependencies.getDependenciesToAddManually(packageJson, this.allDependencies);
    if (!dependencies) return;
    const {
      components,
      packages
    } = dependencies;
    _constants().DEPENDENCIES_FIELDS.forEach(depField => {
      if (components[depField] && components[depField].length) {
        components[depField].forEach(depData => this.allDependencies[depField].push({
          id: depData.componentId,
          relativePaths: [],
          packageName: depData.packageName
        }));
      }
      if (packages[depField] && !(0, _isEmpty2().default)(packages[depField])) {
        Object.assign(this.allPackagesDependencies[this._pkgFieldMapping(depField)], packages[depField]);
      }
    });
    // The automatic dependency detector considers all found dependencies to be runtime dependencies.
    // But this breaks proper installation of injected subdependencies that are resolved from workspace components.
    if (this.allPackagesDependencies.packageDependencies && packages.peerDependencies) {
      for (const peerName of Object.keys(packages.peerDependencies)) {
        delete this.allPackagesDependencies.packageDependencies[peerName];
      }
    }
    if (this.allPackagesDependencies.packageDependencies && packages.peerPackageDependencies) {
      for (const peerName of Object.keys(packages.peerPackageDependencies)) {
        delete this.allPackagesDependencies.packageDependencies[peerName];
      }
    }
  }
  traverseTreeForComponentId(depFile) {
    if (!this.tree[depFile] || !this.tree[depFile].files && !this.tree[depFile].components) return undefined;
    if (!this.componentMap.rootDir) {
      throw Error('traverseTreeForComponentId should get called only when rootDir is set');
    }
    const rootDirFullPath = path().join(this.consumerPath, this.componentMap.rootDir);
    const files = this.tree[depFile].files || [];
    if (files && !(0, _isEmpty2().default)(files)) {
      for (const file of files) {
        const fullDepFile = path().resolve(rootDirFullPath, file.file);
        const depRelativeToConsumer = (0, _utils().pathNormalizeToLinux)(path().relative(this.consumerPath, fullDepFile));
        const componentId = this.consumer.bitMap.getComponentIdByPath(depRelativeToConsumer);
        if (componentId) return componentId;
      }
    }
    if (this.tree[depFile].components && !(0, _isEmpty2().default)(this.tree[depFile].components)) {
      const components = this.tree[depFile].components || [];
      for (const comp of components) {
        return this.getComponentIdByResolvedPackageData(comp);
      }
    }
    if (files && !(0, _isEmpty2().default)(files)) {
      for (const file of files) {
        if (file.file !== depFile) {
          const componentId = this.traverseTreeForComponentId(file.file);
          if (componentId) return componentId;
        } else {
          _logger().default.warn(`traverseTreeForComponentId found a cyclic dependency. ${file.file} depends on itself`);
        }
      }
    }
    return undefined;
  }
  getComponentIdByResolvedPackageData(bit) {
    if (!bit.componentId) {
      throw new Error(`resolved Bit component must have componentId prop in the package.json file`);
    }
    return bit.componentId;
  }

  /**
   * this happens when using relative paths between components, which is allowed on Legacy only.
   * on Harmony, during the execution of this function, it recognizes the use of relative-paths, enter
   * it to the "issues", then, later, it shows a warning on bit-status and block tagging.
   */
  getComponentIdByDepFile(depFile) {
    let depFileRelative = depFile; // dependency file path relative to consumer root
    let componentId;
    let destination;
    const rootDir = this.componentMap.rootDir;
    if (rootDir) {
      // The depFileRelative is relative to rootDir, change it to be relative to current consumer.
      // We can't use path.resolve(rootDir, fileDep) because this might not work when running
      // bit commands not from root, because resolve take by default the process.cwd
      const rootDirFullPath = path().join(this.consumerPath, rootDir);
      const fullDepFile = path().resolve(rootDirFullPath, depFile);
      depFileRelative = (0, _utils().pathNormalizeToLinux)(path().relative(this.consumerPath, fullDepFile));
    }
    componentId = this.consumer.bitMap.getComponentIdByPath(depFileRelative);
    if (!componentId) componentId = this._getComponentIdFromCustomResolveToPackageWithDist(depFileRelative);
    return {
      componentId,
      depFileRelative,
      destination
    };
  }

  /**
   * this is a workaround for cases where an alias points to a package with dist.
   * normally, aliases are created for local directories.
   * they can be however useful when a source code can't be touched and `require` to one package
   * needs to be replaced with a `require` to a component. in that case, our options are:
   * 1) point the alias to the package name.
   * 2) point the alias to the relative directory of the imported component
   * the ideal solution is #1, however, it requires changes in the Tree structure, which should
   * allow "components" to have more data, such as importSource.
   * here, we go option #2, the alias is a relative path to the component. however, when there is
   * dist directory, the resolved path contains the "dist", which doesn't exist in the ComponentMap,
   * the solution we take is to identify such cases, strip the dist, then try to find them again.
   */
  _getComponentIdFromCustomResolveToPackageWithDist(depFile) {
    if (!depFile.includes('dist')) return null;
    const resolveModules = this.consumer.config._resolveModules;
    if (!resolveModules || !resolveModules.aliases) return null;
    const aliases = resolveModules.aliases;
    const foundAlias = Object.keys(aliases).find(alias => depFile.startsWith(aliases[alias]));
    if (!foundAlias) return null;
    const newDepFile = depFile.replace(`${resolveModules.aliases[foundAlias]}/dist`, resolveModules.aliases[foundAlias]);
    return this.consumer.bitMap.getComponentIdByPath(newDepFile);
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  getDependencyPathsFromModel(componentId, depFile, rootDir) {
    const dependency = this.componentFromModel.getAllDependencies().find(dep => dep.id.isEqualWithoutVersion(componentId));
    if (!dependency) {
      throw new (_showDoctorError().default)(
      // $FlowFixMe
      `the auto-generated file ${depFile} should be connected to ${componentId}, however, it's not part of the model dependencies of ${this.componentFromModel.id}`);
    }
    const relativePath = dependency.relativePaths.find(r => r.sourceRelativePath === depFile);
    if (!relativePath) {
      throw new (_showDoctorError().default)(`unable to find ${relativePath} path in the dependencies relativePaths of ${this.componentFromModel.id}`);
    }
    return {
      componentId: dependency.id,
      destination: relativePath.destinationRelativePath,
      depFileRelative: depFile
    };
  }
  processDepFiles(originFile, fileType, nested = false) {
    // We don't just return because different files of the component might import different things from the depFile
    // See more info here: https://github.com/teambit/bit/issues/1796
    if (!this.processedFiles.includes(originFile)) {
      this.processedFiles.push(originFile);
      // We don't want to calculate nested files again after they calculated as direct files
    } else if (nested) {
      return;
    }
    const allDepsFiles = this.tree[originFile].files;
    if (!allDepsFiles || (0, _isEmpty2().default)(allDepsFiles)) return;
    allDepsFiles.forEach(depFile => {
      if (!nested && this.overridesDependencies.shouldIgnoreFile(depFile.file, fileType)) return;
      if (depFile.isLink) this.processLinkFile(originFile, depFile, fileType);else {
        const isDepFileUntracked = this.processOneDepFile(originFile, depFile.file, depFile.importSpecifiers, undefined, fileType, depFile, nested);
        // Only continue recursively if the dep file is untracked
        // for tracked deps if they have untracked deps they will be shown under their own components
        if (isDepFileUntracked) {
          // Recursively check for untracked files (to show them all in bit status)
          // for nested files we don't really care about the file types since we won't do all the checking
          const dummyFileType = {
            isTestFile: false
          };
          this.processDepFiles(depFile.file, dummyFileType, true);
        }
      }
    });
  }

  // return true if the dep file is untracked
  // eslint-disable-next-line complexity
  processOneDepFile(originFile, depFile, importSpecifiers, linkFile, fileType, depFileObject, nested = false) {
    const {
      componentId,
      depFileRelative,
      destination
    } = this.getComponentIdByDepFile(depFile);
    const importSource = depFileObject.importSource;
    // the file dependency doesn't have any counterpart component. Add it to this.issues.untrackedDependencies
    if (!componentId) {
      var _this$tree$depFile, _this$tree$depFile$mi;
      const missingComponents = (_this$tree$depFile = this.tree[depFile]) === null || _this$tree$depFile === void 0 ? void 0 : (_this$tree$depFile$mi = _this$tree$depFile.missing) === null || _this$tree$depFile$mi === void 0 ? void 0 : _this$tree$depFile$mi.components;
      if (missingComponents) {
        // this depFile is a dependency link and this dependency is missing.
        // it can't happen on Harmony, as Harmony doesn't have the generated dependencies files.
        this.throwForMissingComponentsOnHarmony(missingComponents);
        return false;
      }
      this._pushToUntrackDependenciesIssues(originFile, depFileRelative, nested);
      return true;
    }
    if (this.overridesDependencies.shouldIgnoreComponent(componentId, fileType)) {
      // we can't support it because on the imported side, we don't know to convert the relative path
      // to the component name, as it won't have the component installed
      throw new (_generalError().default)(`unable to ignore "${componentId.toString()}" dependency of "${this.componentId.toString()}" by using ignore components syntax because the component is required with relative path.
either, use the ignore file syntax or change the require statement to have a module path`);
    }
    // happens when in the same component one file requires another one. In this case, there is
    // noting to do regarding the dependencies
    if (componentId.isEqualWithoutVersion(this.componentId)) {
      return false;
    }
    const depComponentMap = this.consumer.bitMap.getComponentIfExist(componentId);
    // found a dependency component. Add it to this.allDependencies.dependencies
    const depRootDir = depComponentMap ? depComponentMap.rootDir : undefined;
    const destinationRelativePath = destination || (depRootDir && depFileRelative.startsWith(depRootDir) ? (0, _utils().pathRelativeLinux)(depRootDir, depFileRelative) : depFileRelative);
    let sourceRelativePath;
    if (linkFile) {
      sourceRelativePath = linkFile;
    } else {
      // when there is no rootDir for the current dependency (it happens when it's AUTHORED), keep the original path
      sourceRelativePath = depRootDir ? depFileRelative : depFile;
    }
    const depsPaths = {
      sourceRelativePath,
      destinationRelativePath
    };
    if (importSpecifiers) {
      importSpecifiers.forEach(importSpecifier => {
        // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
        if (importSpecifier.linkFile) delete importSpecifier.linkFile.exported;
        // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
        if (importSpecifier.mainFile) delete importSpecifier.mainFile.exported;
      });
      depsPaths.importSpecifiers = importSpecifiers;
    }
    if (depFileObject.isCustomResolveUsed) {
      depsPaths.isCustomResolveUsed = depFileObject.isCustomResolveUsed;
      depsPaths.importSource = importSource;
    }
    const currentComponentsDeps = {
      id: componentId,
      relativePaths: [depsPaths]
    };
    this._pushToRelativeComponentsAuthoredIssues(originFile, componentId, importSource, depsPaths);
    const allDependencies = [...this.allDependencies.dependencies, ...this.allDependencies.devDependencies];
    const existingDependency = this.getExistingDependency(allDependencies, componentId);
    if (existingDependency) {
      const existingDepRelativePaths = this.getExistingDepRelativePaths(existingDependency, depsPaths);
      if (!existingDepRelativePaths) {
        // it is another file of an already existing component. Just add the new path
        existingDependency.relativePaths.push(depsPaths);
        return false;
      }
      // The dep path already exists but maybe this dep-file has more importSpecifiers
      if (depsPaths.importSpecifiers) {
        // add them to the existing dep
        if (!existingDepRelativePaths.importSpecifiers) {
          existingDepRelativePaths.importSpecifiers = [...depsPaths.importSpecifiers];
        } else {
          // both have importSpecifiers
          const nonExistingImportSpecifiers = this.getDiffSpecifiers(existingDepRelativePaths.importSpecifiers, depsPaths.importSpecifiers);
          existingDepRelativePaths.importSpecifiers.push(...nonExistingImportSpecifiers);
        }
      }
      // Handle cases when the first dep paths are not custom resolved and the new one is
      if (depsPaths.isCustomResolveUsed && !existingDepRelativePaths.isCustomResolveUsed) {
        existingDepRelativePaths.isCustomResolveUsed = depsPaths.isCustomResolveUsed;
      }
      if (depsPaths.importSource && !existingDepRelativePaths.importSource) {
        existingDepRelativePaths.importSource = depsPaths.importSource;
      }
    } else {
      const depDebug = {
        id: currentComponentsDeps.id,
        importSource
      };
      this.pushToDependenciesArray(currentComponentsDeps, fileType, depDebug);
    }
    return false;
  }
  processLinkFile(originFile, linkFile, fileType) {
    if (!linkFile.linkDependencies || (0, _isEmpty2().default)(linkFile.linkDependencies)) return;
    const nonLinkImportSpecifiers = [];
    linkFile.linkDependencies.forEach(dependency => {
      const component = this.getComponentIdByDepFile(linkFile.file);
      if (component.componentId) {
        // the linkFile is already a component, no need to treat it differently than other depFile
        // aggregate all dependencies using the same linkFile and ultimately run processDepFile
        // with all importSpecifiers of that linkFile.
        // also, delete the linkFile attribute of importSpecifiers so then once the component is
        // imported and the link is generated, it won't be treated as a linkFile.
        // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
        dependency.importSpecifiers.map(a => delete a.linkFile);
        // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
        nonLinkImportSpecifiers.push(dependency.importSpecifiers);
      } else {
        this.processOneDepFile(originFile,
        // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
        dependency.file,
        // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
        dependency.importSpecifiers, linkFile.file, fileType, linkFile);
      }
    });
    if (nonLinkImportSpecifiers.length) {
      this.processOneDepFile(originFile, linkFile.file, (0, _flatten2().default)(nonLinkImportSpecifiers), undefined, fileType, linkFile);
    }
  }

  /**
   * process require/import of Bit components where the require statement is not a relative path
   * but a module path, such as `require('@bit/bit.envs/compiler/babel');`
   */
  processComponents(originFile, fileType) {
    const components = this.tree[originFile].components;
    if (!components || (0, _isEmpty2().default)(components)) return;
    components.forEach(compDep => {
      let componentId = this.getComponentIdByResolvedPackageData(compDep);
      if (componentId.isEqual(this.componentId)) {
        // the component is importing itself, so ignore it. although currently it doesn't cause any issues, (probably
        // because it filtered out later), it's better to remove it as soon as possible, for less-confusing debugging.
        return;
      }
      const depDebug = {
        id: componentId,
        dependencyPackageJsonPath: compDep.packageJsonPath,
        dependentPackageJsonPath: compDep.dependentPackageJsonPath,
        componentIdResolvedFrom: 'DependencyPkgJson',
        packageName: compDep.name
      };
      const getVersionFromPkgJson = () => {
        const versionFromDependencyPkgJson = this.getValidVersion(compDep.concreteVersion);
        if (versionFromDependencyPkgJson) {
          depDebug.versionResolvedFrom = 'DependencyPkgJson';
          return versionFromDependencyPkgJson;
        }
        const versionFromDependentPkgJson = this.getValidVersion(compDep.versionUsedByDependent);
        if (versionFromDependentPkgJson) {
          depDebug.versionResolvedFrom = 'DependentPkgJson';
          return versionFromDependentPkgJson;
        }
        return null;
      };
      const version = getVersionFromPkgJson();
      if (version) {
        componentId = componentId.changeVersion(version);
      }
      if (this.overridesDependencies.shouldIgnoreComponent(componentId, fileType)) {
        return;
      }
      const getExistingIdFromBitMapOrModel = () => {
        const existingIds = this.consumer.bitmapIdsFromCurrentLane.filterWithoutVersion(componentId);
        if (existingIds.length === 1) {
          depDebug.versionResolvedFrom = 'BitMap';
          return existingIds[0];
        }
        if (this.componentFromModel) {
          const modelDep = this.componentFromModel.getAllDependenciesIds().searchWithoutVersion(componentId);
          if (modelDep) {
            depDebug.versionResolvedFrom = 'Model';
            return modelDep;
          }
        }
        return undefined;
      };
      const getExistingId = () => {
        var _getExistingIdFromBit;
        // Happens when the dep is not in the node_modules or it's there without a version
        // (it's there without a version when it's in the workspace and it's linked)
        if (!version) return getExistingIdFromBitMapOrModel();

        // In case it's resolved from the node_modules, and it's also in the ws policy or variants,
        // use the resolved version from the node_modules / package folder
        if (this.isPkgInWorkspacePolicies(compDep.name) || this.isPkgInVariants(compDep.name)) {
          return componentId;
        }

        // If there is a version in the node_modules/package folder, but it's not in the ws policy,
        // prefer the version from the bitmap/model over the version from the node_modules
        return (_getExistingIdFromBit = getExistingIdFromBitMapOrModel()) !== null && _getExistingIdFromBit !== void 0 ? _getExistingIdFromBit : componentId;
      };
      const existingId = getExistingId();
      if (existingId) {
        if (existingId.isEqual(this.componentId)) {
          // happens when one of the component files requires another using module path
          // no need to enter anything to the dependencies
          return;
        }
        this.addImportNonMainIssueIfNeeded(originFile, compDep);
        const currentComponentsDeps = {
          id: existingId,
          relativePaths: [],
          packageName: compDep.name
        };
        this._pushToDependenciesIfNotExist(currentComponentsDeps, fileType, depDebug);
      } else {
        this._pushToMissingComponentsIssues(originFile, componentId);
      }
    });
  }
  isPkgInWorkspacePolicies(pkgName) {
    var _DependencyResolver$g;
    return (_DependencyResolver$g = DependencyResolver.getWorkspacePolicy().dependencies) === null || _DependencyResolver$g === void 0 ? void 0 : _DependencyResolver$g[pkgName];
  }
  isPkgInVariants(pkgName) {
    const dependencies = this.overridesDependencies.getDependenciesToAddManually(undefined, this.allDependencies);
    if (!dependencies) return false;
    const {
      components
    } = dependencies;
    return _constants().DEPENDENCIES_FIELDS.some(depField => components[depField] && components[depField].some(depData => depData.packageName === pkgName));
  }
  addImportNonMainIssueIfNeeded(filePath, dependencyPkgData) {
    var _dependencyPkgData$pa, _dependencyPkgData$pa2, _this$issues$getOrCre;
    const depMain = (_dependencyPkgData$pa = dependencyPkgData.packageJsonContent) === null || _dependencyPkgData$pa === void 0 ? void 0 : _dependencyPkgData$pa.main;
    if (!depMain) {
      return;
    }
    const depFullPath = (0, _utils().pathNormalizeToLinux)(dependencyPkgData.fullPath);
    if (depFullPath.endsWith(depMain)) {
      // it requires the main-file. all is good.
      return;
    }
    const extDisallowNonMain = ['.ts', '.tsx', '.js', '.jsx'];
    if (!extDisallowNonMain.includes(path().extname(depFullPath))) {
      // some files such as scss/json are needed to be imported as non-main
      return;
    }
    const pkgRootDir = (_dependencyPkgData$pa2 = dependencyPkgData.packageJsonContent) === null || _dependencyPkgData$pa2 === void 0 ? void 0 : _dependencyPkgData$pa2.componentRootFolder;
    if (pkgRootDir && !_fsExtra().default.existsSync(path().join(pkgRootDir, _constants().DEFAULT_DIST_DIRNAME))) {
      // the dependency wasn't compiled yet. the issue is probably because depMain points to the dist
      // and depFullPath is in the source.
      return;
    }
    const nonMainFileSplit = depFullPath.split(`node_modules/`);
    const nonMainFileShort = nonMainFileSplit[1] || nonMainFileSplit[0];
    ((_this$issues$getOrCre = this.issues.getOrCreate(_componentIssues().IssuesClasses.ImportNonMainFiles).data)[filePath] || (_this$issues$getOrCre[filePath] = [])).push(nonMainFileShort);
  }
  getValidVersion(version) {
    if (!version) {
      return null;
    }
    if (_semver().default.valid(version)) {
      // this takes care of pre-releases as well, as they're considered valid semver.
      return version;
    }
    if (_semver().default.validRange(version)) {
      // if this is a range, e.g. ^1.0.0, return a valid version: 1.0.0.
      const coerced = _semver().default.coerce(version);
      if (coerced) {
        return coerced.version;
      }
    }
    // it's probably a relative path to the component
    return null;
  }
  processPackages(originFile, fileType) {
    const packages = this.tree[originFile].packages;
    if (this.componentFromModel) {
      const modelDeps = this.componentFromModel.getAllPackageDependencies();
      // If a package is not in the policies, then we resolve the package from the model.
      for (const pkgName of Object.keys(packages)) {
        if (!this.isPkgInWorkspacePolicies(pkgName) && modelDeps[pkgName]) {
          packages[pkgName] = modelDeps[pkgName];
        }
      }
    }
    if (!packages || (0, _isEmpty2().default)(packages)) return;
    if (fileType.isTestFile) {
      Object.assign(this.allPackagesDependencies.devPackageDependencies, packages);
    } else {
      Object.assign(this.allPackagesDependencies.packageDependencies, packages);
    }
    this._addTypesPackagesForTypeScript(packages, originFile);
  }
  processMissing(originFile, fileType) {
    const missing = this.tree[originFile].missing;
    if (!missing) return;
    const processMissingFiles = () => {
      // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
      if ((0, _lodash().isEmpty)(missing.files)) return;
      const absOriginFile = this.consumer.toAbsolutePath(originFile);
      // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
      const missingFiles = missing.files.filter(missingFile => {
        // convert from importSource (the string inside the require/import call) to the path relative to consumer
        const resolvedPath = path().resolve(path().dirname(absOriginFile), missingFile);
        const relativeToConsumer = this.consumer.getPathRelativeToConsumer(resolvedPath);
        return !this.overridesDependencies.shouldIgnoreFile(relativeToConsumer, fileType);
      });
      if ((0, _isEmpty2().default)(missingFiles)) return;
      this._pushToMissingDependenciesOnFs(originFile, missingFiles);
    };
    const processMissingPackages = () => {
      if ((0, _lodash().isEmpty)(missing.packages)) return;
      const missingPackages = missing.packages.filter(pkg => !this.overridesDependencies.shouldIgnorePackage(pkg, fileType));
      if ((0, _isEmpty2().default)(missingPackages)) return;
      const customResolvedDependencies = this.findOriginallyCustomModuleResolvedDependencies(missingPackages);
      const missingPackagesWithoutCustomResolved = customResolvedDependencies ? (0, _difference2().default)(missingPackages, Object.keys(customResolvedDependencies)) : missingPackages;
      if (!(0, _isEmpty2().default)(missingPackagesWithoutCustomResolved)) {
        this._pushToMissingPackagesDependenciesIssues(originFile, missingPackages);
      }
    };
    const processMissingComponents = () => {
      if ((0, _lodash().isEmpty)(missing.components)) return;
      this.throwForMissingComponentsOnHarmony(missing.components);
    };
    processMissingFiles();
    processMissingPackages();
    processMissingComponents();
  }
  throwForMissingComponentsOnHarmony(missingComponents) {
    // on Harmony we don't guess whether a path is a component or a package based on the path only,
    // see missing-handler.groupMissingByType() for more info.
    throw new Error(`Harmony should not have "missing.components" (only "missing.packages"). got ${missingComponents.join(',')}`);
  }
  processErrors(originFile) {
    const error = this.tree[originFile].error;
    if (!error) return;
    _logger().default.errorAndAddBreadCrumb('dependency-resolver.processErrors', 'got an error from the driver while resolving dependencies');
    _logger().default.error('dependency-resolver.processErrors', error);
    if (error.code === 'PARSING_ERROR') {
      const location = error.lineNumber && error.column ? ` (line: ${error.lineNumber}, column: ${error.column})` : '';
      this.issues.getOrCreate(_componentIssues().IssuesClasses.ParseErrors).data[originFile] = error.message + location;
    } else this.issues.getOrCreate(_componentIssues().IssuesClasses.ResolveErrors).data[originFile] = error.message;
  }

  /**
   * when a user uses core-extensions these core-extensions should not be dependencies.
   * here, we filter them out from all places they could entered as dependencies.
   * an exception is when running this method on bit-core-extensions themselves (dogfooding), in
   * which case we recognizes that the current originFile is a core-extension and avoid filtering.
   */
  processCoreAspects(originFile) {
    var _DependencyResolver$g2, _this$tree$originFile, _this$tree$originFile2;
    const coreAspects = (_DependencyResolver$g2 = DependencyResolver.getCoreAspectsPackagesAndIds) === null || _DependencyResolver$g2 === void 0 ? void 0 : _DependencyResolver$g2.call(DependencyResolver);
    if (!coreAspects) {
      return;
    }

    // const scopes = coreAspects.map((id) => {
    //   const id = id.split()
    // });

    const coreAspectIds = Object.values(coreAspects);
    const defaultScope = this.component.defaultScope;
    let id;
    if (this.component.id.scope) {
      id = this.component.id;
    } else {
      id = this.component.id.changeScope(defaultScope);
    }
    if (coreAspectIds.includes(id.toStringWithoutVersion())) {
      return;
    }
    const coreAspectsPackages = Object.keys(coreAspects);
    const components = this.tree[originFile].components;
    const unidentifiedPackages = this.tree[originFile].unidentifiedPackages;
    const missingComponents = (_this$tree$originFile = this.tree[originFile]) === null || _this$tree$originFile === void 0 ? void 0 : (_this$tree$originFile2 = _this$tree$originFile.missing) === null || _this$tree$originFile2 === void 0 ? void 0 : _this$tree$originFile2.components;
    const usedCoreAspects = [];
    const findMatchingCoreAspect = packageName => {
      return coreAspectsPackages.find(coreAspectName => packageName === coreAspectName);
    };
    const unidentifiedPackagesFiltered = unidentifiedPackages === null || unidentifiedPackages === void 0 ? void 0 : unidentifiedPackages.filter(packageName => {
      const matchingCoreAspectPackageName = findMatchingCoreAspect(packageName);
      if (matchingCoreAspectPackageName) {
        usedCoreAspects.push(coreAspects[matchingCoreAspectPackageName]);
      }
      return !matchingCoreAspectPackageName;
    });
    const bitsFiltered = components === null || components === void 0 ? void 0 : components.filter(packageInfo => {
      const matchingCoreAspectPackageName = findMatchingCoreAspect(packageInfo.name);
      if (matchingCoreAspectPackageName) {
        usedCoreAspects.push(coreAspects[matchingCoreAspectPackageName]);
      }
      return !matchingCoreAspectPackageName;
    });
    const missingBitsFiltered = missingComponents === null || missingComponents === void 0 ? void 0 : missingComponents.filter(packageName => {
      const matchingCoreAspectPackageName = findMatchingCoreAspect(packageName);
      if (matchingCoreAspectPackageName) {
        usedCoreAspects.push(coreAspects[matchingCoreAspectPackageName]);
      }
      return !matchingCoreAspectPackageName;
    });
    if (missingComponents) {
      // @ts-ignore not clear why this error happens, it is verified that missing.components exist
      this.tree[originFile].missing.components = missingBitsFiltered;
    }
    this.tree[originFile].unidentifiedPackages = unidentifiedPackagesFiltered;
    this.tree[originFile].components = bitsFiltered;
    this.coreAspects.push(...(0, _uniq2().default)(usedCoreAspects));
  }

  /**
   * currently the only unidentified packages being process are the ones coming from custom-modules-resolution.
   * assuming the author used custom-resolution, which enable using non-relative import syntax,
   * for example, requiring the file 'src/utils/is-string' from anywhere as require('utils/is-string');
   * now, when the component is imported, the driver recognizes 'utils/is-string' as a package,
   * because it's not relative.
   * the goal here is to use the 'package' the driver found and match it with one of the
   * dependencies from the model. In the example above, we might find in the model, a dependency
   * is-string with importSource of 'utils/is-string'.
   * Once a match is found, copy the relativePaths from the model.
   *
   * keep in mind that this custom-modules-resolution supported on legacy components only.
   * as such, no need to find the packageName to pass to _pushToDependenciesIfNotExist method.
   */
  processUnidentifiedPackages(originFile, fileType) {
    const unidentifiedPackages = this.tree[originFile].unidentifiedPackages;
    if (!unidentifiedPackages || !unidentifiedPackages.length) return;
    this.debugDependenciesData.unidentifiedPackages = unidentifiedPackages;
    if (!this.componentFromModel) return; // not relevant, the component is not imported
    const getDependencies = () => {
      if (fileType.isTestFile) return this.componentFromModel.devDependencies;
      return this.componentFromModel.dependencies;
    };
    const dependencies = getDependencies();
    if (dependencies.isEmpty()) return;
    const importSourceMap = dependencies.getCustomResolvedData();
    if ((0, _isEmpty2().default)(importSourceMap)) return;
    const clonedDependencies = new (_dependencies().default)(dependencies.getClone());
    unidentifiedPackages.forEach(unidentifiedPackage => {
      const packageLinuxFormat = (0, _utils().pathNormalizeToLinux)(unidentifiedPackage);
      const packageWithNoNodeModules = packageLinuxFormat.replace('node_modules/', '');
      const foundImportSource = Object.keys(importSourceMap).find(importSource => packageWithNoNodeModules.startsWith(importSource));
      if (foundImportSource) {
        const dependencyId = importSourceMap[foundImportSource];
        const currentComponentDeps = {
          id: dependencyId,
          // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
          relativePaths: clonedDependencies.getById(dependencyId).relativePaths
        };
        const depData = {
          id: dependencyId,
          importSource: foundImportSource
        };
        this._pushToDependenciesIfNotExist(currentComponentDeps, fileType, depData);
      }
    });
  }
  _pushToDependenciesIfNotExist(dependency, fileType, depDebug) {
    const existingDependency = this.getExistingDependency(this.allDependencies.dependencies, dependency.id);
    const existingDevDependency = this.getExistingDependency(this.allDependencies.devDependencies, dependency.id);
    // no need to enter dev dependency to devDependencies if it exists already in dependencies
    if (existingDependency || existingDevDependency && fileType.isTestFile) {
      return;
    }
    // at this point, either, it doesn't exist at all and should be entered.
    // or it exists in devDependencies but now it comes from non-dev file, which should be entered
    // as non-dev.
    this.pushToDependenciesArray(dependency, fileType, depDebug);
  }
  pushToDependenciesArray(currentComponentsDeps, fileType, depDebug) {
    if (fileType.isTestFile) {
      this.allDependencies.devDependencies.push(currentComponentsDeps);
    } else {
      this.allDependencies.dependencies.push(currentComponentsDeps);
    }
    this.debugDependenciesData.components.push(depDebug);
  }

  /**
   * Remove the dependencies which appear both in dev and regular deps from the dev
   * Because if a dependency is both dev dependency and regular dependency it should be treated as regular one
   * Apply for both packages and components dependencies
   */
  removeDevAndEnvDepsIfTheyAlsoRegulars() {
    // remove dev and env packages that are also regular packages
    const getNotRegularPackages = packages => (0, _difference2().default)((0, _keys2().default)(packages), (0, _keys2().default)(this.allPackagesDependencies.packageDependencies));
    this.allPackagesDependencies.devPackageDependencies = (0, _pick2().default)(getNotRegularPackages(this.allPackagesDependencies.devPackageDependencies), this.allPackagesDependencies.devPackageDependencies);
    // remove dev dependencies that are also regular dependencies
    // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
    const componentDepsIds = new (_bitId().BitIds)(...this.allDependencies.dependencies.map(c => c.id));
    this.allDependencies.devDependencies = this.allDependencies.devDependencies.filter(d => !componentDepsIds.has(d.id));
  }

  /**
   * given missing packages name, find whether they were dependencies with custom-resolve before.
   */
  findOriginallyCustomModuleResolvedDependencies(packages) {
    if (!packages) return undefined;
    if (!this.componentFromModel) return undefined; // not relevant, the component is not imported
    const dependencies = new (_dependencies().default)(this.componentFromModel.getAllDependencies());
    if (dependencies.isEmpty()) return undefined;
    const importSourceMap = dependencies.getCustomResolvedData();
    if ((0, _isEmpty2().default)(importSourceMap)) return undefined;
    const foundPackages = packages.reduce((acc, value) => {
      const packageLinuxFormat = (0, _utils().pathNormalizeToLinux)(value);
      const foundImportSource = Object.keys(importSourceMap).find(importSource => importSource.startsWith(packageLinuxFormat));
      if (foundImportSource) {
        const dependencyId = importSourceMap[foundImportSource];
        acc[value] = dependencyId;
      }
      return acc;
    }, {});
    return (0, _isEmpty2().default)(foundPackages) ? undefined : foundPackages;
  }
  getExistingDependency(dependencies, id) {
    return dependencies.find(d => d.id.isEqualWithoutVersion(id));
  }
  getExistingDepRelativePaths(dependency, relativePath) {
    if (!dependency.relativePaths || (0, _isEmpty2().default)(dependency.relativePaths)) return null;
    return dependency.relativePaths.find(paths => paths.sourceRelativePath === relativePath.sourceRelativePath && paths.destinationRelativePath === relativePath.destinationRelativePath);
  }
  getDiffSpecifiers(originSpecifiers, targetSpecifiers) {
    const cmp = (specifier1, specifier2) => specifier1.mainFile.name === specifier2.mainFile.name;
    return (0, _differenceWith2().default)(cmp, targetSpecifiers, originSpecifiers);
  }
  applyPeersFromComponentModel() {
    const getPeerDependencies = () => {
      const packageJson = this._getPackageJsonFromComponentModel();
      if (packageJson && packageJson.peerDependencies) return packageJson.peerDependencies;
      return {};
    };
    const projectPeerDependencies = getPeerDependencies();
    const peerPackages = {};
    if ((0, _isEmpty2().default)(projectPeerDependencies)) return;

    // check whether the peer-dependencies was actually require in the code. if so, remove it from
    // the packages/dev-packages and add it as a peer-package.
    // if it was not required in the code, don't add it to the peerPackages
    Object.keys(projectPeerDependencies).forEach(pkg => {
      if (this.overridesDependencies.shouldIgnorePeerPackage(pkg)) return;
      ['packageDependencies', 'devPackageDependencies'].forEach(field => {
        if (Object.keys(this.allPackagesDependencies[field]).includes(pkg)) {
          delete this.allPackagesDependencies[field][pkg];
          peerPackages[pkg] = projectPeerDependencies[pkg];
        }
      });
    });
    this.allPackagesDependencies.peerPackageDependencies = peerPackages;
  }
  applyPackageJson() {
    const packageJson = this._getPackageJson();
    if (!packageJson) return;
    const pkgJsonPeer = packageJson.peerDependencies || {};
    const pkgJsonRegularDeps = packageJson.dependencies || {};
    const peerDeps = this.allPackagesDependencies.peerPackageDependencies || {};
    ['packageDependencies', 'devPackageDependencies', 'peerPackageDependencies'].forEach(field => {
      (0, _forEachObjIndexed2().default)((_pkgVal, pkgName) => {
        const peerVersionFromPkgJson = pkgJsonPeer[pkgName];
        const regularVersionFromPkgJson = pkgJsonRegularDeps[pkgName];
        if (peerVersionFromPkgJson) {
          delete this.allPackagesDependencies[field][pkgName];
          peerDeps[pkgName] = peerVersionFromPkgJson;
        } else if (regularVersionFromPkgJson) {
          var _this$allPackagesDepe;
          (_this$allPackagesDepe = this.allPackagesDependencies.peerPackageDependencies) === null || _this$allPackagesDepe === void 0 ? true : delete _this$allPackagesDepe[pkgName];
          this.allPackagesDependencies[field][pkgName] = regularVersionFromPkgJson;
        }
      }, this.allPackagesDependencies[field]);
    });
    this.allPackagesDependencies.peerPackageDependencies = peerDeps;
  }
  applyWorkspacePolicy() {
    const wsPolicy = DependencyResolver.getWorkspacePolicy();
    if (!wsPolicy) return;
    const wsPeer = wsPolicy.peerDependencies || {};
    const wsRegular = wsPolicy.dependencies || {};
    const peerDeps = this.allPackagesDependencies.peerPackageDependencies || {};
    // we are not iterate component deps since they are resolved from what actually installed
    // the policy used for installation only in that case
    ['packageDependencies', 'devPackageDependencies', 'peerPackageDependencies'].forEach(field => {
      (0, _forEachObjIndexed2().default)((_pkgVal, pkgName) => {
        const peerVersionFromWsPolicy = wsPeer[pkgName];
        const regularVersionFromWsPolicy = wsRegular[pkgName];
        if (peerVersionFromWsPolicy) {
          delete this.allPackagesDependencies[field][pkgName];
          peerDeps[pkgName] = peerVersionFromWsPolicy;
        } else if (regularVersionFromWsPolicy) {
          var _this$allPackagesDepe2;
          (_this$allPackagesDepe2 = this.allPackagesDependencies.peerPackageDependencies) === null || _this$allPackagesDepe2 === void 0 ? true : delete _this$allPackagesDepe2[pkgName];
          this.allPackagesDependencies[field][pkgName] = regularVersionFromWsPolicy;
        }
      }, this.allPackagesDependencies[field]);
    });
    this.allPackagesDependencies.peerPackageDependencies = peerDeps;
  }
  async applyAutoDetectedPeersFromEnvOnComponent() {
    var _this$issues$getIssue;
    const envPolicy = await DependencyResolver.getHarmonyEnvPeersPolicyForComponent(this.component.extensions);
    if (!envPolicy || !Object.keys(envPolicy).length) {
      return;
    }
    const peerDeps = this.allPackagesDependencies.peerPackageDependencies || {};
    // we are not iterate component deps since they are resolved from what actually installed
    // the policy used for installation only in that case
    ['packageDependencies', 'devPackageDependencies', 'peerPackageDependencies'].forEach(field => {
      (0, _forEachObjIndexed2().default)((_pkgVal, pkgName) => {
        const peerVersionFromEnvPolicy = envPolicy[pkgName];
        if (this.overridesDependencies.shouldIgnorePeerPackage(pkgName)) return;
        if (peerVersionFromEnvPolicy) {
          delete this.allPackagesDependencies[field][pkgName];
          peerDeps[pkgName] = peerVersionFromEnvPolicy;
        }
      }, this.allPackagesDependencies[field]);
    });
    // We want to also add missing packages to the peer list as we know to resolve the version from the env anyway
    // @ts-ignore
    const missingsData = (_this$issues$getIssue = this.issues.getIssueByName('MissingPackagesDependenciesOnFs')) === null || _this$issues$getIssue === void 0 ? void 0 : _this$issues$getIssue.data;
    if (missingsData) {
      // @ts-ignore
      const missingPackages = (0, _lodash().union)(...(Object.values(missingsData) || []));
      missingPackages.forEach(pkgName => {
        const peerVersionFromEnvPolicy = envPolicy[pkgName];
        if (peerVersionFromEnvPolicy) {
          peerDeps[pkgName] = peerVersionFromEnvPolicy;
        }
      });
    }

    // TODO: handle component deps once we support peers between components
    this.allPackagesDependencies.peerPackageDependencies = peerDeps;
  }
  async applyAutoDetectedPeersFromEnvOnEnvItSelf() {
    const envPolicy = await DependencyResolver.getHarmonyEnvPeersPolicyForEnvItself(this.component.id);
    if (!envPolicy || !Object.keys(envPolicy).length) {
      return;
    }
    const deps = this.allPackagesDependencies.packageDependencies || {};
    // we are not iterate component deps since they are resolved from what actually installed
    // the policy used for installation only in that case
    ['packageDependencies', 'devPackageDependencies', 'peerPackageDependencies'].forEach(field => {
      (0, _forEachObjIndexed2().default)((_pkgVal, pkgName) => {
        const peerVersionFromEnvPolicy = envPolicy[pkgName];
        if (this.overridesDependencies.shouldIgnorePackageByType(pkgName, 'dependencies')) return;
        if (peerVersionFromEnvPolicy) {
          delete this.allPackagesDependencies[field][pkgName];
        }
      }, this.allPackagesDependencies[field]);
    });
    Object.assign(deps, envPolicy);
    // TODO: handle component deps once we support peers between components
    this.allPackagesDependencies.packageDependencies = deps;
  }

  /**
   * returns `package.json` of the component when it's imported, or `package.json` of the workspace
   * when it's authored.
   */
  _getPackageJson() {
    return this.consumer.packageJson.packageJsonObject;
  }
  _getPackageJsonFromComponentModel() {
    if (this.componentFromModel && this.component.componentMap) {
      // a component is imported but the package.json file is missing or never written
      // read the values from the model
      const packageJson = _packageJsonFile().default.createFromComponent(this.component.componentMap.rootDir, this.componentFromModel);
      return packageJson.packageJsonObject;
    }
    return undefined;
  }
  setLegacyInsideHarmonyIssue() {
    if (this.componentFromModel && this.componentFromModel.isLegacy) {
      this.issues.getOrCreate(_componentIssues().IssuesClasses.LegacyInsideHarmony).data = true;
    }
  }

  /**
   * when requiring packages in typescript, sometimes there are the types packages with the same
   * name, which the user probably wants as well. for example, requiring `foo` package, will also
   * add `@types/foo` to the devDependencies if it has been found in the user `package.json` file.
   *
   * ideally this should be in bit-javascript. however, the decision where to put these `@types`
   * packages (dependencies/devDependencies) is done here according to the user `package.json`
   * and can't be done there because the `Tree` we get from bit-javascript doesn't have this
   * distinction.
   */
  _addTypesPackagesForTypeScript(packages, originFile) {
    const isTypeScript = (0, _utils().getExt)(originFile) === 'ts' || (0, _utils().getExt)(originFile) === 'tsx';
    if (!isTypeScript) return;
    const depsHost = DependencyResolver.getWorkspacePolicy();
    const addFromConfig = packageName => {
      if (!depsHost) return false;
      return _constants().DEPENDENCIES_FIELDS.some(depField => {
        if (!depsHost[depField]) return false;
        const typesPackage = (0, _packageToDefinetlyTyped().packageToDefinetlyTyped)(packageName);
        if (!depsHost[depField][typesPackage]) return false;
        Object.assign(this.allPackagesDependencies.devPackageDependencies, {
          [typesPackage]: depsHost[depField][typesPackage]
        });
        return true;
      });
    };
    const addFromModel = packageName => {
      if (!this.componentFromModel) return;
      const typesPackage = (0, _packageToDefinetlyTyped().packageToDefinetlyTyped)(packageName);
      const typedPackageFromModel = this.componentFromModel.devPackageDependencies[typesPackage];
      if (!typedPackageFromModel) return;
      Object.assign(this.allPackagesDependencies.devPackageDependencies, {
        [typesPackage]: typedPackageFromModel
      });
    };
    Object.keys(packages).forEach(packageName => {
      const added = addFromConfig(packageName);
      if (!added) addFromModel(packageName);
    });
  }
  _pkgFieldMapping(field) {
    switch (field) {
      case 'dependencies':
        return 'packageDependencies';
      case 'devDependencies':
        return 'devPackageDependencies';
      case 'peerDependencies':
        return 'peerPackageDependencies';
      default:
        throw new Error(`${field} is not recognized`);
    }
  }
  _pushToUntrackDependenciesIssues(originFile, depFileRelative, nested = false) {
    const findExisting = () => {
      var _this$issues$getIssue2;
      let result;
      (0, _forEachObjIndexed2().default)(currentUntracked => {
        const found = currentUntracked.untrackedFiles.find(file => {
          return file.relativePath === depFileRelative;
        });
        if (found) {
          result = found;
        }
      }, ((_this$issues$getIssue2 = this.issues.getIssue(_componentIssues().IssuesClasses.UntrackedDependencies)) === null || _this$issues$getIssue2 === void 0 ? void 0 : _this$issues$getIssue2.data) || {});
      return result;
    };
    const existing = findExisting();
    const newUntrackedFile = {
      relativePath: depFileRelative,
      existing: false
    };
    // If it's already found mark them both as existing
    if (existing) {
      newUntrackedFile.existing = true;
      existing.existing = true;
    }
    const untrackIssue = this.issues.getOrCreate(_componentIssues().IssuesClasses.UntrackedDependencies);
    const untrackedCurrentFile = untrackIssue === null || untrackIssue === void 0 ? void 0 : untrackIssue.data[originFile];
    if (untrackedCurrentFile) {
      untrackedCurrentFile.untrackedFiles.push(newUntrackedFile);
    } else {
      untrackIssue.data[originFile] = {
        nested,
        untrackedFiles: [newUntrackedFile]
      };
    }
  }
  _pushToRelativeComponentsIssues(originFile, componentId) {
    var _this$issues$getOrCre2;
    ((_this$issues$getOrCre2 = this.issues.getOrCreate(_componentIssues().IssuesClasses.RelativeComponents).data)[originFile] || (_this$issues$getOrCre2[originFile] = [])).push(componentId);
  }
  _pushToRelativeComponentsAuthoredIssues(originFile, componentId, importSource, relativePath) {
    var _this$issues$getOrCre3;
    ((_this$issues$getOrCre3 = this.issues.getOrCreate(_componentIssues().IssuesClasses.RelativeComponentsAuthored).data)[originFile] || (_this$issues$getOrCre3[originFile] = [])).push({
      importSource,
      componentId,
      relativePath
    });
  }
  _pushToMissingDependenciesOnFs(originFile, missingFiles) {
    var _this$issues$getOrCre4;
    ((_this$issues$getOrCre4 = this.issues.getOrCreate(_componentIssues().IssuesClasses.MissingDependenciesOnFs).data)[originFile] || (_this$issues$getOrCre4[originFile] = [])).push(...missingFiles);
  }
  _pushToMissingPackagesDependenciesIssues(originFile, missingPackages) {
    var _this$issues$getOrCre5;
    ((_this$issues$getOrCre5 = this.issues.getOrCreate(_componentIssues().IssuesClasses.MissingPackagesDependenciesOnFs).data)[originFile] || (_this$issues$getOrCre5[originFile] = [])).push(...(0, _lodash().uniq)(missingPackages));
  }
  _pushToMissingComponentsIssues(originFile, componentId) {
    var _this$issues$getOrCre6;
    ((_this$issues$getOrCre6 = this.issues.getOrCreate(_componentIssues().IssuesClasses.MissingComponents).data)[originFile] || (_this$issues$getOrCre6[originFile] = [])).push(componentId);
  }
}
exports.default = DependencyResolver;
(0, _defineProperty2().default)(DependencyResolver, "getWorkspacePolicy", void 0);
(0, _defineProperty2().default)(DependencyResolver, "getHarmonyEnvPeersPolicyForComponent", void 0);
(0, _defineProperty2().default)(DependencyResolver, "getHarmonyEnvPeersPolicyForEnvItself", void 0);
(0, _defineProperty2().default)(DependencyResolver, "getDepResolverAspectName", void 0);
(0, _defineProperty2().default)(DependencyResolver, "getCoreAspectsPackagesAndIds", void 0);
(0, _defineProperty2().default)(DependencyResolver, "getDevFiles", void 0);
