// @teambit/legacy/dist/cache/lru-cache-adapter.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.array.iterator.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LRUCacheAdapter = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _lruCache() {
  const data = _interopRequireDefault(require("lru-cache"));
  _lruCache = function () {
    return data;
  };
  return data;
}
class LRUCacheAdapter {
  constructor(options) {
    (0, _defineProperty2().default)(this, "cache", void 0);
    const lruOptions = {
      max: options.maxSize || Infinity
    };
    if (options.maxAge) lruOptions.maxAge = options.maxAge;
    this.cache = new (_lruCache().default)(lruOptions);
  }
  set(key, value) {
    this.cache.set(key, value);
  }
  get(key) {
    return this.cache.get(key);
  }
  delete(key) {
    this.cache.del(key);
  }
  has(key) {
    return this.cache.has(key);
  }
  deleteAll() {
    this.cache.reset();
  }
  keys() {
    return this.cache.keys();
  }
}
exports.LRUCacheAdapter = LRUCacheAdapter;
