// @teambit/bundler/dist/events/components-server-started-event.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ComponentsServerStartedEvent = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _pubsub() {
  const data = require("@teambit/pubsub");
  _pubsub = function () {
    return data;
  };
  return data;
}
/* eslint-disable max-classes-per-file */

class ComponentsServerStartedEventData {
  constructor(componentsServer, context, hostname, port) {
    this.componentsServer = componentsServer;
    this.context = context;
    this.hostname = hostname;
    this.port = port;
  }
}
class ComponentsServerStartedEvent extends _pubsub().BitBaseEvent {
  constructor(timestamp, componentsServer, context, hostname, port) {
    super(ComponentsServerStartedEvent.TYPE, '0.0.1', timestamp, new ComponentsServerStartedEventData(componentsServer, context, hostname, port));
    this.timestamp = timestamp;
    this.componentsServer = componentsServer;
    this.context = context;
    this.hostname = hostname;
    this.port = port;
  }
}
exports.ComponentsServerStartedEvent = ComponentsServerStartedEvent;
(0, _defineProperty2().default)(ComponentsServerStartedEvent, "TYPE", 'components-server-started');
