// @teambit/export/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ExportAspect", {
  enumerable: true,
  get: function () {
    return _export().ExportAspect;
  }
});
Object.defineProperty(exports, "ExportPersist", {
  enumerable: true,
  get: function () {
    return _actions().ExportPersist;
  }
});
Object.defineProperty(exports, "ExportValidate", {
  enumerable: true,
  get: function () {
    return _actions().ExportValidate;
  }
});
Object.defineProperty(exports, "FetchMissingDeps", {
  enumerable: true,
  get: function () {
    return _actions().FetchMissingDeps;
  }
});
Object.defineProperty(exports, "Network", {
  enumerable: true,
  get: function () {
    return _network().Network;
  }
});
Object.defineProperty(exports, "ObjectList", {
  enumerable: true,
  get: function () {
    return _objectList().ObjectList;
  }
});
Object.defineProperty(exports, "PushOptions", {
  enumerable: true,
  get: function () {
    return _put().PushOptions;
  }
});
Object.defineProperty(exports, "RemovePendingDir", {
  enumerable: true,
  get: function () {
    return _actions().RemovePendingDir;
  }
});
function _export() {
  const data = require("./export.aspect");
  _export = function () {
    return data;
  };
  return data;
}
function _network() {
  const data = require("@teambit/legacy/dist/scope/network/network");
  _network = function () {
    return data;
  };
  return data;
}
function _put() {
  const data = require("@teambit/legacy/dist/api/scope/lib/put");
  _put = function () {
    return data;
  };
  return data;
}
function _actions() {
  const data = require("@teambit/legacy/dist/scope/actions");
  _actions = function () {
    return data;
  };
  return data;
}
function _objectList() {
  const data = require("@teambit/legacy/dist/scope/objects/object-list");
  _objectList = function () {
    return data;
  };
  return data;
}
