// @teambit/legacy/dist/cli/default-error-handler.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.array.iterator.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
exports.findErrorDefinition = findErrorDefinition;
exports.sendToAnalyticsAndSentry = sendToAnalyticsAndSentry;
function _chalk() {
  const data = _interopRequireDefault(require("chalk"));
  _chalk = function () {
    return data;
  };
  return data;
}
function _bitError() {
  const data = require("@teambit/bit-error");
  _bitError = function () {
    return data;
  };
  return data;
}
function _analytics() {
  const data = require("../analytics/analytics");
  _analytics = function () {
    return data;
  };
  return data;
}
function _configKeyNotFound() {
  const data = _interopRequireDefault(require("../api/consumer/lib/exceptions/config-key-not-found"));
  _configKeyNotFound = function () {
    return data;
  };
  return data;
}
function _diagnosisNotFound() {
  const data = _interopRequireDefault(require("../api/consumer/lib/exceptions/diagnosis-not-found"));
  _diagnosisNotFound = function () {
    return data;
  };
  return data;
}
function _idExportedAlready() {
  const data = _interopRequireDefault(require("../api/consumer/lib/exceptions/id-exported-already"));
  _idExportedAlready = function () {
    return data;
  };
  return data;
}
function _invalidVersion() {
  const data = _interopRequireDefault(require("../api/consumer/lib/exceptions/invalid-version"));
  _invalidVersion = function () {
    return data;
  };
  return data;
}
function _missingDiagnosisName() {
  const data = _interopRequireDefault(require("../api/consumer/lib/exceptions/missing-diagnosis-name"));
  _missingDiagnosisName = function () {
    return data;
  };
  return data;
}
function _noIdMatchWildcard() {
  const data = _interopRequireDefault(require("../api/consumer/lib/exceptions/no-id-match-wildcard"));
  _noIdMatchWildcard = function () {
    return data;
  };
  return data;
}
function _nothingToCompareTo() {
  const data = _interopRequireDefault(require("../api/consumer/lib/exceptions/nothing-to-compare-to"));
  _nothingToCompareTo = function () {
    return data;
  };
  return data;
}
function _objectsWithoutConsumer() {
  const data = _interopRequireDefault(require("../api/consumer/lib/exceptions/objects-without-consumer"));
  _objectsWithoutConsumer = function () {
    return data;
  };
  return data;
}
function _constants() {
  const data = require("../constants");
  _constants = function () {
    return data;
  };
  return data;
}
function _exceptions() {
  const data = require("../consumer/bit-map/exceptions");
  _exceptions = function () {
    return data;
  };
  return data;
}
function _outsideRootDir() {
  const data = _interopRequireDefault(require("../consumer/bit-map/exceptions/outside-root-dir"));
  _outsideRootDir = function () {
    return data;
  };
  return data;
}
function _exceptions2() {
  const data = require("../consumer/component-ops/add-components/exceptions");
  _exceptions2 = function () {
    return data;
  };
  return data;
}
function _addingIndividualFiles() {
  const data = require("../consumer/component-ops/add-components/exceptions/adding-individual-files");
  _addingIndividualFiles = function () {
    return data;
  };
  return data;
}
function _componentsPendingImport() {
  const data = _interopRequireDefault(require("../consumer/component-ops/exceptions/components-pending-import"));
  _componentsPendingImport = function () {
    return data;
  };
  return data;
}
function _componentsPendingMerge() {
  const data = _interopRequireDefault(require("../consumer/component-ops/exceptions/components-pending-merge"));
  _componentsPendingMerge = function () {
    return data;
  };
  return data;
}
function _ejectNoDir() {
  const data = _interopRequireDefault(require("../consumer/component-ops/exceptions/eject-no-dir"));
  _ejectNoDir = function () {
    return data;
  };
  return data;
}
function _componentNotFoundInPath() {
  const data = _interopRequireDefault(require("../consumer/component/exceptions/component-not-found-in-path"));
  _componentNotFoundInPath = function () {
    return data;
  };
  return data;
}
function _externalBuildErrors() {
  const data = _interopRequireDefault(require("../consumer/component/exceptions/external-build-errors"));
  _externalBuildErrors = function () {
    return data;
  };
  return data;
}
function _externalTestErrors() {
  const data = _interopRequireDefault(require("../consumer/component/exceptions/external-test-errors"));
  _externalTestErrors = function () {
    return data;
  };
  return data;
}
function _fileSourceNotFound() {
  const data = _interopRequireDefault(require("../consumer/component/exceptions/file-source-not-found"));
  _fileSourceNotFound = function () {
    return data;
  };
  return data;
}
function _injectNonEjected() {
  const data = _interopRequireDefault(require("../consumer/component/exceptions/inject-non-ejected"));
  _injectNonEjected = function () {
    return data;
  };
  return data;
}
function _mainFileRemoved() {
  const data = _interopRequireDefault(require("../consumer/component/exceptions/main-file-removed"));
  _mainFileRemoved = function () {
    return data;
  };
  return data;
}
function _missingFilesFromComponent() {
  const data = _interopRequireDefault(require("../consumer/component/exceptions/missing-files-from-component"));
  _missingFilesFromComponent = function () {
    return data;
  };
  return data;
}
function _noComponentDir() {
  const data = require("../consumer/component/exceptions/no-component-dir");
  _noComponentDir = function () {
    return data;
  };
  return data;
}
function _invalidBitJson() {
  const data = _interopRequireDefault(require("../consumer/config/exceptions/invalid-bit-json"));
  _invalidBitJson = function () {
    return data;
  };
  return data;
}
function _invalidConfigPropPath() {
  const data = _interopRequireDefault(require("../consumer/config/exceptions/invalid-config-prop-path"));
  _invalidConfigPropPath = function () {
    return data;
  };
  return data;
}
function _invalidPackageJson() {
  const data = _interopRequireDefault(require("../consumer/config/exceptions/invalid-package-json"));
  _invalidPackageJson = function () {
    return data;
  };
  return data;
}
function _invalidPackageManager() {
  const data = _interopRequireDefault(require("../consumer/config/exceptions/invalid-package-manager"));
  _invalidPackageManager = function () {
    return data;
  };
  return data;
}
function _exceptions3() {
  const data = require("../consumer/exceptions");
  _exceptions3 = function () {
    return data;
  };
  return data;
}
function _lanesIsDisabled() {
  const data = require("../consumer/lanes/exceptions/lanes-is-disabled");
  _lanesIsDisabled = function () {
    return data;
  };
  return data;
}
function _exceptions4() {
  const data = require("../consumer/login/exceptions");
  _exceptions4 = function () {
    return data;
  };
  return data;
}
function _generalError() {
  const data = _interopRequireDefault(require("../error/general-error"));
  _generalError = function () {
    return data;
  };
  return data;
}
function _hashErrorObject() {
  const data = _interopRequireDefault(require("../error/hash-error-object"));
  _hashErrorObject = function () {
    return data;
  };
  return data;
}
function _validationError() {
  const data = _interopRequireDefault(require("../error/validation-error"));
  _validationError = function () {
    return data;
  };
  return data;
}
function _promptCanceled() {
  const data = _interopRequireDefault(require("../prompts/exceptions/prompt-canceled"));
  _promptCanceled = function () {
    return data;
  };
  return data;
}
function _remoteNotFound() {
  const data = _interopRequireDefault(require("../remotes/exceptions/remote-not-found"));
  _remoteNotFound = function () {
    return data;
  };
  return data;
}
function _exceptions5() {
  const data = require("../scope/exceptions");
  _exceptions5 = function () {
    return data;
  };
  return data;
}
function _exceptions6() {
  const data = require("../scope/network/exceptions");
  _exceptions6 = function () {
    return data;
  };
  return data;
}
function _exportAnotherOwnerPrivate() {
  const data = _interopRequireDefault(require("../scope/network/exceptions/export-another-owner-private"));
  _exportAnotherOwnerPrivate = function () {
    return data;
  };
  return data;
}
function _remoteResolverError() {
  const data = _interopRequireDefault(require("../scope/network/exceptions/remote-resolver-error"));
  _remoteResolverError = function () {
    return data;
  };
  return data;
}
function _gitNotFound() {
  const data = _interopRequireDefault(require("../utils/git/exceptions/git-not-found"));
  _gitNotFound = function () {
    return data;
  };
  return data;
}
function _remoteUndefined() {
  const data = _interopRequireDefault(require("./commands/exceptions/remote-undefined"));
  _remoteUndefined = function () {
    return data;
  };
  return data;
}
function _newerVersionTemplate() {
  const data = _interopRequireDefault(require("./templates/newer-version-template"));
  _newerVersionTemplate = function () {
    return data;
  };
  return data;
}
// all errors that the command does not handle comes to this switch statement
// if you handle the error, then return true

const reportIssueToGithubMsg = 'This error should have never happened. Please report this issue on Github https://github.com/teambit/bit/issues';

// @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
// @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
const errorsMap = [[_remoteUndefined().default, () => _chalk().default.red('error: remote url must be defined. please use: `ssh://`, `file://` or `bit://` protocols to define remote access')], [_exceptions3().ConsumerAlreadyExists, () => 'workspace already exists'], [_generalError().default, err => `${err.msg}`], [_exceptions5().VersionAlreadyExists, err => `error: version ${err.version} already exists for ${err.componentId}`], [_exceptions3().ConsumerNotFound, () => 'workspace not found. to initiate a new workspace, please use `bit init`'], [_exceptions3().LoginFailed, () => 'error: there was a problem with web authentication'],
// [
//   PluginNotFound,
//   err => `error: The compiler "${err.plugin}" is not installed, please use "bit install ${err.plugin}" to install it.`
// ],
[_fileSourceNotFound().default, err => `file or directory "${err.path}" was not found`], [_lanesIsDisabled().LanesIsDisabled, () => `lanes/snaps features are disabled. upgrade your workspace to Harmony to enable them`], [_outsideRootDir().default, err => `unable to add file ${err.filePath} because it's located outside the component root dir ${err.rootDir}`], [_addingIndividualFiles().AddingIndividualFiles, err => `error: adding individual files is blocked ("${err.file}"), and only directories can be added`], [_exceptions6().ProtocolNotSupported, () => 'error: remote scope protocol is not supported, please use: `ssh://`, `file://`, `http://`, `https://` or `bit://`'], [_exceptions6().RemoteScopeNotFound, err => `error: remote scope "${_chalk().default.bold(err.name)}" was not found.`], [_injectNonEjected().default, () => 'error: could not inject config for already injected component'], [_componentsPendingImport().default, () => _constants().IMPORT_PENDING_MSG],
// TODO: improve error
[_componentsPendingMerge().default, err => {
  const componentsStr = err.divergeData.map(d => `${_chalk().default.bold(d.id)} has ${_chalk().default.bold(d.snapsLocal)} snaps locally only and ${_chalk().default.bold(d.snapsRemote)} snaps remotely only`).join('\n');
  return `the local and remote history of the following component(s) have diverged\n${componentsStr}\nPlease use --merge to merge them`;
}], [_ejectNoDir().default, err => `error: could not eject config for ${_chalk().default.bold(err.compId)}, please make sure it's under a track directory`], [_exceptions5().ComponentNotFound, err => {
  return err.dependentId ? `error: the component dependency "${_chalk().default.bold(err.id)}" required by "${_chalk().default.bold(err.dependentId)}" was not found` : `error: component "${_chalk().default.bold(err.id)}" was not found`;
}], [_exceptions5().CorruptedComponent, err => `error: the model representation of "${_chalk().default.bold(err.id)}" is corrupted, the object of version ${err.version} is missing.\n${reportIssueToGithubMsg}`], [_validationError().default, err => `${err.message}\n${reportIssueToGithubMsg}`], [_componentNotFoundInPath().default, err => `error: component in path "${_chalk().default.bold(err.path)}" was not found`], [_remoteNotFound().default, err => `error: remote "${_chalk().default.bold(err.name)}" was not found`], [_exceptions6().NetworkError, err => `error: remote failed with error the following error:\n "${_chalk().default.bold(err.remoteErr)}"`], [_exceptions5().HashMismatch, err => `found hash mismatch of ${_chalk().default.bold(err.id)}, version ${_chalk().default.bold(err.version)}.
  originalHash: ${_chalk().default.bold(err.originalHash)}.
  currentHash: ${_chalk().default.bold(err.currentHash)}
  this usually happens when a component is old and the migration script was not running or interrupted`], [_exceptions5().HashNotFound, err => `hash ${_chalk().default.bold(err.hash)} not found`], [_exceptions5().HeadNotFound, err => `head snap ${_chalk().default.bold(err.headHash)} was not found for a component ${_chalk().default.bold(err.id)}`], [_exceptions5().OutdatedIndexJson, err => `error: ${_chalk().default.bold(err.id)} found in the index.json file, however, is missing from the scope.
the cache is deleted and will be rebuilt on the next command. please re-run the command.`], [_exceptions5().CyclicDependencies, err => `${err.msg.toString().toLocaleLowerCase()}`], [_exceptions6().UnexpectedNetworkError, err => `unexpected network error has occurred.
${err.message ? `server responded with: "${err.message}"` : ''}`], [_remoteResolverError().default, err => `error: ${err.message ? `${err.message}` : 'unexpected remote resolver error has occurred'}`], [_exportAnotherOwnerPrivate().default, err => `error: unable to export components to ${err.destinationScope} because they have dependencies on components in ${err.sourceScope}.
bit does not allow setting dependencies between components in private collections managed by different owners.

see troubleshooting at https://${_constants().BASE_DOCS_DOMAIN}/docs/bit-dev#permissions-for-collections`], [_exceptions6().SSHInvalidResponse, () => `error: received an invalid response from the remote SSH server.
to see the invalid response, have a look at the log, located at ${_constants().DEBUG_LOG}`], [_exceptions5().InvalidIndexJson, err => `fatal: your .bit/index.json is not a valid JSON file.
To rebuild the file, please run ${_chalk().default.bold('bit init --reset')}.
Original Error: ${err.message}`], [_exceptions5().ScopeNotFound, err => `error: scope not found at ${_chalk().default.bold(err.scopePath)}`], [_exceptions5().ScopeJsonNotFound, err => `error: scope.json file was not found at ${_chalk().default.bold(err.path)}, please use ${_chalk().default.bold('bit init')} to recreate the file`], [_missingDiagnosisName().default, () => 'error: please provide a diagnosis name'], [_diagnosisNotFound().default, err => `error: diagnosis ${_chalk().default.bold(err.diagnosisName)} not found`], [_exceptions3().ComponentOutOfSync, err => `component ${_chalk().default.bold(err.id)} is not in-sync between the consumer and the scope.
if it is originated from another git branch, go back to that branch to continue working on the component.
if possible, remove the component using "bit remove" and re-import or re-create it.
to re-start Bit from scratch, deleting all objects from the scope, use "bit init --reset-hard"`], [_exceptions3().NothingToImport, () => _chalk().default.yellow('nothing to import. please use `bit import [component_id]`')], [_invalidBitJson().default, err => `error: invalid bit.json: ${_chalk().default.bold(err.path)} is not a valid JSON file.
consider running ${_chalk().default.bold('bit init --reset')} to recreate the file`], [_invalidPackageManager().default, err => `error: the package manager provided ${_chalk().default.bold(err.packageManager)} is not a valid package manager.
please specify 'npm' or 'yarn'`], [_invalidPackageJson().default, err => `error: package.json at ${_chalk().default.bold(err.path)} is not a valid JSON file.
please fix the file in order to run bit commands`], [_invalidConfigPropPath().default, err => `error: the path "${_chalk().default.bold(err.fieldValue)}" of "${_chalk().default.bold(err.fieldName)}" in your bit.json or package.json file is invalid.
please make sure it's not absolute and doesn't contain invalid characters`], [_exceptions().MissingMainFile, err => `error: the component ${_chalk().default.bold(err.componentId)} does not contain a main file.\nplease either use --id to group all added files as one component or use our DSL to define the main file dynamically.\nsee troubleshooting at https://${_constants().BASE_DOCS_DOMAIN}/components/component-main-file`], [_noComponentDir().NoComponentDir, err => `"${err.id}" doesn't have a component directory, which is invalid.
please run "bit status" to get more info.\nLearn more at https://${_constants().BASE_DOCS_DOMAIN}/workspace/component-directory`], [_exceptions2().MissingMainFileMultipleComponents, err => `error: the components ${_chalk().default.bold(err.componentIds.join(', '))} does not contain a main file.\nplease either use --id to group all added files as one component or use our DSL to define the main file dynamically.\nsee troubleshooting at https://${_constants().BASE_DOCS_DOMAIN}/components/component-main-file`], [_exceptions().InvalidBitMap, err => `error: unable to parse your bitMap file at ${_chalk().default.bold(err.path)}, due to an error ${_chalk().default.bold(err.errorMessage)}.
      consider running ${_chalk().default.bold('bit init --reset')} to recreate the file`], [_exceptions2().ExcludedMainFile, err => `error: main file ${_chalk().default.bold(err.mainFile)} was excluded from file list`], [_mainFileRemoved().default, err => `error: main file ${_chalk().default.bold(err.mainFile)} was removed from ${_chalk().default.bold(err.id)}.
please use "bit remove" to delete the component or "bit add" with "--main" and "--id" flags to add a new mainFile`], [_exceptions2().VersionShouldBeRemoved, err => `please remove the version part from the specified id ${_chalk().default.bold(err.id)} and try again`], [_exceptions2().MainFileIsDir, err => `error: the specified main path ${_chalk().default.bold(err.mainFile)} is a directory, please specify a file or a pattern DSL`], [_missingFilesFromComponent().default, err => {
  return `component ${err.id} is invalid as part or all of the component files were deleted. please use 'bit remove' to resolve the issue`;
}], [_exceptions2().PathsNotExist, err => `error: file or directory "${_chalk().default.bold(err.paths.join(', '))}" was not found.`], [_exceptions2().PathOutsideConsumer, err => `error: file or directory "${_chalk().default.bold(err.path)}" is located outside of the workspace.`], [_configKeyNotFound().default, err => `unable to find a key "${_chalk().default.bold(err.key)}" in your bit config`], [_exceptions4().WriteToNpmrcError, err => `unable to add @bit as a scoped registry at "${_chalk().default.bold(err.path)}"`], [_exceptions4().PathToNpmrcNotExist, err => `error: file or directory "${_chalk().default.bold(err.path)}" was not found.`], [_exceptions5().ParentNotFound, err => `component ${_chalk().default.bold(err.id)} missing data. parent ${err.parentHash} of version ${err.versionHash} was not found.`], [_exceptions2().NoFiles, err => _chalk().default.yellow('warning: no files to add') + _chalk().default.yellow(err.ignoredFiles ? `, the following files were ignored: ${_chalk().default.bold(err.ignoredFiles)}` : '')], [_exceptions2().DuplicateIds, err => Object.keys(err.componentObject).map(key => {
  return `unable to add ${Object.keys(err.componentObject[key]).length} components with the same ID: ${_chalk().default.bold(key)} : ${err.componentObject[key]}\n`;
}).join(' ')], [_idExportedAlready().default, err => `component ${_chalk().default.bold(err.id)} has been already exported to ${_chalk().default.bold(err.remote)}`], [_invalidVersion().default, err => `error: version ${_chalk().default.bold(err.version)} is not a valid semantic version. learn more: https://semver.org`], [_noIdMatchWildcard().default, err => `unable to find component ids that match the following: ${err.idsWithWildcards.join(', ')}`], [_nothingToCompareTo().default, () => 'no previous versions to compare'], [_exceptions3().NewerVersionFound,
// err => JSON.stringify(err.newerVersions)
err => (0, _newerVersionTemplate().default)(err.newerVersions)], [_promptCanceled().default, () => _chalk().default.yellow('operation aborted')], [_externalTestErrors().default, err => `error: bit failed to test ${err.id} with the following exception:\n${getExternalErrorsMessageAndStack(err.originalErrors)}`], [_externalBuildErrors().default, err => `error: bit failed to build ${err.id} with the following exception:\n${getExternalErrorsMessageAndStack(err.originalErrors)}`], [_exceptions5().ResolutionException, err => `error: bit failed to require ${err.filePath} due to the following exception:\n${getExternalErrorMessage(err.originalError)}.\n${err.originalError.stack}`], [_gitNotFound().default, () => "error: unable to run command because git executable not found. please ensure git is installed and/or git_path is configured using the 'bit config set git_path <GIT_PATH>'"], [_exceptions6().AuthenticationFailed, err => `authentication failed. see troubleshooting at https://${_constants().BASE_LEGACY_DOCS_DOMAIN}/setup-authentication#autentication-issues.html\n\n${err.debugInfo}`], [_objectsWithoutConsumer().default, err => `error: unable to initialize a bit workspace. bit has found undeleted local objects at ${_chalk().default.bold(err.scopePath)}.
1. use the ${_chalk().default.bold('--reset-hard')} flag to clear all data and initialize an empty workspace.
2. if deleted by mistake, please restore .bitmap and bit.json.
3. force workspace initialization without clearing data use the ${_chalk().default.bold('--force')} flag.`]];
function findErrorDefinition(err) {
  const error = errorsMap.find(([ErrorType]) => {
    return err instanceof ErrorType || err && err.name === ErrorType.name; // in some cases, such as forked process, the received err is serialized.
  });

  return error;
}
function getErrorFunc(errorDefinition) {
  if (!errorDefinition) return null;
  const [, func] = errorDefinition;
  return func;
}
function getErrorMessage(error, func) {
  if (!error || !func) return '';
  let errorMessage = func(error);
  // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
  if (error.showDoctorMessage) {
    errorMessage = `${errorMessage}

run 'bit doctor' to get detailed workspace diagnosis and issue resolution.`;
  }
  return errorMessage;
}
function getExternalErrorMessage(externalError) {
  if (!externalError) return '';

  // In case an error is not a real error
  if (!(externalError instanceof Error)) {
    return externalError;
  }
  // In case it's not a bit error
  if (externalError.message) {
    return externalError.message;
  }
  const errorDefinition = findErrorDefinition(externalError);
  const func = getErrorFunc(errorDefinition);
  const errorMessage = getErrorMessage(externalError, func);
  return errorMessage;
}
function getExternalErrorsMessageAndStack(errors) {
  const result = errors.map(e => {
    const msg = getExternalErrorMessage(e);
    const stack = e.stack || '';
    return `${msg}\n${stack}\n`;
  }).join('~~~~~~~~~~~~~\n');
  return result;
}

/**
 * if err.userError is set, it inherits from AbstractError, which are user errors not Bit errors
 * and should not be reported to Sentry.
 * reason why we don't check (err instanceof AbstractError) is that it could be thrown from a fork,
 * in which case, it loses its class and has only the fields.
 */
function sendToAnalyticsAndSentry(err) {
  const possiblyHashedError = (0, _hashErrorObject().default)(err);
  // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
  // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
  const shouldNotReportToSentry = Boolean(err.isUserError || err.code === 'EACCES');
  // only level FATAL are reported to Sentry.
  const level = shouldNotReportToSentry ? _analytics().LEVEL.INFO : _analytics().LEVEL.FATAL;
  _analytics().Analytics.setError(level, possiblyHashedError);
}
function handleNonBitCustomErrors(err) {
  // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
  if (err.code === 'EACCES' && err.path) {
    // see #1774
    return _chalk().default.red(
    // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
    `error: you do not have permissions to access '${err.path}', were you running bit, npm or git as root?`);
  }
  // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
  return _chalk().default.red(err.message || err);
}
var _default = err => {
  const errorDefinition = findErrorDefinition(err);
  const getErrMsg = () => {
    if (err instanceof _bitError().BitError) {
      return err.report();
    }
    if (!errorDefinition) {
      return handleNonBitCustomErrors(err);
    }
    const func = getErrorFunc(errorDefinition);
    const errorMessage = getErrorMessage(err, func) || 'unknown error';
    err.message = errorMessage;
    return errorMessage;
  };
  sendToAnalyticsAndSentry(err);
  const errorMessage = getErrMsg();
  return {
    message: _chalk().default.red(errorMessage),
    error: err
  };
};
exports.default = _default;
