// @teambit/legacy/dist/links/link-content.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.JAVASCRIPT_FLAVORS_EXTENSIONS = exports.EXTENSIONS_TO_STRIP_FROM_PACKAGES = exports.EXTENSIONS_TO_REPLACE_TO_JS_IN_PACKAGES = exports.EXTENSIONS_NOT_SUPPORT_DIRS = void 0;
exports.isSupportedExtension = isSupportedExtension;
function _utils() {
  const data = require("../utils");
  _utils = function () {
    return data;
  };
  return data;
}
const LINKS_CONTENT_TEMPLATES = {
  js: "module.exports = require('{filePath}');",
  ts: "export * from '{filePath}';",
  jsx: "export * from '{filePath}';",
  tsx: "export * from '{filePath}';",
  css: "@import '{filePath}.css';",
  scss: "@import '{filePath}.scss';",
  sass: "@import '{filePath}.sass';",
  less: "@import '{filePath}.less';"
};
const JAVASCRIPT_FLAVORS_EXTENSIONS = ['js', 'ts', 'jsx', 'tsx'];
exports.JAVASCRIPT_FLAVORS_EXTENSIONS = JAVASCRIPT_FLAVORS_EXTENSIONS;
const EXTENSIONS_TO_STRIP_FROM_PACKAGES = ['js', 'ts', 'jsx', 'tsx', 'd.ts'];
exports.EXTENSIONS_TO_STRIP_FROM_PACKAGES = EXTENSIONS_TO_STRIP_FROM_PACKAGES;
const EXTENSIONS_TO_REPLACE_TO_JS_IN_PACKAGES = ['ts', 'jsx', 'tsx'];
// node-sass doesn't resolve directories to 'index.scss', @see https://github.com/sass/sass/issues/690
exports.EXTENSIONS_TO_REPLACE_TO_JS_IN_PACKAGES = EXTENSIONS_TO_REPLACE_TO_JS_IN_PACKAGES;
const EXTENSIONS_NOT_SUPPORT_DIRS = ['scss', 'sass'];
exports.EXTENSIONS_NOT_SUPPORT_DIRS = EXTENSIONS_NOT_SUPPORT_DIRS;
function isSupportedExtension(filePath) {
  const ext = (0, _utils().getExt)(filePath);
  const supportedExtensions = getSupportedExtensions();
  return supportedExtensions.includes(ext);
}
function getSupportedExtensions() {
  return Object.keys(LINKS_CONTENT_TEMPLATES);
}
