// @teambit/react/dist/apps/web/ssr/ssr-express.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.array.iterator.js");
require("core-js/modules/es.promise.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createExpressSsr = createExpressSsr;
exports.loadSsrApp = loadSsrApp;
exports.parseAssets = parseAssets;
function _uiFoundationUiPages() {
  const data = require("@teambit/ui-foundation.ui.pages.static-error");
  _uiFoundationUiPages = function () {
    return data;
  };
  return data;
}
function _reactRendering() {
  const data = require("@teambit/react.rendering.ssr");
  _reactRendering = function () {
    return data;
  };
  return data;
}
function _express() {
  const data = _interopRequireDefault(require("express"));
  _express = function () {
    return data;
  };
  return data;
}
function fs() {
  const data = _interopRequireWildcard(require("fs-extra"));
  fs = function () {
    return data;
  };
  return data;
}
function _path() {
  const data = require("path");
  _path = function () {
    return data;
  };
  return data;
}
function _urlJoin() {
  const data = _interopRequireDefault(require("url-join"));
  _urlJoin = function () {
    return data;
  };
  return data;
}
function _webpackAppSsr() {
  const data = require("../webpack/webpack.app.ssr.config");
  _webpackAppSsr = function () {
    return data;
  };
  return data;
}
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
const MAGIC_FOLDER = 'public'; // idk where this is comping from

function createExpressSsr({
  name,
  workdir,
  port,
  app,
  assets,
  logger
}) {
  const express = (0, _express().default)();
  const publicFolder = (0, _path().resolve)(workdir, (0, _webpackAppSsr().calcOutputPath)(name, 'browser'), MAGIC_FOLDER);
  express.use(_webpackAppSsr().PUBLIC_PATH, _express().default.static(publicFolder));
  express.use((request, response, next) => {
    if (request.query._rendering !== 'client') {
      next();
      return;
    }
    response.sendFile((0, _path().resolve)(workdir, (0, _webpackAppSsr().calcOutputPath)(name, 'browser'), MAGIC_FOLDER, 'index.html'));
  });
  express.use(async (request, response) => {
    logger === null || logger === void 0 ? void 0 : logger.info(`[react.application] [ssr] handling "${request.url}"`);
    const browser = (0, _reactRendering().browserFromExpress)(request, port);
    const session = {
      assets,
      browser,
      request,
      response
    };
    try {
      // the app itself controls the response
      await app(session);
      logger === null || logger === void 0 ? void 0 : logger.console(`[react.application] [ssr] success "${request.url}"`);
    } catch (error) {
      logger === null || logger === void 0 ? void 0 : logger.consoleError(`[react.application] [ssr] error at "${request.url}"`, error);
      response.status(500).send((0, _uiFoundationUiPages().serverError)());
    }
  });
  return express;
}
async function loadSsrApp(workdir, appName) {
  const entryFile = (0, _path().resolve)(workdir, (0, _webpackAppSsr().calcOutputPath)(appName, 'ssr'), MAGIC_FOLDER, _webpackAppSsr().SSR_ENTRY_FILE);
  if (!fs().existsSync(entryFile)) throw new Error(`expected ssr bundle entry file at "${entryFile}"`);
  const entry = await Promise.resolve(`${entryFile}`).then(s => _interopRequireWildcard(require(s)));
  const app = entry === null || entry === void 0 ? void 0 : entry.default;
  if (!app) throw new Error(`bundle entry file has no default export (at "${entryFile}")`);
  return app;
}
function parseAssets(assets, publicPath = _webpackAppSsr().PUBLIC_PATH) {
  const deadAssets = assets.filter(x => !x.name);
  if (deadAssets.length > 0) throw new Error('missing some build assets (maybe need to turn on cachedAssets, etc)');
  return {
    css: assets.map(x => x.name).filter(name => name === null || name === void 0 ? void 0 : name.endsWith('.css')).map(name => (0, _urlJoin().default)(publicPath, name)),
    js: assets.map(x => x.name).filter(name => name === null || name === void 0 ? void 0 : name.endsWith('.js')).map(name => (0, _urlJoin().default)(publicPath, name))
  };
}
