// @teambit/typescript/dist/transformers/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "BindingElementTransformer", {
  enumerable: true,
  get: function () {
    return _bindingElement().BindingElementTransformer;
  }
});
Object.defineProperty(exports, "ClassDecelerationTransformer", {
  enumerable: true,
  get: function () {
    return _classDeceleration().ClassDecelerationTransformer;
  }
});
Object.defineProperty(exports, "EnumDeclarationTransformer", {
  enumerable: true,
  get: function () {
    return _enumDeclaration().EnumDeclarationTransformer;
  }
});
Object.defineProperty(exports, "ExportDeclaration", {
  enumerable: true,
  get: function () {
    return _exportDeclaration().ExportDeclaration;
  }
});
Object.defineProperty(exports, "FunctionDeclaration", {
  enumerable: true,
  get: function () {
    return _functionDeclaration().FunctionDeclaration;
  }
});
Object.defineProperty(exports, "InterfaceDeclarationTransformer", {
  enumerable: true,
  get: function () {
    return _interfaceDeclaration().InterfaceDeclarationTransformer;
  }
});
Object.defineProperty(exports, "SourceFileTransformer", {
  enumerable: true,
  get: function () {
    return _sourceFileTransformer().SourceFileTransformer;
  }
});
Object.defineProperty(exports, "TypeAliasTransformer", {
  enumerable: true,
  get: function () {
    return _typeAlias().TypeAliasTransformer;
  }
});
Object.defineProperty(exports, "VariableDeclaration", {
  enumerable: true,
  get: function () {
    return _variableDeclaration().VariableDeclaration;
  }
});
Object.defineProperty(exports, "VariableStatementTransformer", {
  enumerable: true,
  get: function () {
    return _variableStatement().VariableStatementTransformer;
  }
});
function _exportDeclaration() {
  const data = require("./export-declaration");
  _exportDeclaration = function () {
    return data;
  };
  return data;
}
function _functionDeclaration() {
  const data = require("./function-declaration");
  _functionDeclaration = function () {
    return data;
  };
  return data;
}
function _variableStatement() {
  const data = require("./variable-statement");
  _variableStatement = function () {
    return data;
  };
  return data;
}
function _variableDeclaration() {
  const data = require("./variable-declaration");
  _variableDeclaration = function () {
    return data;
  };
  return data;
}
function _sourceFileTransformer() {
  const data = require("./source-file-transformer");
  _sourceFileTransformer = function () {
    return data;
  };
  return data;
}
function _typeAlias() {
  const data = require("./type-alias");
  _typeAlias = function () {
    return data;
  };
  return data;
}
function _classDeceleration() {
  const data = require("./class-deceleration");
  _classDeceleration = function () {
    return data;
  };
  return data;
}
function _interfaceDeclaration() {
  const data = require("./interface-declaration");
  _interfaceDeclaration = function () {
    return data;
  };
  return data;
}
function _enumDeclaration() {
  const data = require("./enum-declaration");
  _enumDeclaration = function () {
    return data;
  };
  return data;
}
function _bindingElement() {
  const data = require("./binding-element");
  _bindingElement = function () {
    return data;
  };
  return data;
}
