// @teambit/install/dist/link/core-aspects-links.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CoreAspectsLinks = CoreAspectsLinks;
function _chalk() {
  const data = _interopRequireDefault(require("chalk"));
  _chalk = function () {
    return data;
  };
  return data;
}
function _getPackageNameFromTarget() {
  const data = require("./get-package-name-from-target");
  _getPackageNameFromTarget = function () {
    return data;
  };
  return data;
}
function _linkRow() {
  const data = require("./link-row");
  _linkRow = function () {
    return data;
  };
  return data;
}
function CoreAspectsLinks({
  coreAspectsLinks,
  verbose = false
}) {
  if (!coreAspectsLinks || !coreAspectsLinks.length) {
    return _chalk().default.cyan('No core aspects were linked');
  }
  const title = _chalk().default.cyan('Core aspects links');
  const links = coreAspectsLinks.map(link => CoreAspectLinkRow({
    coreAspectLink: link,
    verbose
  })).join('\n');
  return `${title}\n${links}`;
}
function CoreAspectLinkRow({
  coreAspectLink,
  verbose = false
}) {
  if (verbose) return VerboseCoreAspectLink({
    coreAspectLink
  });
  return RegularCoreAspectLink({
    coreAspectLink
  });
}
function RegularCoreAspectLink({
  coreAspectLink
}) {
  const id = coreAspectLink.aspectId.toString();
  const packagePath = (0, _getPackageNameFromTarget().getPackageNameFromTarget)(coreAspectLink.linkDetail.to);
  return (0, _linkRow().LinkRow)({
    title: id,
    target: packagePath,
    padding: 50
  });
}
function VerboseCoreAspectLink({
  coreAspectLink
}) {
  const id = coreAspectLink.aspectId.toString();
  const title = _chalk().default.bold.cyan(id);
  const link = (0, _linkRow().VerboseLinkRow)({
    from: coreAspectLink.linkDetail.from,
    to: coreAspectLink.linkDetail.to
  });
  return `${title}\n${link}`;
}
