// @teambit/legacy/dist/consumer/exceptions/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ComponentOutOfSync", {
  enumerable: true,
  get: function () {
    return _componentOutOfSync().default;
  }
});
Object.defineProperty(exports, "ConsumerAlreadyExists", {
  enumerable: true,
  get: function () {
    return _consumerAlreadyExists().default;
  }
});
Object.defineProperty(exports, "ConsumerNotFound", {
  enumerable: true,
  get: function () {
    return _consumerNotFound().default;
  }
});
Object.defineProperty(exports, "LoginFailed", {
  enumerable: true,
  get: function () {
    return _loginFailed().default;
  }
});
Object.defineProperty(exports, "NewerVersionFound", {
  enumerable: true,
  get: function () {
    return _newerVersionFound().default;
  }
});
Object.defineProperty(exports, "NothingToImport", {
  enumerable: true,
  get: function () {
    return _nothingToImport().default;
  }
});
function _componentOutOfSync() {
  const data = _interopRequireDefault(require("./component-out-of-sync"));
  _componentOutOfSync = function () {
    return data;
  };
  return data;
}
function _consumerAlreadyExists() {
  const data = _interopRequireDefault(require("./consumer-already-exists"));
  _consumerAlreadyExists = function () {
    return data;
  };
  return data;
}
function _consumerNotFound() {
  const data = _interopRequireDefault(require("./consumer-not-found"));
  _consumerNotFound = function () {
    return data;
  };
  return data;
}
function _loginFailed() {
  const data = _interopRequireDefault(require("./login-failed"));
  _loginFailed = function () {
    return data;
  };
  return data;
}
function _newerVersionFound() {
  const data = _interopRequireDefault(require("./newer-version-found"));
  _newerVersionFound = function () {
    return data;
  };
  return data;
}
function _nothingToImport() {
  const data = _interopRequireDefault(require("./nothing-to-import"));
  _nothingToImport = function () {
    return data;
  };
  return data;
}
