// @teambit/node/dist/templates/express-app/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.expressAppTemplate = void 0;
function _index() {
  const data = require("./files/index");
  _index = function () {
    return data;
  };
  return data;
}
function _doc() {
  const data = require("./files/doc");
  _doc = function () {
    return data;
  };
  return data;
}
function _appRoot() {
  const data = require("./files/app-root");
  _appRoot = function () {
    return data;
  };
  return data;
}
function _nodeApp() {
  const data = require("./files/node-app");
  _nodeApp = function () {
    return data;
  };
  return data;
}
function _getPort() {
  const data = require("./files/get-port");
  _getPort = function () {
    return data;
  };
  return data;
}
function _mockRoute() {
  const data = require("./files/mock-route");
  _mockRoute = function () {
    return data;
  };
  return data;
}
function _route() {
  const data = require("./files/route");
  _route = function () {
    return data;
  };
  return data;
}
// import { testFile } from './files/test';
// import { compositionFile } from './files/composition';

const expressAppTemplate = {
  name: 'express-app',
  description: 'a bit express application',
  generateFiles: context => {
    return [(0, _index().indexFile)(context), (0, _doc().docFile)(context), (0, _appRoot().appRoot)(context), (0, _nodeApp().nodeApp)(context), (0, _getPort().getPort)(), (0, _mockRoute().mockRoute)(), (0, _route().route)()
    // compositionFile(context),
    // testFile(context),
    ];
  },

  config: {
    'teambit.harmony/aspect': {},
    'teambit.envs/envs': {
      env: 'teambit.harmony/aspect'
    }
  }
};
exports.expressAppTemplate = expressAppTemplate;
