// @teambit/ui/dist/events/ui-server-started-event.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UiServerStartedEvent = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _pubsub() {
  const data = require("@teambit/pubsub");
  _pubsub = function () {
    return data;
  };
  return data;
}
/* eslint-disable max-classes-per-file */

class UiServerStartedEventData {
  constructor(targetHost, targetPort, uiRoot) {
    this.targetHost = targetHost;
    this.targetPort = targetPort;
    this.uiRoot = uiRoot;
  }
}
class UiServerStartedEvent extends _pubsub().BitBaseEvent {
  constructor(timestamp, targetHost, targetPort, uiRoot) {
    super(UiServerStartedEvent.TYPE, '0.0.1', timestamp, new UiServerStartedEventData(targetHost, targetPort, uiRoot));
    this.timestamp = timestamp;
    this.targetHost = targetHost;
    this.targetPort = targetPort;
    this.uiRoot = uiRoot;
  }
}
exports.UiServerStartedEvent = UiServerStartedEvent;
(0, _defineProperty2().default)(UiServerStartedEvent, "TYPE", 'ui-server-started');
