// @teambit/bit/dist/bit.main.runtime.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BitMain = void 0;
exports.registerCoreExtensions = registerCoreExtensions;
function _cli() {
  const data = require("@teambit/cli");
  _cli = function () {
    return data;
  };
  return data;
}
function _extensionData() {
  const data = require("@teambit/legacy/dist/consumer/config/extension-data");
  _extensionData = function () {
    return data;
  };
  return data;
}
function _bit() {
  const data = require("./bit.aspect");
  _bit = function () {
    return data;
  };
  return data;
}
function _bit2() {
  const data = require("./bit.provider");
  _bit2 = function () {
    return data;
  };
  return data;
}
function _manifests() {
  const data = require("./manifests");
  _manifests = function () {
    return data;
  };
  return data;
}
const manifests = Object.values(_manifests().manifestsMap);
function registerCoreExtensions() {
  const allCoreExtensionsNames = Object.keys(_manifests().manifestsMap);
  _extensionData().ExtensionDataList.registerManyCoreExtensionNames(allCoreExtensionsNames);
}
const BitMain = {
  name: 'bit',
  runtime: _cli().MainRuntime,
  dependencies: manifests,
  provider: _bit2().provideBit
};
exports.BitMain = BitMain;
_bit().BitAspect.addRuntime(BitMain);
