// @teambit/bit/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Aspect", {
  enumerable: true,
  get: function () {
    return _harmony().Aspect;
  }
});
Object.defineProperty(exports, "BitAspect", {
  enumerable: true,
  get: function () {
    return _bit().BitAspect;
  }
});
Object.defineProperty(exports, "Hook", {
  enumerable: true,
  get: function () {
    return _hooks().Hook;
  }
});
Object.defineProperty(exports, "HookRegistry", {
  enumerable: true,
  get: function () {
    return _hooks().HookRegistry;
  }
});
Object.defineProperty(exports, "RuntimeDefinition", {
  enumerable: true,
  get: function () {
    return _harmony().RuntimeDefinition;
  }
});
Object.defineProperty(exports, "RuntimeManifest", {
  enumerable: true,
  get: function () {
    return _harmony().RuntimeManifest;
  }
});
Object.defineProperty(exports, "getAllCoreAspectsIds", {
  enumerable: true,
  get: function () {
    return _manifests().getAllCoreAspectsIds;
  }
});
Object.defineProperty(exports, "isCoreAspect", {
  enumerable: true,
  get: function () {
    return _manifests().isCoreAspect;
  }
});
Object.defineProperty(exports, "loadBit", {
  enumerable: true,
  get: function () {
    return _loadBit().loadBit;
  }
});
Object.defineProperty(exports, "manifestsMap", {
  enumerable: true,
  get: function () {
    return _manifests().manifestsMap;
  }
});
Object.defineProperty(exports, "registerCoreExtensions", {
  enumerable: true,
  get: function () {
    return _bitMain().registerCoreExtensions;
  }
});
function _harmony() {
  const data = require("@teambit/harmony");
  _harmony = function () {
    return data;
  };
  return data;
}
function _hooks() {
  const data = require("./hooks");
  _hooks = function () {
    return data;
  };
  return data;
}
function _manifests() {
  const data = require("./manifests");
  _manifests = function () {
    return data;
  };
  return data;
}
function _bitMain() {
  const data = require("./bit.main.runtime");
  _bitMain = function () {
    return data;
  };
  return data;
}
function _bit() {
  const data = require("./bit.aspect");
  _bit = function () {
    return data;
  };
  return data;
}
function _loadBit() {
  const data = require("./load-bit");
  _loadBit = function () {
    return data;
  };
  return data;
}
