// @teambit/legacy/dist/scope/objects/bit-object-list.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BitObjectList = void 0;
function _models() {
  const data = require("../models");
  _models = function () {
    return data;
  };
  return data;
}
class BitObjectList {
  constructor(objects) {
    this.objects = objects;
  }
  getComponents() {
    return this.objects.filter(object => object instanceof _models().ModelComponent);
  }
  getVersions() {
    return this.objects.filter(object => object instanceof _models().Version);
  }
  getLanes() {
    return this.objects.filter(object => object instanceof _models().Lane);
  }
  getAll() {
    return this.objects;
  }
  getExportMetadata() {
    return this.objects.find(object => object instanceof _models().ExportMetadata);
  }

  /**
   * object that needs merge operation before saving them into the scope, such as ModelComponent
   */
  getObjectsRequireMerge() {
    const typeRequireMerge = this.objectTypesRequireMerge();
    return this.objects.filter(object => typeRequireMerge.some(ObjClass => object instanceof ObjClass));
  }

  /**
   * object that don't need merge operation and can be saved immediately into the scope.
   * such as Source or Version
   */
  getObjectsNotRequireMerge() {
    const typeRequireMerge = this.objectTypesRequireMerge();
    return this.objects.filter(object => typeRequireMerge.every(ObjClass => !(object instanceof ObjClass)));
  }
  objectTypesRequireMerge() {
    return [_models().ModelComponent, _models().Lane];
  }
}
exports.BitObjectList = BitObjectList;
