// @teambit/legacy/dist/consumer/config/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ExtensionDataEntry", {
  enumerable: true,
  get: function () {
    return _extensionData().ExtensionDataEntry;
  }
});
Object.defineProperty(exports, "ExtensionDataList", {
  enumerable: true,
  get: function () {
    return _extensionData().ExtensionDataList;
  }
});
Object.defineProperty(exports, "ILegacyWorkspaceConfig", {
  enumerable: true,
  get: function () {
    return _legacyWorkspaceConfigInterface().ILegacyWorkspaceConfig;
  }
});
Object.defineProperty(exports, "LegacyWorkspaceConfig", {
  enumerable: true,
  get: function () {
    return _workspaceConfig().default;
  }
});
Object.defineProperty(exports, "LegacyWorkspaceConfigProps", {
  enumerable: true,
  get: function () {
    return _workspaceConfig().WorkspaceConfigProps;
  }
});
Object.defineProperty(exports, "REMOVE_EXTENSION_SPECIAL_SIGN", {
  enumerable: true,
  get: function () {
    return _extensionData().REMOVE_EXTENSION_SPECIAL_SIGN;
  }
});
Object.defineProperty(exports, "WorkspaceConfigEnsureFunction", {
  enumerable: true,
  get: function () {
    return _workspaceConfig().WorkspaceConfigEnsureFunction;
  }
});
Object.defineProperty(exports, "WorkspaceConfigLoadFunction", {
  enumerable: true,
  get: function () {
    return _workspaceConfig().WorkspaceConfigLoadFunction;
  }
});
exports.default = void 0;
function _componentConfig() {
  const data = _interopRequireDefault(require("./component-config"));
  _componentConfig = function () {
    return data;
  };
  return data;
}
function _workspaceConfig() {
  const data = _interopRequireWildcard(require("./workspace-config"));
  _workspaceConfig = function () {
    return data;
  };
  return data;
}
function _legacyWorkspaceConfigInterface() {
  const data = require("./legacy-workspace-config-interface");
  _legacyWorkspaceConfigInterface = function () {
    return data;
  };
  return data;
}
function _extensionData() {
  const data = require("./extension-data");
  _extensionData = function () {
    return data;
  };
  return data;
}
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
var _default = _componentConfig().default;
exports.default = _default;
