// @teambit/compiler/dist/templates/compiler/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.compilerTemplate = void 0;
function _indexFile() {
  const data = require("./files/index-file");
  _indexFile = function () {
    return data;
  };
  return data;
}
function _aspectFile() {
  const data = require("./files/aspect-file");
  _aspectFile = function () {
    return data;
  };
  return data;
}
function _mainRuntimeFile() {
  const data = require("./files/main-runtime-file");
  _mainRuntimeFile = function () {
    return data;
  };
  return data;
}
function _compilerFile() {
  const data = require("./files/compiler-file");
  _compilerFile = function () {
    return data;
  };
  return data;
}
const compilerTemplate = {
  name: 'compiler',
  description: 'A compiler implementation',
  hidden: true,
  generateFiles: context => {
    return [{
      relativePath: 'index.ts',
      content: (0, _indexFile().indexFile)(context),
      isMain: true
    }, {
      relativePath: `${context.name}.aspect.ts`,
      content: (0, _aspectFile().aspectFile)(context)
    }, {
      relativePath: `${context.name}.compiler.ts`,
      content: (0, _compilerFile().compilerFile)(context)
    }, {
      relativePath: `${context.name}.main.runtime.ts`,
      content: (0, _mainRuntimeFile().mainRuntimeFile)(context)
    }];
  }
};
exports.compilerTemplate = compilerTemplate;
