// @teambit/generator/dist/templates/component-generator/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.componentGeneratorTemplate = void 0;
function _index() {
  const data = require("./files/index");
  _index = function () {
    return data;
  };
  return data;
}
function _aspectFile() {
  const data = require("./files/aspect-file");
  _aspectFile = function () {
    return data;
  };
  return data;
}
function _docsFile() {
  const data = require("./files/docs-file");
  _docsFile = function () {
    return data;
  };
  return data;
}
function _mainRuntime() {
  const data = require("./files/main-runtime");
  _mainRuntime = function () {
    return data;
  };
  return data;
}
const componentGeneratorTemplate = {
  name: 'component-generator',
  description: 'create your own component generator \nDocs: https://bit.dev/docs/dev-services/generator/generate-component',
  generateFiles: context => {
    return [{
      relativePath: 'index.ts',
      content: (0, _index().indexFile)(context),
      isMain: true
    }, {
      relativePath: `${context.name}.aspect.ts`,
      content: (0, _aspectFile().aspectFile)(context)
    }, {
      relativePath: `${context.name}.docs.mdx`,
      content: (0, _docsFile().docsFile)()
    }, {
      relativePath: `${context.name}.main.runtime.ts`,
      content: (0, _mainRuntime().mainRuntime)(context)
    }];
  },
  config: {
    'teambit.harmony/aspect': {},
    'teambit.envs/envs': {
      env: 'teambit.harmony/aspect'
    }
  }
};
exports.componentGeneratorTemplate = componentGeneratorTemplate;
