// @teambit/express/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ExpressAspect", {
  enumerable: true,
  get: function () {
    return _express().ExpressAspect;
  }
});
Object.defineProperty(exports, "NextFunction", {
  enumerable: true,
  get: function () {
    return _types().NextFunction;
  }
});
Object.defineProperty(exports, "Request", {
  enumerable: true,
  get: function () {
    return _types().Request;
  }
});
Object.defineProperty(exports, "Response", {
  enumerable: true,
  get: function () {
    return _types().Response;
  }
});
Object.defineProperty(exports, "Route", {
  enumerable: true,
  get: function () {
    return _types().Route;
  }
});
Object.defineProperty(exports, "RouteSlot", {
  enumerable: true,
  get: function () {
    return _expressMain().RouteSlot;
  }
});
Object.defineProperty(exports, "Verb", {
  enumerable: true,
  get: function () {
    return _types().Verb;
  }
});
function _expressMain() {
  const data = require("./express.main.runtime");
  _expressMain = function () {
    return data;
  };
  return data;
}
function _types() {
  const data = require("./types");
  _types = function () {
    return data;
  };
  return data;
}
function _express() {
  const data = require("./express.aspect");
  _express = function () {
    return data;
  };
  return data;
}
