// @teambit/legacy/dist/api/consumer/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "add", {
  enumerable: true,
  get: function () {
    return _add().addOne;
  }
});
Object.defineProperty(exports, "addMany", {
  enumerable: true,
  get: function () {
    return _add().addMany;
  }
});
Object.defineProperty(exports, "clearCache", {
  enumerable: true,
  get: function () {
    return _clearCache().clearCache;
  }
});
Object.defineProperty(exports, "config", {
  enumerable: true,
  get: function () {
    return _globalConfig().default;
  }
});
Object.defineProperty(exports, "dependencyStatus", {
  enumerable: true,
  get: function () {
    return _dependency_status().default;
  }
});
Object.defineProperty(exports, "diff", {
  enumerable: true,
  get: function () {
    return _diff().default;
  }
});
Object.defineProperty(exports, "getComponentLogs", {
  enumerable: true,
  get: function () {
    return _getComponentLogs().default;
  }
});
Object.defineProperty(exports, "getConsumerComponent", {
  enumerable: true,
  get: function () {
    return _getConsumerComponent().default;
  }
});
Object.defineProperty(exports, "getScopeComponent", {
  enumerable: true,
  get: function () {
    return _getScopeComponent().default;
  }
});
Object.defineProperty(exports, "init", {
  enumerable: true,
  get: function () {
    return _init().default;
  }
});
Object.defineProperty(exports, "link", {
  enumerable: true,
  get: function () {
    return _link().default;
  }
});
Object.defineProperty(exports, "listScope", {
  enumerable: true,
  get: function () {
    return _listScope().listScope;
  }
});
Object.defineProperty(exports, "login", {
  enumerable: true,
  get: function () {
    return _login().default;
  }
});
Object.defineProperty(exports, "migrate", {
  enumerable: true,
  get: function () {
    return _migrate().default;
  }
});
Object.defineProperty(exports, "move", {
  enumerable: true,
  get: function () {
    return _move().default;
  }
});
Object.defineProperty(exports, "remoteAdd", {
  enumerable: true,
  get: function () {
    return _remote().add;
  }
});
Object.defineProperty(exports, "remoteList", {
  enumerable: true,
  get: function () {
    return _remote().list;
  }
});
Object.defineProperty(exports, "remoteRm", {
  enumerable: true,
  get: function () {
    return _remote().remove;
  }
});
Object.defineProperty(exports, "show", {
  enumerable: true,
  get: function () {
    return _show().default;
  }
});
function _diff() {
  const data = _interopRequireDefault(require("./lib//diff"));
  _diff = function () {
    return data;
  };
  return data;
}
function _add() {
  const data = require("./lib/add");
  _add = function () {
    return data;
  };
  return data;
}
function _dependency_status() {
  const data = _interopRequireDefault(require("./lib/dependency_status"));
  _dependency_status = function () {
    return data;
  };
  return data;
}
function _getComponentLogs() {
  const data = _interopRequireDefault(require("./lib/get-component-logs"));
  _getComponentLogs = function () {
    return data;
  };
  return data;
}
function _getConsumerComponent() {
  const data = _interopRequireDefault(require("./lib/get-consumer-component"));
  _getConsumerComponent = function () {
    return data;
  };
  return data;
}
function _getScopeComponent() {
  const data = _interopRequireDefault(require("./lib/get-scope-component"));
  _getScopeComponent = function () {
    return data;
  };
  return data;
}
function _globalConfig() {
  const data = _interopRequireDefault(require("./lib/global-config"));
  _globalConfig = function () {
    return data;
  };
  return data;
}
function _init() {
  const data = _interopRequireDefault(require("./lib/init"));
  _init = function () {
    return data;
  };
  return data;
}
function _link() {
  const data = _interopRequireDefault(require("./lib/link"));
  _link = function () {
    return data;
  };
  return data;
}
function _listScope() {
  const data = require("./lib/list-scope");
  _listScope = function () {
    return data;
  };
  return data;
}
function _login() {
  const data = _interopRequireDefault(require("./lib/login"));
  _login = function () {
    return data;
  };
  return data;
}
function _migrate() {
  const data = _interopRequireDefault(require("./lib/migrate"));
  _migrate = function () {
    return data;
  };
  return data;
}
function _move() {
  const data = _interopRequireDefault(require("./lib/move"));
  _move = function () {
    return data;
  };
  return data;
}
function _remote() {
  const data = require("./lib/remote");
  _remote = function () {
    return data;
  };
  return data;
}
function _show() {
  const data = _interopRequireDefault(require("./lib/show"));
  _show = function () {
    return data;
  };
  return data;
}
function _clearCache() {
  const data = require("./lib/clear-cache");
  _clearCache = function () {
    return data;
  };
  return data;
}
