// @teambit/legacy/dist/global-config/config.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.array.iterator.js");
require("core-js/modules/es.promise.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
function _fsExtra() {
  const data = _interopRequireDefault(require("fs-extra"));
  _fsExtra = function () {
    return data;
  };
  return data;
}
function path() {
  const data = _interopRequireWildcard(require("path"));
  path = function () {
    return data;
  };
  return data;
}
function _constants() {
  const data = require("../constants");
  _constants = function () {
    return data;
  };
  return data;
}
function _utils() {
  const data = require("../utils");
  _utils = function () {
    return data;
  };
  return data;
}
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function getPath() {
  return path().join(_constants().GLOBAL_CONFIG, _constants().GLOBAL_CONFIG_FILE);
}
class Config extends Map {
  toPlainObject() {
    return (0, _utils().mapToObject)(this);
  }
  toJson() {
    return JSON.stringify(this.toPlainObject());
  }
  write() {
    return _fsExtra().default.outputFile(getPath(), this.toJson());
  }
  writeSync() {
    return _fsExtra().default.outputFileSync(getPath(), this.toJson());
  }
  static loadSync() {
    const configPath = getPath();
    if (!_fsExtra().default.existsSync(configPath)) {
      const config = new Config([]);
      config.writeSync();
      return config;
    }
    const contents = _fsExtra().default.readFileSync(configPath);
    return new Config(Object.entries(JSON.parse(contents.toString())));
  }
  static async load() {
    const configPath = getPath();
    const exists = await _fsExtra().default.pathExists(configPath);
    if (!exists) {
      const config = new Config([]);
      await config.write();
      return config;
    }
    const contents = await _fsExtra().default.readFile(configPath);
    return new Config(Object.entries(JSON.parse(contents.toString())));
  }
}
exports.default = Config;
