// @teambit/legacy/dist/scope/objects-fetcher/write-components-queue.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WriteComponentsQueue = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _pQueue() {
  const data = _interopRequireDefault(require("p-queue"));
  _pQueue = function () {
    return data;
  };
  return data;
}
class WriteComponentsQueue {
  constructor(concurrency = 1) {
    (0, _defineProperty2().default)(this, "queue", void 0);
    (0, _defineProperty2().default)(this, "processedIds", []);
    this.queue = new (_pQueue().default)({
      concurrency,
      autoStart: true
    });
  }
  addComponent(id, fn) {
    this.processedIds.push(id);
    return this.add(fn);
  }
  getQueue() {
    return this.queue;
  }
  add(fn, priority) {
    return this.queue.add(fn, {
      priority
    });
  }
  onIdle() {
    return this.queue.onIdle();
  }
}
exports.WriteComponentsQueue = WriteComponentsQueue;
