// @teambit/legacy/dist/scope/objects-fetcher/write-objects-queue.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WriteObjectsQueue = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _pQueue() {
  const data = _interopRequireDefault(require("p-queue"));
  _pQueue = function () {
    return data;
  };
  return data;
}
function _concurrency() {
  const data = require("../../utils/concurrency");
  _concurrency = function () {
    return data;
  };
  return data;
}
class WriteObjectsQueue {
  constructor(concurrency = (0, _concurrency().concurrentIOLimit)()) {
    (0, _defineProperty2().default)(this, "queue", void 0);
    (0, _defineProperty2().default)(this, "addedHashes", []);
    (0, _defineProperty2().default)(this, "added", 0);
    this.queue = new (_pQueue().default)({
      concurrency,
      autoStart: true
    });
  }
  addImmutableObject(hash, fn) {
    if (this.addedHashes.includes(hash)) {
      return null;
    }
    this.addedHashes.push(hash);
    return this.add(fn);
  }
  getQueue() {
    return this.queue;
  }
  add(fn, priority) {
    this.added += 1;
    return this.queue.add(fn, {
      priority
    });
  }
  onIdle() {
    return this.queue.onIdle();
  }
}
exports.WriteObjectsQueue = WriteObjectsQueue;
