// @teambit/schema/dist/mock/button/button.composition.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ButtonWithIcon = exports.ButtonWithCustomStyles = exports.ButtonAsALink = exports.BasicButton = void 0;
exports.Footer = Footer;
function _react() {
  const data = _interopRequireDefault(require("react"));
  _react = function () {
    return data;
  };
  return data;
}
function _button() {
  const data = require("./button");
  _button = function () {
    return data;
  };
  return data;
}
/* eslint-disable no-alert */
/* eslint-disable @typescript-eslint/no-unused-vars */

const BasicButton = () => {
  return /*#__PURE__*/_react().default.createElement(_button().Button, null, "click me");
};

/**
 * example of passing a JSX Element as a parameter
 */
exports.BasicButton = BasicButton;
function Footer({
  children = /*#__PURE__*/_react().default.createElement(BasicButton, null)
}) {
  return /*#__PURE__*/_react().default.createElement(_button().Button, null, children);
}
const ButtonWithCustomStyles = () => {
  return /*#__PURE__*/_react().default.createElement(_button().Button, {
    style: {
      background: 'red'
    }
  }, "click me");
};
exports.ButtonWithCustomStyles = ButtonWithCustomStyles;
const ButtonWithIcon = () => {
  return /*#__PURE__*/_react().default.createElement(_button().Button, null, "click me");
};
exports.ButtonWithIcon = ButtonWithIcon;
const ButtonAsALink = () => {
  return /*#__PURE__*/_react().default.createElement(_button().Button, {
    href: "https://bit.dev"
  }, "Bit");
};
exports.ButtonAsALink = ButtonAsALink;
