// @teambit/schema/dist/mock/button/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.array.iterator.js");
require("core-js/modules/es.promise.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  myFunc: true,
  Hi: true,
  a: true,
  b: true,
  HiThere: true,
  Function: true,
  Array: true,
  ClassSomething: true,
  a1: true,
  getBar: true,
  tuple: true,
  Food: true,
  getPromise: true,
  typePredicateFn: true,
  typePredicateNoTypeFn: true,
  objectBindingElements: true,
  arrayBindingElements: true,
  Compositions: true
};
exports.Function = exports.Food = exports.Compositions = exports.ClassSomething = exports.Array = void 0;
exports.Hi = Hi;
exports.a1 = exports.a = exports.HiThere = void 0;
exports.arrayBindingElements = arrayBindingElements;
exports.getBar = exports.b = void 0;
exports.getPromise = getPromise;
exports.myFunc = myFunc;
exports.objectBindingElements = objectBindingElements;
exports.tuple = void 0;
exports.typePredicateFn = typePredicateFn;
exports.typePredicateNoTypeFn = typePredicateNoTypeFn;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
var _button = require("./button");
Object.keys(_button).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _button[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _button[key];
    }
  });
});
function _Compositions() {
  const data = _interopRequireWildcard(require("./button.composition"));
  _Compositions = function () {
    return data;
  };
  return data;
}
Object.defineProperty(exports, "Compositions", {
  enumerable: true,
  get: function () {
    return _Compositions();
  }
});
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
/* eslint-disable @typescript-eslint/no-empty-function */
/* eslint-disable @typescript-eslint/no-shadow */
/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable @typescript-eslint/prefer-as-const */
/* eslint-disable one-var */
/* eslint-disable max-classes-per-file */

/**
 * General comment of the myFunc
 * @deprecate example of deprecation tag
 * @param a { number } this is A
 * @param b this is B
 * @returns { number } results of adding a to b
 */
function myFunc(a = 4, b = 5) {
  return a + b;
}
function Hi() {}
const a = 4,
  b = 5;
exports.b = b;
exports.a = a;
const HiThere = 'HiThere';
exports.HiThere = HiThere;
const Function = () => {};
exports.Function = Function;
const Array = ['hi', 'there'];
exports.Array = Array;
class Foo {}
class ClassSomething {
  constructor(da) {
    this.da = da;
    (0, _defineProperty2().default)(this, "app", '');
  }
  a() {
    return new Foo();
  }
  get getter() {
    return 'hi';
  }
  set setter(a) {}
}
exports.ClassSomething = ClassSomething;
const obj = {
  a: 1,
  b: 2
};
const a1 = {
  a: 5,
  b: 9
};
exports.a1 = a1;
// this is for Jump in the definition
class Bar {
  foo() {}
}
const getBar = bar => new Bar();
exports.getBar = getBar;
const tuple = ([a, b, c]) => {};
exports.tuple = tuple;
let Food;
exports.Food = Food;
(function (Food) {
  Food[Food["Falafel"] = 0] = "Falafel";
  Food[Food["Hummus"] = 1] = "Hummus";
  Food[Food["Tahini"] = 2] = "Tahini";
})(Food || (exports.Food = Food = {}));
async function getPromise() {
  return 'promise';
}
class T1 {}
class T2 {}
class T3 {}
function typePredicateFn(str) {
  return str;
}
function typePredicateNoTypeFn(condition, msg) {}
async function objectBindingElements({
  prop = 1
}) {
  return prop;
}
async function arrayBindingElements([prop]) {
  return prop;
}
const computedName = 'str';
