// @teambit/cloud/dist/cloud.main.runtime.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.promise.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CloudMain = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _cli() {
  const data = require("@teambit/cli");
  _cli = function () {
    return data;
  };
  return data;
}
function _constants() {
  const data = require("@teambit/legacy/dist/constants");
  _constants = function () {
    return data;
  };
  return data;
}
function _cloud() {
  const data = require("./cloud.aspect");
  _cloud = function () {
    return data;
  };
  return data;
}
class CloudMain {
  constructor(config) {
    this.config = config;
  }
  getCloudDomain() {
    return this.config.cloudDomain;
  }
  getCloudHubDomain() {
    return this.config.cloudHubDomain;
  }
  getCloudApi() {
    return this.config.cloudApi;
  }
  getCloudGraphQL() {
    return this.config.cloudGraphQL;
  }
  getLoginDomain() {
    return this.config.loginDomain;
  }
  getAnalyticsDomain() {
    return this.config.analyticsDomain;
  }
  getRegistryUrl() {
    return this.config.registryUrl;
  }
  getCloudExporterUrl() {
    return this.config.cloudExporterUrl;
  }
  getHubName() {
    return this.config.cloudHubName;
  }
  static async provider(_deps, config) {
    return new CloudMain(config);
  }
}
exports.CloudMain = CloudMain;
(0, _defineProperty2().default)(CloudMain, "slots", []);
(0, _defineProperty2().default)(CloudMain, "dependencies", []);
(0, _defineProperty2().default)(CloudMain, "runtime", _cli().MainRuntime);
(0, _defineProperty2().default)(CloudMain, "defaultConfig", {
  cloudDomain: _constants().BASE_CLOUD_DOMAIN,
  cloudHubDomain: _constants().DEFAULT_HUB_DOMAIN,
  cloudApi: _constants().SYMPHONY_URL,
  cloudGraphQL: _constants().SYMPHONY_GRAPHQL,
  loginDomain: _constants().DEFAULT_HUB_LOGIN,
  analyticsDomain: _constants().DEFAULT_ANALYTICS_DOMAIN,
  registryUrl: _constants().DEFAULT_REGISTRY_URL,
  cloudExporterUrl: _constants().CENTRAL_BIT_HUB_URL,
  cloudHubName: _constants().CENTRAL_BIT_HUB_NAME
});
_cloud().CloudAspect.addRuntime(CloudMain);
