// @teambit/notifications/dist/notification.ui.runtime.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.array.iterator.js");
require("core-js/modules/es.promise.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
function _extends2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
  _extends2 = function () {
    return data;
  };
  return data;
}
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _ui() {
  const data = require("@teambit/ui");
  _ui = function () {
    return data;
  };
  return data;
}
function _react() {
  const data = _interopRequireWildcard(require("react"));
  _react = function () {
    return data;
  };
  return data;
}
function _uuid() {
  const data = require("uuid");
  _uuid = function () {
    return data;
  };
  return data;
}
function _uiFoundationUiNotifications() {
  const data = require("@teambit/ui-foundation.ui.notifications.notification-context");
  _uiFoundationUiNotifications = function () {
    return data;
  };
  return data;
}
function _uiFoundationUiNotifications2() {
  const data = require("@teambit/ui-foundation.ui.notifications.notification-center");
  _uiFoundationUiNotifications2 = function () {
    return data;
  };
  return data;
}
function _uiFoundationUiNotifications3() {
  const data = require("@teambit/ui-foundation.ui.notifications.store");
  _uiFoundationUiNotifications3 = function () {
    return data;
  };
  return data;
}
function _notificationReducer() {
  const data = require("./notification-reducer");
  _notificationReducer = function () {
    return data;
  };
  return data;
}
function _notifications() {
  const data = require("./notifications.aspect");
  _notifications = function () {
    return data;
  };
  return data;
}
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
/**
 * extension
 */
class NotificationUI {
  static async provider([uiRuntimeExtension]) {
    return new NotificationUI(uiRuntimeExtension);
  }
  constructor(uiRuntimeExtension) {
    (0, _defineProperty2().default)(this, "dispatch", void 0);
    (0, _defineProperty2().default)(this, "add", (message, level) => {
      var _this$dispatch;
      const id = (0, _uuid().v1)();
      (_this$dispatch = this.dispatch) === null || _this$dispatch === void 0 ? void 0 : _this$dispatch.call(this, {
        type: 'add',
        content: {
          id,
          message,
          level,
          time: new Date().toISOString()
        }
      });
      return id;
    });
    (0, _defineProperty2().default)(this, "log", message => this.add(message, _uiFoundationUiNotifications3().MessageLevel.info));
    (0, _defineProperty2().default)(this, "warn", message => this.add(message, _uiFoundationUiNotifications3().MessageLevel.warning));
    (0, _defineProperty2().default)(this, "error", message => this.add(message, _uiFoundationUiNotifications3().MessageLevel.error));
    (0, _defineProperty2().default)(this, "success", message => this.add(message, _uiFoundationUiNotifications3().MessageLevel.success));
    (0, _defineProperty2().default)(this, "clear", () => {
      var _this$dispatch2;
      (_this$dispatch2 = this.dispatch) === null || _this$dispatch2 === void 0 ? void 0 : _this$dispatch2.call(this, {
        type: 'clear'
      });
    });
    (0, _defineProperty2().default)(this, "render", props => {
      // this code assumes a single place of render per instance of NotificationUI
      const [messages, dispatch] = (0, _react().useReducer)(_notificationReducer().notificationReducer, []);
      this.dispatch = dispatch;
      return /*#__PURE__*/_react().default.createElement(_uiFoundationUiNotifications2().NotificationCenter, (0, _extends2().default)({}, props, {
        notifications: messages
      }));
    });
    (0, _defineProperty2().default)(this, "renderContext", ({
      children
    }) => {
      return /*#__PURE__*/_react().default.createElement(_uiFoundationUiNotifications().NotificationContext.Provider, {
        value: this
      }, children);
    });
    uiRuntimeExtension.registerHudItem( /*#__PURE__*/_react().default.createElement(this.render, {
      key: "NotificationUI"
    }));
    uiRuntimeExtension.registerRenderHooks({
      reactContext: this.renderContext
    });
  }
  /** removes/archives a message from the log */
  dismiss(id) {
    var _this$dispatch3;
    (_this$dispatch3 = this.dispatch) === null || _this$dispatch3 === void 0 ? void 0 : _this$dispatch3.call(this, {
      type: 'dismiss',
      id
    });
  }

  /** adds a message with level "info" to the log */
}
exports.default = NotificationUI;
(0, _defineProperty2().default)(NotificationUI, "dependencies", [_ui().UIAspect]);
(0, _defineProperty2().default)(NotificationUI, "runtime", _ui().UIRuntime);
_notifications().NotificationsAspect.addRuntime(NotificationUI);
