// @teambit/express/dist/types/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Middleware", {
  enumerable: true,
  get: function () {
    return _route().Middleware;
  }
});
Object.defineProperty(exports, "NextFunction", {
  enumerable: true,
  get: function () {
    return _next().NextFunction;
  }
});
Object.defineProperty(exports, "Request", {
  enumerable: true,
  get: function () {
    return _request().Request;
  }
});
Object.defineProperty(exports, "Response", {
  enumerable: true,
  get: function () {
    return _response().Response;
  }
});
Object.defineProperty(exports, "Route", {
  enumerable: true,
  get: function () {
    return _route().Route;
  }
});
Object.defineProperty(exports, "Verb", {
  enumerable: true,
  get: function () {
    return _route().Verb;
  }
});
function _request() {
  const data = require("./request");
  _request = function () {
    return data;
  };
  return data;
}
function _response() {
  const data = require("./response");
  _response = function () {
    return data;
  };
  return data;
}
function _next() {
  const data = require("./next");
  _next = function () {
    return data;
  };
  return data;
}
function _route() {
  const data = require("./route");
  _route = function () {
    return data;
  };
  return data;
}
