// @teambit/html/dist/templates/html-component/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.htmlComponentTemplate = exports.deprecatedHtmlComponentTemplate = void 0;
function _component() {
  const data = require("./files/component");
  _component = function () {
    return data;
  };
  return data;
}
function _composition() {
  const data = require("./files/composition");
  _composition = function () {
    return data;
  };
  return data;
}
function _docs() {
  const data = require("./files/docs");
  _docs = function () {
    return data;
  };
  return data;
}
function _test() {
  const data = require("./files/test");
  _test = function () {
    return data;
  };
  return data;
}
const htmlComponentTemplate = {
  name: 'html',
  description: 'a basic html component',
  generateFiles: context => {
    const {
      name,
      namePascalCase: Name
    } = context;
    const indexFile = {
      relativePath: 'index.ts',
      content: `export { ${Name} } from './${name}';
export type { ${Name}Props } from './${name}';
`
    };
    return [indexFile, (0, _component().componentFile)(context), (0, _composition().compositionFile)(context), (0, _docs().docsFile)(context), (0, _test().testFile)(context)];
  },
  config: {
    'teambit.html/html': {},
    'teambit.envs/envs': {
      env: 'teambit.html/html'
    }
  }
};
exports.htmlComponentTemplate = htmlComponentTemplate;
const deprecatedHtmlComponentTemplate = {
  name: 'html-component',
  description: 'a basic html component',
  hidden: true,
  generateFiles: context => {
    const {
      name,
      namePascalCase: Name
    } = context;
    const indexFile = {
      relativePath: 'index.ts',
      content: `export { ${Name} } from './${name}';
export type { ${Name}Props } from './${name}';
`
    };
    return [indexFile, (0, _component().componentFile)(context), (0, _composition().compositionFile)(context), (0, _docs().docsFile)(context), (0, _test().testFile)(context)];
  },
  config: {
    'teambit.html/html': {},
    'teambit.envs/envs': {
      env: 'teambit.html/html'
    }
  }
};
exports.deprecatedHtmlComponentTemplate = deprecatedHtmlComponentTemplate;
