// @teambit/application/dist/app-context.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AppContext = void 0;
function _envs() {
  const data = require("@teambit/envs");
  _envs = function () {
    return data;
  };
  return data;
}
class AppContext extends _envs().ExecutionContext {
  constructor(appName,
  /**
   * determine whether to serve the application in dev mode.
   */
  dev, appComponent, workdir, execContext,
  /**
   * A path for the host root dir
   * Host root dir is the dir where we run the app from
   * This can be used in different bundle options which run require.resolve
   * for example when configuring webpack aliases or webpack expose loader on the peers deps
   */
  hostRootDir) {
    super(execContext.upper, execContext.envRuntime, execContext.components);
    this.appName = appName;
    this.dev = dev;
    this.appComponent = appComponent;
    this.workdir = workdir;
    this.hostRootDir = hostRootDir;
  }
  // static fromExecContext() {
  // return new AppContext();
  // }
}
exports.AppContext = AppContext;
