// @teambit/application/dist/app-type.plugin.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AppTypePlugin = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _bitError() {
  const data = require("@teambit/bit-error");
  _bitError = function () {
    return data;
  };
  return data;
}
function _cli() {
  const data = require("@teambit/cli");
  _cli = function () {
    return data;
  };
  return data;
}
class AppTypePlugin {
  constructor(pattern, appType, appSlot) {
    this.pattern = pattern;
    this.appType = appType;
    this.appSlot = appSlot;
    (0, _defineProperty2().default)(this, "runtimes", [_cli().MainRuntime.name]);
  }
  register(object) {
    const app = this.appType.createApp(object);
    this.validateApp(app);
    this.appSlot.register([app]);
  }
  validateApp(app) {
    if (app.name.includes(' ')) {
      throw new (_bitError().BitError)(`app name "${app.name}" is invalid. spaces are not permitted`);
    }
  }
}
exports.AppTypePlugin = AppTypePlugin;
