// @teambit/application/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "APPS_ARTIFACTS_DIR_NAME", {
  enumerable: true,
  get: function () {
    return _build().ARTIFACTS_DIR_NAME;
  }
});
Object.defineProperty(exports, "AppBuildContext", {
  enumerable: true,
  get: function () {
    return _appBuildContext().AppBuildContext;
  }
});
Object.defineProperty(exports, "AppBuildResult", {
  enumerable: true,
  get: function () {
    return _appBuildResult().AppBuildResult;
  }
});
Object.defineProperty(exports, "AppContext", {
  enumerable: true,
  get: function () {
    return _appContext().AppContext;
  }
});
Object.defineProperty(exports, "AppDeployContext", {
  enumerable: true,
  get: function () {
    return _appDeployContext().AppDeployContext;
  }
});
Object.defineProperty(exports, "ApplicationAspect", {
  enumerable: true,
  get: function () {
    return _application().ApplicationAspect;
  }
});
Object.defineProperty(exports, "ApplicationType", {
  enumerable: true,
  get: function () {
    return _applicationType().ApplicationType;
  }
});
Object.defineProperty(exports, "DeploymentProvider", {
  enumerable: true,
  get: function () {
    return _deploymentProvider().DeploymentProvider;
  }
});
function _application() {
  const data = require("./application.aspect");
  _application = function () {
    return data;
  };
  return data;
}
function _appContext() {
  const data = require("./app-context");
  _appContext = function () {
    return data;
  };
  return data;
}
function _deploymentProvider() {
  const data = require("./deployment-provider");
  _deploymentProvider = function () {
    return data;
  };
  return data;
}
function _applicationType() {
  const data = require("./application-type");
  _applicationType = function () {
    return data;
  };
  return data;
}
function _appDeployContext() {
  const data = require("./app-deploy-context");
  _appDeployContext = function () {
    return data;
  };
  return data;
}
function _appBuildContext() {
  const data = require("./app-build-context");
  _appBuildContext = function () {
    return data;
  };
  return data;
}
function _appBuildResult() {
  const data = require("./app-build-result");
  _appBuildResult = function () {
    return data;
  };
  return data;
}
function _build() {
  const data = require("./build.task");
  _build = function () {
    return data;
  };
  return data;
}
