// @teambit/dev-files/dist/dev-files.fragment.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.array.iterator.js");
require("core-js/modules/es.promise.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DevFilesFragment = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
class DevFilesFragment {
  constructor(devFiles) {
    this.devFiles = devFiles;
    (0, _defineProperty2().default)(this, "title", 'dev files');
    (0, _defineProperty2().default)(this, "weight", 7);
  }
  stringifyDevFiles(component) {
    const devFiles = this.devFiles.getDevFiles(component);
    const tuples = devFiles.toTupleArray();
    return tuples.map(([file, aspectId]) => {
      return `${file} (${aspectId})`;
    }).join('\n');
  }
  async renderRow(component) {
    return {
      title: this.title,
      content: this.stringifyDevFiles(component)
    };
  }
  async json(component) {
    return {
      title: this.title,
      json: this.devFiles.getDevFiles(component).toObject()
    };
  }
}
exports.DevFilesFragment = DevFilesFragment;
