// @teambit/graph/dist/ui/dependencies-graph/dependencies-graph.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DependenciesGraph = DependenciesGraph;
function _extends2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
  _extends2 = function () {
    return data;
  };
  return data;
}
function _objectWithoutProperties2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
  _objectWithoutProperties2 = function () {
    return data;
  };
  return data;
}
function _react() {
  const data = _interopRequireWildcard(require("react"));
  _react = function () {
    return data;
  };
  return data;
}
function _classnames() {
  const data = _interopRequireDefault(require("classnames"));
  _classnames = function () {
    return data;
  };
  return data;
}
function _reactFlowRenderer() {
  const data = _interopRequireWildcard(require("react-flow-renderer"));
  _reactFlowRenderer = function () {
    return data;
  };
  return data;
}
function _componentNode() {
  const data = require("../component-node");
  _componentNode = function () {
    return data;
  };
  return data;
}
function _calcElements() {
  const data = require("./calc-elements");
  _calcElements = function () {
    return data;
  };
  return data;
}
function _minimap() {
  const data = require("./minimap");
  _minimap = function () {
    return data;
  };
  return data;
}
function _graphContext() {
  const data = require("./graph-context");
  _graphContext = function () {
    return data;
  };
  return data;
}
function _dependenciesGraphModule() {
  const data = _interopRequireDefault(require("./dependencies-graph.module.scss"));
  _dependenciesGraphModule = function () {
    return data;
  };
  return data;
}
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function ComponentNodeContainer(props) {
  const {
    sourcePosition = _reactFlowRenderer().Position.Top,
    targetPosition = _reactFlowRenderer().Position.Bottom,
    data,
    id
  } = props;
  return /*#__PURE__*/_react().default.createElement("div", {
    key: id
  }, /*#__PURE__*/_react().default.createElement(_reactFlowRenderer().Handle, {
    type: "target",
    position: targetPosition,
    isConnectable: false
  }), /*#__PURE__*/_react().default.createElement(_reactFlowRenderer().Handle, {
    type: "source",
    position: sourcePosition,
    isConnectable: false
  }), /*#__PURE__*/_react().default.createElement(_componentNode().ComponentNode, {
    node: data.node,
    type: data.type
  }));
}

// @ts-ignore - incorrect NodeTypes https://github.com/wbkd/react-flow/issues/2101 (#5746)
const NodeTypes = {
  ComponentNode: ComponentNodeContainer
};
function DependenciesGraph(_ref) {
  let {
      graph,
      rootNode,
      componentWidgets,
      className,
      onLoad,
      children
    } = _ref,
    rest = (0, _objectWithoutProperties2().default)(_ref, ["graph", "rootNode", "componentWidgets", "className", "onLoad", "children"]);
  const graphRef = (0, _react().useRef)();
  const elements = (0, _calcElements().calcElements)(graph, {
    rootNode
  });
  const context = (0, _react().useMemo)(() => ({
    componentWidgets
  }), [componentWidgets]);
  const handleLoad = (0, _react().useCallback)(instance => {
    graphRef.current = instance;
    instance.fitView();
    onLoad === null || onLoad === void 0 ? void 0 : onLoad(instance);
  }, [onLoad]);

  // clear ref on unmount
  (0, _react().useEffect)(() => () => graphRef.current = undefined, []);
  (0, _react().useEffect)(() => {
    var _graphRef$current;
    (_graphRef$current = graphRef.current) === null || _graphRef$current === void 0 ? void 0 : _graphRef$current.fitView();
  }, [graph]);
  return /*#__PURE__*/_react().default.createElement(_graphContext().ComponentGraphContext.Provider, {
    value: context
  }, /*#__PURE__*/_react().default.createElement(_reactFlowRenderer().ReactFlowProvider, null, /*#__PURE__*/_react().default.createElement(_reactFlowRenderer().default, (0, _extends2().default)({
    draggable: false,
    nodesDraggable: true,
    selectNodesOnDrag: false,
    nodesConnectable: false,
    zoomOnDoubleClick: false,
    elementsSelectable: false,
    maxZoom: 1
  }, rest, {
    className: (0, _classnames().default)(_dependenciesGraphModule().default.graph, className),
    elements: elements,
    nodeTypes: NodeTypes,
    onLoad: handleLoad
  }), /*#__PURE__*/_react().default.createElement(_reactFlowRenderer().Background, null), /*#__PURE__*/_react().default.createElement(_reactFlowRenderer().Controls, {
    className: _dependenciesGraphModule().default.controls
  }), /*#__PURE__*/_react().default.createElement(_reactFlowRenderer().MiniMap, {
    nodeColor: _minimap().calcMinimapColors,
    className: _dependenciesGraphModule().default.minimap
  }), children)));
}
