// @teambit/graph/dist/ui/dependencies-graph/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ComponentGraph", {
  enumerable: true,
  get: function () {
    return _graphContext().ComponentGraph;
  }
});
Object.defineProperty(exports, "ComponentGraphContext", {
  enumerable: true,
  get: function () {
    return _graphContext().ComponentGraphContext;
  }
});
Object.defineProperty(exports, "DependenciesGraph", {
  enumerable: true,
  get: function () {
    return _dependenciesGraph().DependenciesGraph;
  }
});
Object.defineProperty(exports, "calcElements", {
  enumerable: true,
  get: function () {
    return _calcElements().calcElements;
  }
});
Object.defineProperty(exports, "calcLayout", {
  enumerable: true,
  get: function () {
    return _calcLayout().calcLayout;
  }
});
Object.defineProperty(exports, "calcMinimapColors", {
  enumerable: true,
  get: function () {
    return _minimap().calcMinimapColors;
  }
});
Object.defineProperty(exports, "depTypeToClass", {
  enumerable: true,
  get: function () {
    return _depEdge().depTypeToClass;
  }
});
Object.defineProperty(exports, "depTypeToLabel", {
  enumerable: true,
  get: function () {
    return _depEdge().depTypeToLabel;
  }
});
exports.styles = void 0;
function _dependenciesGraphModule() {
  const data = _interopRequireDefault(require("./dependencies-graph.module.scss"));
  _dependenciesGraphModule = function () {
    return data;
  };
  return data;
}
function _dependenciesGraph() {
  const data = require("./dependencies-graph");
  _dependenciesGraph = function () {
    return data;
  };
  return data;
}
function _graphContext() {
  const data = require("./graph-context");
  _graphContext = function () {
    return data;
  };
  return data;
}
function _depEdge() {
  const data = require("./dep-edge");
  _depEdge = function () {
    return data;
  };
  return data;
}
function _minimap() {
  const data = require("./minimap");
  _minimap = function () {
    return data;
  };
  return data;
}
function _calcLayout() {
  const data = require("./calc-layout");
  _calcLayout = function () {
    return data;
  };
  return data;
}
function _calcElements() {
  const data = require("./calc-elements");
  _calcElements = function () {
    return data;
  };
  return data;
}
const {
  graph,
  minimap,
  controls
} = _dependenciesGraphModule().default;
const styles = {
  graph,
  minimap,
  controls
};
exports.styles = styles;
