// @teambit/dependency-resolver/dist/dependencies/base-dependency.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BaseDependency = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
class BaseDependency {
  constructor(_id, _version, _lifecycle, _source) {
    this._id = _id;
    this._version = _version;
    this._lifecycle = _lifecycle;
    this._source = _source;
    (0, _defineProperty2().default)(this, "_type", void 0);
  }
  get id() {
    return this._id;
  }
  set id(newId) {
    this._id = newId;
  }
  get version() {
    return this._version;
  }
  get type() {
    return this._type;
  }
  get lifecycle() {
    return this._lifecycle;
  }
  get source() {
    return this._source;
  }
  set source(source) {
    this._source = source;
  }
  serialize() {
    return {
      id: this.id,
      version: this.version,
      __type: this.type,
      lifecycle: this.lifecycle.toString(),
      source: this.source
    };
  }
  setVersion(newVersion) {
    this._version = newVersion;
  }
  getPackageName() {
    return this.id;
  }
  toManifest() {
    var _this$getPackageName;
    const packageName = (_this$getPackageName = this.getPackageName) === null || _this$getPackageName === void 0 ? void 0 : _this$getPackageName.call(this);
    const version = this.version;
    return {
      packageName,
      version
    };
  }
}
exports.BaseDependency = BaseDependency;
