// @teambit/dependency-resolver/dist/dependencies/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RUNTIME_DEP_LIFECYCLE_TYPE = exports.ROOT_NAME = exports.PEER_DEP_LIFECYCLE_TYPE = exports.LIFECYCLE_TYPE_BY_KEY_NAME = exports.KEY_NAME_BY_LIFECYCLE_TYPE = exports.DEV_DEP_LIFECYCLE_TYPE = void 0;
const RUNTIME_DEP_LIFECYCLE_TYPE = 'runtime';
exports.RUNTIME_DEP_LIFECYCLE_TYPE = RUNTIME_DEP_LIFECYCLE_TYPE;
const DEV_DEP_LIFECYCLE_TYPE = 'dev';
exports.DEV_DEP_LIFECYCLE_TYPE = DEV_DEP_LIFECYCLE_TYPE;
const PEER_DEP_LIFECYCLE_TYPE = 'peer';
exports.PEER_DEP_LIFECYCLE_TYPE = PEER_DEP_LIFECYCLE_TYPE;
const ROOT_NAME = 'workspace';
exports.ROOT_NAME = ROOT_NAME;
const LIFECYCLE_TYPE_BY_KEY_NAME = {
  dependencies: RUNTIME_DEP_LIFECYCLE_TYPE,
  devDependencies: DEV_DEP_LIFECYCLE_TYPE,
  peerDependencies: PEER_DEP_LIFECYCLE_TYPE
};
exports.LIFECYCLE_TYPE_BY_KEY_NAME = LIFECYCLE_TYPE_BY_KEY_NAME;
const KEY_NAME_BY_LIFECYCLE_TYPE = {
  [RUNTIME_DEP_LIFECYCLE_TYPE]: 'dependencies',
  [DEV_DEP_LIFECYCLE_TYPE]: 'devDependencies',
  [PEER_DEP_LIFECYCLE_TYPE]: 'peerDependencies'
};
exports.KEY_NAME_BY_LIFECYCLE_TYPE = KEY_NAME_BY_LIFECYCLE_TYPE;
