// @teambit/dependency-resolver/dist/dependencies/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "BaseDependency", {
  enumerable: true,
  get: function () {
    return _baseDependency().BaseDependency;
  }
});
Object.defineProperty(exports, "COMPONENT_DEP_TYPE", {
  enumerable: true,
  get: function () {
    return _componentDependency().COMPONENT_DEP_TYPE;
  }
});
Object.defineProperty(exports, "ComponentDependency", {
  enumerable: true,
  get: function () {
    return _componentDependency().ComponentDependency;
  }
});
Object.defineProperty(exports, "ComponentDependencyFactory", {
  enumerable: true,
  get: function () {
    return _componentDependency().ComponentDependencyFactory;
  }
});
Object.defineProperty(exports, "DependenciesManifest", {
  enumerable: true,
  get: function () {
    return _dependencyList().DependenciesManifest;
  }
});
Object.defineProperty(exports, "Dependency", {
  enumerable: true,
  get: function () {
    return _dependency().Dependency;
  }
});
Object.defineProperty(exports, "DependencyFactory", {
  enumerable: true,
  get: function () {
    return _dependencyFactory().DependencyFactory;
  }
});
Object.defineProperty(exports, "DependencyLifecycleType", {
  enumerable: true,
  get: function () {
    return _dependency().DependencyLifecycleType;
  }
});
Object.defineProperty(exports, "DependencyList", {
  enumerable: true,
  get: function () {
    return _dependencyList().DependencyList;
  }
});
Object.defineProperty(exports, "DependencyListFactory", {
  enumerable: true,
  get: function () {
    return _dependencyListFactory().DependencyListFactory;
  }
});
Object.defineProperty(exports, "KEY_NAME_BY_LIFECYCLE_TYPE", {
  enumerable: true,
  get: function () {
    return _constants().KEY_NAME_BY_LIFECYCLE_TYPE;
  }
});
Object.defineProperty(exports, "LIFECYCLE_TYPE_BY_KEY_NAME", {
  enumerable: true,
  get: function () {
    return _constants().LIFECYCLE_TYPE_BY_KEY_NAME;
  }
});
Object.defineProperty(exports, "PackageName", {
  enumerable: true,
  get: function () {
    return _dependency().PackageName;
  }
});
Object.defineProperty(exports, "SemverVersion", {
  enumerable: true,
  get: function () {
    return _dependency().SemverVersion;
  }
});
Object.defineProperty(exports, "SerializedDependency", {
  enumerable: true,
  get: function () {
    return _dependency().SerializedDependency;
  }
});
Object.defineProperty(exports, "WorkspaceDependencyLifecycleType", {
  enumerable: true,
  get: function () {
    return _dependency().WorkspaceDependencyLifecycleType;
  }
});
function _dependencyList() {
  const data = require("./dependency-list");
  _dependencyList = function () {
    return data;
  };
  return data;
}
function _dependency() {
  const data = require("./dependency");
  _dependency = function () {
    return data;
  };
  return data;
}
function _baseDependency() {
  const data = require("./base-dependency");
  _baseDependency = function () {
    return data;
  };
  return data;
}
function _dependencyListFactory() {
  const data = require("./dependency-list-factory");
  _dependencyListFactory = function () {
    return data;
  };
  return data;
}
function _componentDependency() {
  const data = require("./component-dependency");
  _componentDependency = function () {
    return data;
  };
  return data;
}
function _dependencyFactory() {
  const data = require("./dependency-factory");
  _dependencyFactory = function () {
    return data;
  };
  return data;
}
function _constants() {
  const data = require("./constants");
  _constants = function () {
    return data;
  };
  return data;
}
