// @teambit/yarn/dist/yarn.ui.runtime.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.array.iterator.js");
require("core-js/modules/es.promise.js");
require("core-js/modules/es.regexp.exec.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.YarnUI = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _react() {
  const data = _interopRequireDefault(require("react"));
  _react = function () {
    return data;
  };
  return data;
}
function _ui() {
  const data = require("@teambit/ui");
  _ui = function () {
    return data;
  };
  return data;
}
function _uiFoundationUiUseBox() {
  const data = require("@teambit/ui-foundation.ui.use-box.menu");
  _uiFoundationUiUseBox = function () {
    return data;
  };
  return data;
}
function _component() {
  const data = _interopRequireDefault(require("@teambit/component"));
  _component = function () {
    return data;
  };
  return data;
}
function _yarn() {
  const data = require("./yarn.aspect");
  _yarn = function () {
    return data;
  };
  return data;
}
class YarnUI {
  constructor() {
    (0, _defineProperty2().default)(this, "consumeMethod", (comp, options) => {
      if (options !== null && options !== void 0 && options.currentLane) return undefined;
      const registry = comp.packageName.split('/')[0];
      const packageVersion = comp.version === comp.latest ? '' : `@${comp.version}`;
      return {
        Title: /*#__PURE__*/_react().default.createElement("img", {
          style: {
            height: '17px',
            paddingTop: '4px'
          },
          src: "https://static.bit.dev/brands/logo-yarn-text.svg"
        }),
        Component: /*#__PURE__*/_react().default.createElement(_uiFoundationUiUseBox().Install, {
          config: `npm config set '${registry}:registry' https://node.bit.cloud`,
          componentName: comp.id.name,
          packageManager: "yarn",
          copyString: `yarn add ${comp.packageName}${packageVersion}`,
          registryName: registry
        }),
        order: 20
      };
    });
  }
  static async provider([componentUI]) {
    const yarn = new YarnUI();
    componentUI.registerConsumeMethod(yarn.consumeMethod);
    return yarn;
  }
}
exports.YarnUI = YarnUI;
(0, _defineProperty2().default)(YarnUI, "runtime", _ui().UIRuntime);
(0, _defineProperty2().default)(YarnUI, "dependencies", [_component().default]);
var _default = YarnUI;
exports.default = _default;
_yarn().YarnAspect.addRuntime(YarnUI);
