// @teambit/component/dist/aspect-entry.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AspectEntry = void 0;
/**
 * avoid importing any (non-type) legacy code here. otherwise, PreviewTask will throw cryptic errors
 */

class AspectEntry {
  constructor(id, legacyEntry) {
    this.id = id;
    this.legacyEntry = legacyEntry;
  }
  get legacy() {
    return this.legacyEntry;
  }
  get isLegacy() {
    var _this$config;
    if ((_this$config = this.config) !== null && _this$config !== void 0 && _this$config.__legacy) return true;
    return false;
  }
  get config() {
    return this.legacy.config;
  }
  set config(config) {
    this.legacy.config = config;
  }
  get data() {
    return this.legacy.data;
  }
  get isRemoved() {
    return this.legacy.isRemoved;
  }

  // eslint-disable-next-line @typescript-eslint/adjacent-overload-signatures
  set data(val) {
    this.legacy.data = val;
  }
  transform(newData) {
    const newEntry = this.clone();
    newEntry.data = newData;
    return new AspectEntry(this.id, newEntry.legacy);
  }
  clone() {
    return new AspectEntry(this.id, this.legacyEntry.clone());
  }
  serialize() {
    return {
      id: this.id.toString(),
      config: this.config,
      data: this.data,
      icon: 'https://static.bit.dev/extensions-icons/default.svg' // TODO @gilad - once you connect the icon here please use this url as the default icon
    };
  }
}
exports.AspectEntry = AspectEntry;
