// @teambit/component/dist/aspect.section.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AspectSection = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _react() {
  const data = _interopRequireDefault(require("react"));
  _react = function () {
    return data;
  };
  return data;
}
function _uiFoundationUi() {
  const data = require("@teambit/ui-foundation.ui.menu-widget-icon");
  _uiFoundationUi = function () {
    return data;
  };
  return data;
}
function _aspectPage() {
  const data = require("./ui/aspect-page");
  _aspectPage = function () {
    return data;
  };
  return data;
}
class AspectSection {
  constructor() {
    (0, _defineProperty2().default)(this, "route", {
      path: '~aspect',
      element: /*#__PURE__*/_react().default.createElement(_aspectPage().AspectPage, null)
    });
    (0, _defineProperty2().default)(this, "navigationLink", {
      href: '~aspect',
      children: /*#__PURE__*/_react().default.createElement(_uiFoundationUi().MenuWidgetIcon, {
        icon: "configuration",
        tooltipContent: "Configuration"
      }),
      displayName: 'Configuration'
    });
    (0, _defineProperty2().default)(this, "order", 50);
  }
}
exports.AspectSection = AspectSection;
