// @teambit/component/dist/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AspectData", {
  enumerable: true,
  get: function () {
    return _aspectEntry().AspectData;
  }
});
Object.defineProperty(exports, "AspectEntry", {
  enumerable: true,
  get: function () {
    return _aspectEntry().AspectEntry;
  }
});
Object.defineProperty(exports, "Component", {
  enumerable: true,
  get: function () {
    return _component2().Component;
  }
});
Object.defineProperty(exports, "ComponentAspect", {
  enumerable: true,
  get: function () {
    return _component().ComponentAspect;
  }
});
Object.defineProperty(exports, "ComponentContext", {
  enumerable: true,
  get: function () {
    return _componentContext().ComponentContext;
  }
});
Object.defineProperty(exports, "ComponentDescriptorContext", {
  enumerable: true,
  get: function () {
    return _componentContext().ComponentDescriptorContext;
  }
});
Object.defineProperty(exports, "ComponentDescriptorProvider", {
  enumerable: true,
  get: function () {
    return _context().ComponentDescriptorProvider;
  }
});
Object.defineProperty(exports, "ComponentFS", {
  enumerable: true,
  get: function () {
    return _componentFs().default;
  }
});
Object.defineProperty(exports, "ComponentID", {
  enumerable: true,
  get: function () {
    return _componentId().ComponentID;
  }
});
Object.defineProperty(exports, "ComponentMap", {
  enumerable: true,
  get: function () {
    return _componentMap().ComponentMap;
  }
});
Object.defineProperty(exports, "ComponentModel", {
  enumerable: true,
  get: function () {
    return _componentModel().ComponentModel;
  }
});
Object.defineProperty(exports, "ComponentModelProps", {
  enumerable: true,
  get: function () {
    return _componentModel().ComponentModelProps;
  }
});
Object.defineProperty(exports, "ComponentProvider", {
  enumerable: true,
  get: function () {
    return _context().ComponentProvider;
  }
});
Object.defineProperty(exports, "ComponentUrlParams", {
  enumerable: true,
  get: function () {
    return _component3().ComponentUrlParams;
  }
});
Object.defineProperty(exports, "Config", {
  enumerable: true,
  get: function () {
    return _config().default;
  }
});
Object.defineProperty(exports, "ConsumePlugin", {
  enumerable: true,
  get: function () {
    return _menu().ConsumePlugin;
  }
});
Object.defineProperty(exports, "InvalidComponent", {
  enumerable: true,
  get: function () {
    return _component2().InvalidComponent;
  }
});
Object.defineProperty(exports, "RegisteredComponentRoute", {
  enumerable: true,
  get: function () {
    return _component3().RegisteredComponentRoute;
  }
});
Object.defineProperty(exports, "ResolveComponentIdFunc", {
  enumerable: true,
  get: function () {
    return _aspectEntry().ResolveComponentIdFunc;
  }
});
Object.defineProperty(exports, "Section", {
  enumerable: true,
  get: function () {
    return _section().Section;
  }
});
Object.defineProperty(exports, "Snap", {
  enumerable: true,
  get: function () {
    return _snap().Snap;
  }
});
Object.defineProperty(exports, "SnapProps", {
  enumerable: true,
  get: function () {
    return _snap().SnapProps;
  }
});
Object.defineProperty(exports, "State", {
  enumerable: true,
  get: function () {
    return _state().State;
  }
});
Object.defineProperty(exports, "Tag", {
  enumerable: true,
  get: function () {
    return _tag().Tag;
  }
});
Object.defineProperty(exports, "TagMap", {
  enumerable: true,
  get: function () {
    return _tagMap().TagMap;
  }
});
Object.defineProperty(exports, "TagProps", {
  enumerable: true,
  get: function () {
    return _tag().TagProps;
  }
});
Object.defineProperty(exports, "TopBarNav", {
  enumerable: true,
  get: function () {
    return _topBarNav().TopBarNav;
  }
});
Object.defineProperty(exports, "componentFields", {
  enumerable: true,
  get: function () {
    return _ui().componentFields;
  }
});
Object.defineProperty(exports, "componentIdFields", {
  enumerable: true,
  get: function () {
    return _ui().componentIdFields;
  }
});
Object.defineProperty(exports, "componentOverviewFields", {
  enumerable: true,
  get: function () {
    return _ui().componentOverviewFields;
  }
});
exports.default = void 0;
Object.defineProperty(exports, "useComponent", {
  enumerable: true,
  get: function () {
    return _ui().useComponent;
  }
});
Object.defineProperty(exports, "useComponentDescriptor", {
  enumerable: true,
  get: function () {
    return _componentContext().useComponentDescriptor;
  }
});
Object.defineProperty(exports, "useComponentHost", {
  enumerable: true,
  get: function () {
    return _host().useComponentHost;
  }
});
Object.defineProperty(exports, "useIdFromLocation", {
  enumerable: true,
  get: function () {
    return _ui().useIdFromLocation;
  }
});
function _component() {
  const data = require("./component.aspect");
  _component = function () {
    return data;
  };
  return data;
}
function _host() {
  const data = require("./host");
  _host = function () {
    return data;
  };
  return data;
}
function _component2() {
  const data = require("./component");
  _component2 = function () {
    return data;
  };
  return data;
}
function _componentId() {
  const data = require("@teambit/component-id");
  _componentId = function () {
    return data;
  };
  return data;
}
function _componentFs() {
  const data = _interopRequireDefault(require("./component-fs"));
  _componentFs = function () {
    return data;
  };
  return data;
}
function _aspectEntry() {
  const data = require("./aspect-entry");
  _aspectEntry = function () {
    return data;
  };
  return data;
}
function _snap() {
  const data = require("./snap/snap");
  _snap = function () {
    return data;
  };
  return data;
}
function _tag() {
  const data = require("./tag/tag");
  _tag = function () {
    return data;
  };
  return data;
}
function _state() {
  const data = require("./state");
  _state = function () {
    return data;
  };
  return data;
}
function _tagMap() {
  const data = require("./tag-map");
  _tagMap = function () {
    return data;
  };
  return data;
}
function _componentMap() {
  const data = require("./component-map");
  _componentMap = function () {
    return data;
  };
  return data;
}
function _section() {
  const data = require("./section");
  _section = function () {
    return data;
  };
  return data;
}
function _componentContext() {
  const data = require("./ui/context/component-context");
  _componentContext = function () {
    return data;
  };
  return data;
}
function _context() {
  const data = require("./ui/context");
  _context = function () {
    return data;
  };
  return data;
}
function _ui() {
  const data = require("./ui");
  _ui = function () {
    return data;
  };
  return data;
}
function _menu() {
  const data = require("./ui/menu");
  _menu = function () {
    return data;
  };
  return data;
}
function _component3() {
  const data = require("./component.route");
  _component3 = function () {
    return data;
  };
  return data;
}
function _componentModel() {
  const data = require("./ui/component-model");
  _componentModel = function () {
    return data;
  };
  return data;
}
function _topBarNav() {
  const data = require("./ui/top-bar-nav");
  _topBarNav = function () {
    return data;
  };
  return data;
}
function _config() {
  const data = _interopRequireDefault(require("./config"));
  _config = function () {
    return data;
  };
  return data;
}
// TODO: check why it's not working when using the index in snap dir like this:
// export { Snap, Author } from './snap';
// TODO: check why it's not working when using the index in tag dir like this:
// export { Tag } from './tag';
// export { AspectList } from './aspect-list';
// export { AspectEntry } from './aspect-entry';
var _default = _component().ComponentAspect;
exports.default = _default;
