// @teambit/compositions/dist/composition.section.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CompositionsSection = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _react() {
  const data = _interopRequireDefault(require("react"));
  _react = function () {
    return data;
  };
  return data;
}
function _compositions() {
  const data = require("./compositions");
  _compositions = function () {
    return data;
  };
  return data;
}
class CompositionsSection {
  constructor(
  /**
   * simulations ui extension.
   */
  compositions, options, emptyStateSlot) {
    this.compositions = compositions;
    this.options = options;
    this.emptyStateSlot = emptyStateSlot;
    (0, _defineProperty2().default)(this, "navigationLink", {
      href: '~compositions',
      children: 'Compositions'
    });
    (0, _defineProperty2().default)(this, "route", {
      path: '~compositions',
      element: /*#__PURE__*/_react().default.createElement(_compositions().Compositions, {
        menuBarWidgets: this.options.menuBarWidgetSlot,
        emptyState: this.emptyStateSlot
      })
    });
    (0, _defineProperty2().default)(this, "order", 20);
  }
}
exports.CompositionsSection = CompositionsSection;
