// @teambit/compositions/dist/composition.compare.section.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CompositionCompareSection = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _react() {
  const data = _interopRequireDefault(require("react"));
  _react = function () {
    return data;
  };
  return data;
}
function _compositionsUi() {
  const data = require("@teambit/compositions.ui.composition-compare");
  _compositionsUi = function () {
    return data;
  };
  return data;
}
class CompositionCompareSection {
  constructor(emptyStateSlot) {
    this.emptyStateSlot = emptyStateSlot;
    (0, _defineProperty2().default)(this, "navigationLink", {
      href: 'compositions',
      children: 'Compositions',
      order: 1
    });
    (0, _defineProperty2().default)(this, "route", {
      path: 'compositions/*',
      element: /*#__PURE__*/_react().default.createElement(_compositionsUi().CompositionCompare, {
        emptyState: this.emptyStateSlot
      })
    });
  }
}
exports.CompositionCompareSection = CompositionCompareSection;
