// @teambit/compositions/dist/compositions.preview-definition.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.promise.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CompositionPreviewDefinition = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
class CompositionPreviewDefinition {
  constructor(compositions) {
    this.compositions = compositions;
    (0, _defineProperty2().default)(this, "prefix", 'compositions');
    (0, _defineProperty2().default)(this, "includePeers", true);
  }
  async renderTemplatePath(context) {
    return this.renderTemplatePathByEnv(context.env);
  }
  async renderTemplatePathByEnv(env) {
    return env.getMounter();
  }
  async getModuleMap(components) {
    const map = this.compositions.getPreviewFiles(components);
    return map;
  }
  async getMetadata(component) {
    const compositions = this.compositions.getCompositions(component).map(composition => ({
      displayName: composition.displayName,
      identifier: composition.identifier
    }));
    return {
      compositions
    };
  }
}
exports.CompositionPreviewDefinition = CompositionPreviewDefinition;
