// @teambit/compositions/dist/composition.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Composition = void 0;
function _compositionsModel() {
  const data = require("@teambit/compositions.model.composition-id");
  _compositionsModel = function () {
    return data;
  };
  return data;
}
class Composition {
  constructor(
  /**
   * identifier of the composition
   */
  identifier,
  /**
   * file path in which the composition is contained.
   */
  filepath,
  /**
   * set explicit display name
   */
  _displayName) {
    this.identifier = identifier;
    this.filepath = filepath;
    this._displayName = _displayName;
  }
  get displayName() {
    return this._displayName || (0, _compositionsModel().humanizeCompositionId)(this.identifier);
  }
  toObject() {
    return {
      identifier: this.identifier,
      filepath: this.filepath,
      displayName: this._displayName
    };
  }
  toBrowserMetaDataObject() {
    return {
      identifier: this.identifier,
      displayName: this._displayName
    };
  }
  static fromArray(compositions) {
    return compositions.map(composition => {
      return new Composition(composition.identifier, composition.filepath, composition.displayName);
    });
  }
}
exports.Composition = Composition;
