// @teambit/compositions/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ComponentComposition", {
  enumerable: true,
  get: function () {
    return _ui().ComponentComposition;
  }
});
Object.defineProperty(exports, "Composition", {
  enumerable: true,
  get: function () {
    return _composition().Composition;
  }
});
Object.defineProperty(exports, "CompositionContent", {
  enumerable: true,
  get: function () {
    return _compositions2().CompositionContent;
  }
});
Object.defineProperty(exports, "CompositionContentProps", {
  enumerable: true,
  get: function () {
    return _compositions2().CompositionContentProps;
  }
});
Object.defineProperty(exports, "CompositionProps", {
  enumerable: true,
  get: function () {
    return _composition().CompositionProps;
  }
});
Object.defineProperty(exports, "CompositionsAspect", {
  enumerable: true,
  get: function () {
    return _compositions().CompositionsAspect;
  }
});
exports.default = void 0;
function _compositions() {
  const data = require("./compositions.aspect");
  _compositions = function () {
    return data;
  };
  return data;
}
function _composition() {
  const data = require("./composition");
  _composition = function () {
    return data;
  };
  return data;
}
function _compositions2() {
  const data = require("./compositions");
  _compositions2 = function () {
    return data;
  };
  return data;
}
function _ui() {
  const data = require("./ui");
  _ui = function () {
    return data;
  };
  return data;
}
var _default = _compositions().CompositionsAspect;
exports.default = _default;
