// @teambit/legacy/dist/scope/models/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ExportMetadata", {
  enumerable: true,
  get: function () {
    return _exportMetadata().default;
  }
});
Object.defineProperty(exports, "Lane", {
  enumerable: true,
  get: function () {
    return _lane().default;
  }
});
Object.defineProperty(exports, "ModelComponent", {
  enumerable: true,
  get: function () {
    return _modelComponent().default;
  }
});
Object.defineProperty(exports, "ScopeMeta", {
  enumerable: true,
  get: function () {
    return _scopeMeta().default;
  }
});
Object.defineProperty(exports, "Source", {
  enumerable: true,
  get: function () {
    return _source().default;
  }
});
Object.defineProperty(exports, "Symlink", {
  enumerable: true,
  get: function () {
    return _symlink().default;
  }
});
Object.defineProperty(exports, "Version", {
  enumerable: true,
  get: function () {
    return _version().default;
  }
});
function _lane() {
  const data = _interopRequireDefault(require("./lane"));
  _lane = function () {
    return data;
  };
  return data;
}
function _modelComponent() {
  const data = _interopRequireDefault(require("./model-component"));
  _modelComponent = function () {
    return data;
  };
  return data;
}
function _scopeMeta() {
  const data = _interopRequireDefault(require("./scopeMeta"));
  _scopeMeta = function () {
    return data;
  };
  return data;
}
function _source() {
  const data = _interopRequireDefault(require("./source"));
  _source = function () {
    return data;
  };
  return data;
}
function _symlink() {
  const data = _interopRequireDefault(require("./symlink"));
  _symlink = function () {
    return data;
  };
  return data;
}
function _version() {
  const data = _interopRequireDefault(require("./version"));
  _version = function () {
    return data;
  };
  return data;
}
function _exportMetadata() {
  const data = _interopRequireDefault(require("./export-metadata"));
  _exportMetadata = function () {
    return data;
  };
  return data;
}
