// @teambit/webpack/dist/transformers/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "BodyInjectionOptions", {
  enumerable: true,
  get: function () {
    return _injectBody().BodyInjectionOptions;
  }
});
Object.defineProperty(exports, "GenerateBodyInjectionTransformer", {
  enumerable: true,
  get: function () {
    return _injectBody().GenerateBodyInjectionTransformer;
  }
});
Object.defineProperty(exports, "GenerateHeadInjectionTransformer", {
  enumerable: true,
  get: function () {
    return _injectHead().GenerateHeadInjectionTransformer;
  }
});
Object.defineProperty(exports, "HeadInjectionOptions", {
  enumerable: true,
  get: function () {
    return _injectHead().HeadInjectionOptions;
  }
});
Object.defineProperty(exports, "generateAddAliasesFromPeersTransformer", {
  enumerable: true,
  get: function () {
    return _transformers().generateAddAliasesFromPeersTransformer;
  }
});
Object.defineProperty(exports, "generateExposePeersTransformer", {
  enumerable: true,
  get: function () {
    return _transformers().generateExposePeersTransformer;
  }
});
Object.defineProperty(exports, "generateExternalsTransformer", {
  enumerable: true,
  get: function () {
    return _transformers().generateExternalsTransformer;
  }
});
function _injectBody() {
  const data = require("./inject-body");
  _injectBody = function () {
    return data;
  };
  return data;
}
function _injectHead() {
  const data = require("./inject-head");
  _injectHead = function () {
    return data;
  };
  return data;
}
function _transformers() {
  const data = require("./transformers");
  _transformers = function () {
    return data;
  };
  return data;
}
