// @teambit/schema/dist/schema.graphql.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
require("core-js/modules/es.promise.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.schemaSchema = schemaSchema;
function _graphqlTypeJson() {
  const data = require("graphql-type-json");
  _graphqlTypeJson = function () {
    return data;
  };
  return data;
}
function _graphqlTag() {
  const data = _interopRequireDefault(require("graphql-tag"));
  _graphqlTag = function () {
    return data;
  };
  return data;
}
function schemaSchema(schema) {
  return {
    typeDefs: (0, _graphqlTag().default)`
      scalar JSONObject
      extend type ComponentHost {
        getSchema(id: String!): JSONObject
      }
    `,
    resolvers: {
      JSONObject: _graphqlTypeJson().GraphQLJSONObject,
      ComponentHost: {
        getSchema: async (host, {
          id
        }) => {
          const componentId = await host.resolveComponentId(id);
          const component = await host.get(componentId);
          const empty = {
            exports: []
          };
          if (!component) return empty;
          const docs = await schema.getSchema(component);
          if (!docs) return empty;
          return docs;
        }
      }
    }
  };
}
