// @teambit/legacy/dist/toolbox/timer/timer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Timer = void 0;
function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
  _defineProperty2 = function () {
    return data;
  };
  return data;
}
function _exceptions() {
  const data = require("./exceptions");
  _exceptions = function () {
    return data;
  };
  return data;
}
function _response() {
  const data = require("./response");
  _response = function () {
    return data;
  };
  return data;
}
class Timer {
  constructor() {
    (0, _defineProperty2().default)(this, "startTime", null);
  }
  /**
   * start the timer.
   */
  start() {
    if (this.startTime) throw new (_exceptions().TimerAlreadyRunning)();
    this.startTime = Date.now();
    return this;
  }

  /**
   * stop the timer and return timer results.
   */
  stop() {
    if (!this.startTime) throw new (_exceptions().TimerNotStarted)();
    const endTime = Date.now();
    return new (_response().TimerResponse)(this.calculateElapsed(this.startTime, endTime));
  }
  calculateElapsed(startTime, endTime) {
    return endTime - startTime;
  }

  /**
   * create a new timer instance.
   */
  static create() {
    return new Timer();
  }
}
exports.Timer = Timer;
